{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "from utils import config, sample_utils as su\n",
    "from utils import parse_midas_data, parse_patric, core_gene_utils, midas_db_utils\n",
    "import numpy as np, pickle, sys\n",
    "from numpy.random import choice\n",
    "from collections import defaultdict\n",
    "\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib.patches import Patch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "plot_dir = \"%s/revs_genes/\" % (config.analysis_directory)\n",
    "ddir = config.data_directory\n",
    "pdir = \"%s/pickles\" % ddir\n",
    "sweep_type = 'full'\n",
    "snp_modifications_by_site = pickle.load(open('%s/snp_modifications_by_site_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "snp_modification_genes = pickle.load(open('%s/snp_modification_genes_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "modification_gene_gains = pickle.load(open('%s/modification_gene_gains_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "modification_gene_losses = pickle.load(open('%s/modification_gene_losses_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "modification_gene_gains_present_null = pickle.load(open('%s/modification_gene_gains_present_null_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "modification_gene_losses_present_null = pickle.load(open('%s/modification_gene_losses_present_null_%s.pkl' % (pdir, sweep_type), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "keywords={}\n",
    "keywords['ABC transporter']=['ABC']\n",
    "keywords['phage']=['hage']\n",
    "keywords['transposon']=['onjugati','anspos']\n",
    "keywords['mobilization']=['mob','obilization','obile']\n",
    "keywords['integrase']=['ntegrase']\n",
    "keywords['plasmid']=['plasmid']\n",
    "keywords['recombinase']=['ecombinase']\n",
    "keywords['tRNA']=['tRNA']\n",
    "keywords['ATP']=['ATP']\n",
    "keywords['excisionase']=['xcisionase']\n",
    "keywords['transmembrane']=['embrane']\n",
    "keywords['replication']=['eplication']\n",
    "keywords['regulator']=['egulator']\n",
    "keywords['transcription']=['anscription']\n",
    "keywords['toxin']=['toxin']\n",
    "keywords['restriction']=['estriction']\n",
    "keywords['replication']=['eplication']\n",
    "keywords['transferase']=['ansferase']\n",
    "keywords['reductase']=['eductase']\n",
    "keywords['phosphatase']=['phosphatase']\n",
    "keywords['helicase']=['elicase']\n",
    "keywords['kinase']=['kinase']\n",
    "keywords['dehydrogenase']=['dehydrogenase']\n",
    "keywords['drug']=['drug']\n",
    "keywords['cell wall']=['ell wall']\n",
    "keywords['primase']=['imase']\n",
    "keywords['resistance']=['resistance']\n",
    "keywords['hydrolase']=['ydrolase']\n",
    "keywords['topoisomerase']=['opoisomerase']\n",
    "keywords['hypothetical'] = ['ypothetical']\n",
    "\n",
    "# since this is a greedy algorithm, order the more important keywords first\n",
    "keyword_order=['ABC transporter','phage','transposon','mobilization','integrase', 'plasmid','recombinase','tRNA','ATP','excisionase','transmembrane','replication','regulator','transcription','toxin','restriction','replication','transferase','reductase','phosphatase','helicase','kinase','dehydrogenase','drug','cell wall','primase','topoisomerase','hypothetical']\n",
    "\n",
    "common_genes={}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "new_dict = defaultdict(int)\n",
    "\n",
    "for keyword in keyword_order:\n",
    "    for gene_name in changed_gene_name_counts_by_tp_type['AA']:\n",
    "        if keyword in gene_name:\n",
    "            new_dict[keyword] += changed_gene_name_counts_by_tp_type['AA'][gene_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'ABC transporter': 27,\n",
       "             'ATP': 11,\n",
       "             'dehydrogenase': 2,\n",
       "             'drug': 2,\n",
       "             'hypothetical': 77,\n",
       "             'kinase': 6,\n",
       "             'phosphatase': 1,\n",
       "             'reductase': 2,\n",
       "             'regulator': 11,\n",
       "             'tRNA': 4,\n",
       "             'transcription': 1,\n",
       "             'transferase': 9})"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "tp_types = ['AA', 'II-premie', 'II-nonpremie','MM', 'MI-premie', 'MI-nonpremie']\n",
    "num_gene_gains_by_tp_type = {tp_type: 0 for tp_type in tp_types}\n",
    "gained_genes_by_tp_type = {tp_type: set() for tp_type in tp_types}\n",
    "gained_genes_counts_by_tp_type = {tp_type: defaultdict(int) for tp_type in tp_types}\n",
    "gained_genes_counts_nulls_by_tp_type = {tp_type: defaultdict(int) for tp_type in tp_types}\n",
    "\n",
    "# Enumerate genes\n",
    "for species in modification_gene_gains:\n",
    "    for cohort in modification_gene_gains[species]:\n",
    "        for subject in modification_gene_gains[species][cohort]:\n",
    "            subdict = modification_gene_gains[species][cohort][subject]\n",
    "            subdict_null = modification_gene_gains_present_null[species][cohort][subject]\n",
    "            # For each sample pair in this cohort/species,\n",
    "            # have a count of how many gains of a particular gene\n",
    "            for tp_pair in subdict:\n",
    "                tp_type = tp_pair_to_tp_type(tp_pair, cohort)\n",
    "                for gene_id, gene_name in subdict[tp_pair]:\n",
    "                    gene_gain_count = subdict[tp_pair][(gene_id, gene_name)]\n",
    "                    gene_gain_count_null = subdict_null[tp_pair][(gene_id, gene_name)]\n",
    "                    num_gene_gains_by_tp_type[tp_type] += gene_gain_count\n",
    "                    gained_genes_by_tp_type[tp_type].add(gene_name)\n",
    "                    gained_genes_counts_by_tp_type[tp_type][gene_name] += gene_gain_count\n",
    "                    gained_genes_counts_nulls_by_tp_type[tp_type][gene_name] += gene_gain_count_null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AA': 4,\n",
       " 'II-nonpremie': 3555,\n",
       " 'II-premie': 93,\n",
       " 'MI-nonpremie': 1,\n",
       " 'MI-premie': 0,\n",
       " 'MM': 0}"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_gene_gains_by_tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 AA\n",
      "30 II-premie\n",
      "0 MI-premie\n",
      "0 MM\n",
      "1 MI-nonpremie\n",
      "637 II-nonpremie\n"
     ]
    }
   ],
   "source": [
    "for tp_type in gained_genes_by_tp_type:\n",
    "    count = len(gained_genes_by_tp_type[tp_type])\n",
    "    print('%i %s' % (count, tp_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA\n",
      "4 hypothetical protein\n",
      "\n",
      "II-premie\n",
      "61 hypothetical protein\n",
      "3 Phage protein\n",
      "2 Doubtful CDS. No database matches\n",
      "\n",
      "MI-premie\n",
      "\n",
      "MM\n",
      "\n",
      "MI-nonpremie\n",
      "\n",
      "II-nonpremie\n",
      "9 RNA-binding protein\n",
      "9 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "8 Putative helicase\n",
      "8 Integrase\n",
      "7 Uncharacterized MFS-type transporter\n",
      "6 cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases\n",
      "6 UpdZ protein\n",
      "6 Two-component transcriptional response regulator, LuxR family\n",
      "6 Glycosyltransferase\n",
      "5 SclB protein\n",
      "5 Putative glycosyltransferase\n",
      "5 O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)\n",
      "5 N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)\n",
      "5 Capsular polysaccharide transcription antitermination protein, UpxY family\n",
      "5 ABC-type antimicrobial peptide transport system, ATPase component\n",
      "46 RNA polymerase ECF-type sigma factor\n",
      "4 beta-glycosyl hydrolase\n",
      "4 [FeFe]-hydrogenase maturation protein HydF\n",
      "4 Vitamin B12 ABC transporter, substrate-binding protein BtuF\n",
      "4 Vitamin B12 ABC transporter, permease protein BtuC\n",
      "4 Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)\n",
      "4 UDP-glucose 4-epimerase (EC 5.1.3.2)\n",
      "4 Tyrosine type site-specific recombinase\n",
      "4 TPR-repeat-containing protein\n",
      "4 Putative periplasmic protein\n",
      "4 Phage integrase\n",
      "4 Oxidoreductase, aldo/keto reductase family\n",
      "4 Nitroreductase family protein\n",
      "4 N-acetylglucosaminyltransferase (EC 2.4.1.-)\n",
      "4 Monofunctional biosynthetic peptidoglycan transglycosylase\n",
      "4 Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids\n",
      "4 MBL-fold metallo-hydrolase superfamily\n",
      "4 ISNCY family transposase\n",
      "4 Flavoredoxin\n",
      "4 Endonuclease\n",
      "4 Efflux ABC transporter, permease protein\n",
      "4 DNA ligase (NAD(+)) (EC 6.5.1.2)\n",
      "4 CAAX amino terminal protease family protein\n",
      "4 3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)\n",
      "3 tmRNA-binding protein SmpB\n",
      "3 tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)\n",
      "3 putative patatin-like phospholipase\n",
      "3 putative integration host factor IHF alpha subunit\n",
      "3 dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)\n",
      "3 cytochrome c biogenesis protein\n",
      "3 Zn-ribbon-containing, possibly RNA-binding protein and truncated derivatives\n",
      "3 Xylulose kinase (EC 2.7.1.17)\n",
      "3 Xylanase\n",
      "3 Uronate isomerase (EC 5.3.1.12)\n",
      "3 Uncharacterized protein YgbA\n",
      "3 Uncharacterized protein BT4380\n",
      "3 Uncharacterized membrane protein PA1577\n",
      "3 UPF0126 inner membrane protein BT_1201\n",
      "3 UPF0056 inner membrane protein MarC\n",
      "3 Two-component transcriptional response regulator, LuxR family BT1217\n",
      "3 Transmembrane protein\n",
      "3 Transcriptional regulator, AsnC family\n",
      "3 TonB family protein\n",
      "3 Tll0814 protein\n",
      "3 Thiol:disulfide interchange protein\n",
      "3 Stress responsive alpha-beta barrel domain protein Dabb\n",
      "3 Small heat shock protein\n",
      "3 Slr1117 protein\n",
      "3 Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)\n",
      "3 Sigma factor RpoE regulatory protein RseC\n",
      "3 Serine protease\n",
      "3 Rrf2-linked NADH-flavin reductase\n",
      "3 RidA/YER057c/UK114 superfamily protein\n",
      "3 Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria\n",
      "3 Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)\n",
      "3 Ribosomal silencing factor RsfA\n",
      "3 ROK family sugar kinase or transcriptional regulator\n",
      "3 RND efflux system, inner membrane transporter\n",
      "3 RNA methyltransferase, TrmH family\n",
      "3 Putative thioredoxin\n",
      "3 Putative permease\n",
      "3 Putative outer membrane protein, probably involved in nutrient binding\n",
      "3 Protein DUF2149, predicted transporter component\n",
      "3 Probable iron export ATP-binding protein FetA\n",
      "3 Probable integral membrane protein Cj1165c\n",
      "3 Possible glyoxylase family protein (Lactoylglutathione lyase) (EC 4.4.1.5)\n",
      "3 Possible exported heavy-metal binding protein\n",
      "3 Phosphoribosylformimino-5-aminoimidazole carboxamide ribotide isomerase (EC 5.3.1.16)\n",
      "3 Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO\n",
      "3 Permease of the drug/metabolite transporter (DMT) superfamily\n",
      "3 Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)\n",
      "3 PAP2 superfamily protein\n",
      "3 Oxidoreductase\n",
      "3 Outer membrane protein romA\n",
      "3 OmpA family outer membrane protein\n",
      "3 Na+/H+-dicarboxylate symporter\n",
      "3 N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)\n",
      "3 N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)\n",
      "3 Multidrug resistance protein\n",
      "3 Mobile element protein\n",
      "3 Membrane protein\n",
      "3 Malonyl CoA-acyl carrier protein transacylase (EC 2.3.1.39)\n",
      "3 Lipoprotein signal peptidase (EC 3.4.23.36)\n",
      "3 Large-conductance mechanosensitive channel\n",
      "3 Integration host factor alpha subunit\n",
      "3 Integrase IntN1\n",
      "3 Integral membrane protein\n",
      "3 Inner membrane protein YhiM\n",
      "3 Indole-3-glycerol phosphate synthase (EC 4.1.1.48)\n",
      "3 Immunoreactive 53 kDa antigen PG123\n",
      "3 Hypothetical Nudix-like regulator\n",
      "3 Helicase, C-terminal:Type III restriction enzyme, res subunit:DEAD/DEAH box helicase, N-terminal\n",
      "3 Glycosyl transferase, family 2\n",
      "3 GTP cyclohydrolase 1 type 2 homolog YbgI @ Bsu YqfO NIF3/CutA domain\n",
      "3 Free methionine-(R)-sulfoxide reductase, contains GAF domain\n",
      "3 Fosmidomycin resistance protein\n",
      "3 FIG137478: Hypothetical protein\n",
      "3 FIG016027: protein of unknown function YeaO\n",
      "3 FIG00490011: hypothetical protein\n",
      "3 FIG000605: protein co-occurring with transport systems (COG1739)\n",
      "3 Endo-1,4-beta-xylanase D\n",
      "3 DoxX family protein\n",
      "3 Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)\n",
      "3 Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)\n",
      "3 Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)\n",
      "3 Diacylglycerol kinase (EC 2.7.1.107)\n",
      "3 DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)\n",
      "3 DNA polymerase III delta subunit (EC 2.7.7.7)\n",
      "3 D-alanine--poly(phosphoribitol) ligase ACP subunit (EC 6.1.1.13)\n",
      "3 CzcABC family efflux RND transporter, outer membrane protein\n",
      "3 Cys-tRNA(Pro) deacylase YbaK\n",
      "3 Conserved protein YuxK\n",
      "3 Conserved hypothetical protein 22\n",
      "3 Co-activator of prophage gene expression IbrB\n",
      "3 Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage\n",
      "3 Aminotransferase, DegT/DnrJ/EryC1/StrS family\n",
      "3 Aminodeoxychorismate lyase (EC 4.1.3.38)\n",
      "3 Aerotolerance protein BatE\n",
      "3 Aerotolerance protein BatD\n",
      "3 Aerotolerance protein BatC\n",
      "3 Aerotolerance protein BatB\n",
      "3 Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)\n",
      "3 Acetyltransferase, GNAT family\n",
      "3 ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)\n",
      "3 ATP-dependent RNA helicase\n",
      "3 ATP synthase protein I\n",
      "3 ATP synthase gamma chain (EC 3.6.3.14)\n",
      "3 ATP synthase epsilon chain (EC 3.6.3.14)\n",
      "3 ATP synthase F0 sector subunit c (EC 3.6.3.14)\n",
      "3 4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)\n",
      "3 4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)\n",
      "3 2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)\n",
      "2262 hypothetical protein\n",
      "2 tRNA(1)(Val) (adenine(37)-N(6))-methyltransferase (EC 2.1.1.223)\n",
      "2 tRNA (5-methylaminomethyl-2-thiouridylate)-methyltransferase (EC 2.1.1.61)\n",
      "2 sodium/iodide co-transporter\n",
      "2 putative pyrogenic exotoxin B\n",
      "2 putative lipoprotein\n",
      "2 putative acetyltransferase\n",
      "2 putative UndPP-QuiNAc-P-transferase\n",
      "2 putative N-acetylmuramoyl-L-alanine amidase\n",
      "2 probable lipoprotein YPO2292\n",
      "2 mobilization protein B\n",
      "2 low molecular weight phosphatase supefamily\n",
      "2 dTDP-4-dehydrorhamnose reductase (EC 1.1.1.133)\n",
      "2 archaeal ATPase, fused to C-terminal DUF234 domain\n",
      "2 Ync\n",
      "2 V-type ATP synthase subunit K (EC 3.6.3.14)\n",
      "2 V-type ATP synthase subunit E (EC 3.6.3.14)\n",
      "2 UspA\n",
      "2 Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)\n",
      "2 Uracil permease @ Uracil:proton symporter UraA\n",
      "2 Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)\n",
      "2 Uncharacterized protein CHU_1311\n",
      "2 UPF0324 inner membrane protein YeiH\n",
      "2 UDP-glucuronate decarboxylase (EC 4.1.1.35)\n",
      "2 Two-component system sensor histidine kinase/response regulator hybrid\n",
      "2 Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)\n",
      "2 Trk potassium uptake system protein TrkH\n",
      "2 Transposase\n",
      "2 Transmembrane component YkoC of energizing module of thiamin-regulated ECF transporter for HydroxyMethylPyrimidine\n",
      "2 Translation initiation factor SUI1-related protein\n",
      "2 Translation initiation factor 3\n",
      "2 Transglycosylase-associated protein\n",
      "2 Transcriptional regulator, PadR family\n",
      "2 Transcriptional regulator, AcrR family\n",
      "2 Transcriptional regulator\n",
      "2 Transcription termination protein NusB\n",
      "2 Transcription regulator [contains diacylglycerol kinase catalytic domain]\n",
      "2 Transcription accessory protein (S1 RNA-binding domain)\n",
      "2 Thiol peroxidase, Bcp-type (EC 1.11.1.15)\n",
      "2 Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)\n",
      "2 Superoxide dismutase [Mn/Fe] (EC 1.15.1.1)\n",
      "2 Sulfur carrier protein ThiS adenylyltransferase (EC 2.7.7.73)\n",
      "2 Sugar-phosphate nucleotidyl transferase\n",
      "2 Sortase A, LPXTG specific\n",
      "2 Similar to conjugative transposon protein TraB\n",
      "2 Signal recognition particle receptor FtsY\n",
      "2 Serine/threonine kinase\n",
      "2 Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog\n",
      "2 SanA protein\n",
      "2 SSU ribosomal protein S20p\n",
      "2 SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent\n",
      "2 SAM-dependent methyltransferase\n",
      "2 Rubredoxin\n",
      "2 Ribosomal protein S12p Asp88 (E. coli) methylthiotransferase (EC 2.8.4.4)\n",
      "2 Ribose-phosphate pyrophosphokinase (EC 2.7.6.1)\n",
      "2 Ribose-5-phosphate isomerase B (EC 5.3.1.6)\n",
      "2 Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor\n",
      "2 Recombination protein RecR\n",
      "2 Pyridoxal phosphate-containing protein YggS\n",
      "2 Putative pre-16S rRNA nuclease YqgF\n",
      "2 Putative phage-encoded peptidoglycan binding protein\n",
      "2 Putative membrane-associated metal-dependent hydrolase\n",
      "2 Putative dipeptidase\n",
      "2 Putative calcineurin superfamily phosphohydrolase\n",
      "2 Putative alpha-1,6-mannanase\n",
      "2 Putative DNA-binding protein\n",
      "2 Protein BF2544, predicted transporter component\n",
      "2 Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)\n",
      "2 Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup\n",
      "2 Polysaccharide deacetylase-like protein\n",
      "2 Polyribonucleotide nucleotidyltransferase (EC 2.7.7.8)\n",
      "2 Phosphoribosylanthranilate isomerase (EC 5.3.1.24)\n",
      "2 Phosphonopyruvate decarboxylase (EC 4.1.1.82)\n",
      "2 Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)\n",
      "2 Peroxide stress regulator; Ferric uptake regulation protein; Fe2+/Zn2+ uptake regulation proteins\n",
      "2 Pentapeptide repeat family protein\n",
      "2 Outer membrane assembly lipoprotein YfiO\n",
      "2 OmpA\n",
      "2 Oligopeptide transporter, OPT family\n",
      "2 Na+-driven multidrug efflux pump\n",
      "2 Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)\n",
      "2 Murein endolytic transglycosylase MltG\n",
      "2 Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps\n",
      "2 Methylmalonyl-CoA epimerase (EC 5.1.99.1)\n",
      "2 Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)\n",
      "2 LysM-repeat proteins and domains\n",
      "2 Long-chain-fatty-acid--CoA ligase (EC 6.2.1.3)\n",
      "2 Lipopolysaccharide export system permease protein LptG\n",
      "2 LemA protein\n",
      "2 LSU ribosomal protein L34p\n",
      "2 LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent\n",
      "2 LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent\n",
      "2 L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)\n",
      "2 Isoleucyl-tRNA synthetase (EC 6.1.1.5)\n",
      "2 Integrase protein\n",
      "2 Inner membrane protein YihY, formerly thought to be RNase BN\n",
      "2 Imidazole glycerol phosphate synthase amidotransferase subunit HisH\n",
      "2 Hemin receptor\n",
      "2 Hcp transcriptional regulator HcpR (Crp/Fnr family)\n",
      "2 Glyoxalase family protein\n",
      "2 Glycosyltransferase involved in cell wall biogenesis\n",
      "2 Glycosylase BF2339\n",
      "2 Glycosyl transferase, group 2 family\n",
      "2 Formate-dependent phosphoribosylglycinamide formyltransferase\n",
      "2 Fluoride ion transporter CrcB\n",
      "2 FIG138056: a glutathione-dependent thiol reductase\n",
      "2 FIG00936810: hypothetical protein\n",
      "2 FIG00491264: hypothetical protein\n",
      "2 FIG00414141: hypothetical protein\n",
      "2 FIG00003370: Multicopper polyphenol oxidase\n",
      "2 FHA domain protein\n",
      "2 Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)\n",
      "2 Electron transport complex protein RnfB\n",
      "2 Electron transport complex protein RnfA\n",
      "2 Efflux transport system, outer membrane factor (OMF) lipoprotein\n",
      "2 DnaK suppressor protein, putative\n",
      "2 Dethiobiotin synthase BioD (EC 6.3.3.3)\n",
      "2 DNA/RNA helicases, SNF2 family\n",
      "2 DNA-binding protein HU-beta\n",
      "2 DNA base-flipping protein\n",
      "2 D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)\n",
      "2 Cytochrome d ubiquinol oxidase subunit II (EC 1.10.3.-)\n",
      "2 Cytidylyltransferase domain / Phosphoenolpyruvate phosphomutase (EC 5.4.2.9)\n",
      "2 Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)\n",
      "2 Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain\n",
      "2 Cobalt-precorrin-7 (C5)-methyltransferase (EC 2.1.1.289) / Cobalt-precorrin-6B C15-methyltransferase [decarboxylating] (EC 2.1.1.196)\n",
      "2 Cobalt-precorrin-2 C(20)-methyltransferase (EC 2.1.1.151)\n",
      "2 Clostripain-related protein\n",
      "2 Chromate transport protein\n",
      "2 Chaperone protein HtpG\n",
      "2 Cation efflux system protein\n",
      "2 Carboxyl-terminal protease\n",
      "2 Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase\n",
      "2 Beta-lactamase class C-like and penicillin binding proteins (PBPs) superfamily\n",
      "2 Aspartate carbamoyltransferase regulatory chain (PyrI)\n",
      "2 Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)\n",
      "2 Arylsulfatase (EC 3.1.6.1)\n",
      "2 Aminobenzoyl-glutamate transport protein\n",
      "2 Amino acid-binding ACT\n",
      "2 Alpha-ribazole-5'-phosphate phosphatase (EC 3.1.3.73)\n",
      "2 Alkaline phosphatase (EC 3.1.3.1)\n",
      "2 Agmatine deiminase (EC 3.5.3.12)\n",
      "2 Aerotolerance protein BatA\n",
      "2 Adenylylsulfate kinase (EC 2.7.1.25)\n",
      "2 Adenosylcobinamide-phosphate synthase (EC 6.3.1.10)\n",
      "2 Adenosine deaminase (EC 3.5.4.4)\n",
      "2 ATPase RavA\n",
      "2 ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis\n",
      "2 ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)\n",
      "2 ATP-dependent DNA helicase\n",
      "2 ATP synthase beta chain (EC 3.6.3.14)\n",
      "2 ATP synthase alpha chain (EC 3.6.3.14)\n",
      "2 ATP synthase F0 sector subunit a (EC 3.6.3.14)\n",
      "2 AMP nucleosidase (EC 3.2.2.4)\n",
      "2 6-phosphogluconolactonase (EC 3.1.1.31), eukaryotic type\n",
      "2 6-phosphogluconate dehydrogenase, decarboxylating (EC 1.1.1.44)\n",
      "2 5-amino-6-(5-phosphoribosylamino)uracil reductase\n",
      "2 5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\n",
      "2 4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase\n",
      "2 4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)\n",
      "2 3-deoxy-manno-octulosonate cytidylyltransferase (EC 2.7.7.38)\n",
      "2 2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)\n",
      "16 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "15 Transcriptional regulator, AraC family\n",
      "14 Two-component system sensor histidine kinase\n",
      "13 Thioredoxin\n",
      "12 putative membrane protein\n",
      "12 Putative anti-sigma factor\n",
      "11 Acidobacterial duplicated orphan permease (function unknown)\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for tp_type in gained_genes_counts_by_tp_type:\n",
    "    print(tp_type)\n",
    "    strs = []\n",
    "    for gene_name in gained_genes_counts_by_tp_type[tp_type]:\n",
    "        count = gained_genes_counts_by_tp_type[tp_type][gene_name]\n",
    "        if count > 1:\n",
    "            strs.append(\"%i %s\" % (count, gene_name))\n",
    "    for string in sorted(strs, reverse=True):\n",
    "        print(string)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AA': defaultdict(int, {}),\n",
       " 'II-nonpremie': defaultdict(int,\n",
       "             {'1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)': 0.14,\n",
       "              \"16S rRNA (cytidine(1402)-2'-O)-methyltransferase (EC 2.1.1.198)\": 0.21999999999999997,\n",
       "              '2-amino-3-ketobutyrate coenzyme A ligase (EC 2.3.1.29)': 0.27999999999999997,\n",
       "              '2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)': 0.41999999999999993,\n",
       "              '2-methylfumaryl-CoA hydratase (EC 4.2.1.148)': 0.25999999999999995,\n",
       "              '2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)': 0.3,\n",
       "              '2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, gamma subunit (EC 1.2.7.-)': 0.27999999999999997,\n",
       "              '3-deoxy-manno-octulosonate cytidylyltransferase (EC 2.7.7.38)': 0.23999999999999996,\n",
       "              '3-isopropylmalate dehydratase small subunit (EC 4.2.1.33)': 0.34,\n",
       "              '3-oxo-5-alpha-steroid 4-dehydrogenase': 0.3,\n",
       "              '3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)': 0.45999999999999996,\n",
       "              '4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)': 0.32,\n",
       "              '4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)': 0.4799999999999999,\n",
       "              '4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)': 0.36000000000000004,\n",
       "              '4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase': 0.18,\n",
       "              \"5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\": 0.3,\n",
       "              '5-amino-6-(5-phosphoribosylamino)uracil reductase': 0.18,\n",
       "              '6-phosphogluconate dehydrogenase, decarboxylating (EC 1.1.1.44)': 0.38,\n",
       "              '6-phosphogluconolactonase (EC 3.1.1.31), eukaryotic type': 0.27999999999999997,\n",
       "              'AAA+ superfamily protein': 0.14,\n",
       "              'ABC-type antimicrobial peptide transport system, ATPase component': 0.82,\n",
       "              'ADP-ribose pyrophosphatase of COG1058 family (EC 3.6.1.13) / Nicotinamide-nucleotide amidase (EC 3.5.1.42)': 0.21999999999999997,\n",
       "              'AMP nucleosidase (EC 3.2.2.4)': 0.18,\n",
       "              'ATP phosphoribosyltransferase (EC 2.4.2.17) => HisGl': 0.16,\n",
       "              'ATP synthase F0 sector subunit a (EC 3.6.3.14)': 0.3,\n",
       "              'ATP synthase F0 sector subunit b (EC 3.6.3.14)': 0.23999999999999996,\n",
       "              'ATP synthase F0 sector subunit c (EC 3.6.3.14)': 0.45999999999999996,\n",
       "              'ATP synthase alpha chain (EC 3.6.3.14)': 0.18,\n",
       "              'ATP synthase beta chain (EC 3.6.3.14)': 0.19999999999999998,\n",
       "              'ATP synthase epsilon chain (EC 3.6.3.14)': 0.4999999999999999,\n",
       "              'ATP synthase gamma chain (EC 3.6.3.14)': 0.54,\n",
       "              'ATP synthase protein I': 0.62,\n",
       "              'ATP-dependent DNA helicase': 0.16,\n",
       "              'ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)': 0.21999999999999997,\n",
       "              'ATP-dependent RNA helicase': 0.6000000000000001,\n",
       "              'ATP-dependent RNA helicase BT1154': 0.23999999999999996,\n",
       "              'ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)': 0.19999999999999998,\n",
       "              'ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis': 0.38000000000000006,\n",
       "              'ATPase RavA': 0.33999999999999997,\n",
       "              'Acetyl esterase': 0.25999999999999995,\n",
       "              'Acetyltransferase': 0.16,\n",
       "              'Acetyltransferase, GNAT family': 0.25999999999999995,\n",
       "              'Acidobacterial duplicated orphan permease (function unknown)': 3.0399999999999996,\n",
       "              'Acyltransferase': 0.25999999999999995,\n",
       "              'Adenosine deaminase (EC 3.5.4.4)': 0.5599999999999999,\n",
       "              'Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)': 0.58,\n",
       "              'Adenosylcobinamide-phosphate synthase (EC 6.3.1.10)': 0.16,\n",
       "              'Adenylylsulfate kinase (EC 2.7.1.25)': 0.66,\n",
       "              'Aerotolerance protein BatA': 0.41999999999999993,\n",
       "              'Aerotolerance protein BatB': 0.43999999999999995,\n",
       "              'Aerotolerance protein BatC': 0.41999999999999993,\n",
       "              'Aerotolerance protein BatD': 0.38,\n",
       "              'Aerotolerance protein BatE': 0.68,\n",
       "              'Agmatine deiminase (EC 3.5.3.12)': 0.27999999999999997,\n",
       "              'Alkaline phosphatase (EC 3.1.3.1)': 0.34,\n",
       "              'Alpha-1,2-fucosyltransferase': 0.36000000000000004,\n",
       "              'Alpha-1,4-N-acetylgalactosamine transferase PglH (EC 2.4.1.-)': 0.25999999999999995,\n",
       "              'Alpha-N-acetylglucosaminidase (EC 3.2.1.50)': 0.18,\n",
       "              \"Alpha-ribazole-5'-phosphate phosphatase (EC 3.1.3.73)\": 0.16,\n",
       "              'Amino acid-binding ACT': 0.58,\n",
       "              'Aminobenzoyl-glutamate transport protein': 0.39999999999999997,\n",
       "              'Aminodeoxychorismate lyase (EC 4.1.3.38)': 0.42,\n",
       "              'Aminotransferase, DegT/DnrJ/EryC1/StrS family': 0.18,\n",
       "              'AmpG protein': 0.27999999999999997,\n",
       "              'Arabinan endo-1,5-alpha-L-arabinosidase A': 0.25999999999999995,\n",
       "              'Arylsulfatase (EC 3.1.6.1)': 0.34,\n",
       "              'Arylsulfatase regulator': 0.1,\n",
       "              'Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)': 0.23999999999999996,\n",
       "              'Aspartate carbamoyltransferase (EC 2.1.3.2)': 0.36000000000000004,\n",
       "              'Aspartate carbamoyltransferase regulatory chain (PyrI)': 0.3,\n",
       "              'Bacterial non-heme ferritin (EC 1.16.3.2)': 0.18,\n",
       "              'Bacterial ribosome SSU maturation protein RimP': 0.23999999999999996,\n",
       "              'Beta-1,3-glucosyltransferase': 0.32,\n",
       "              'Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase': 0.21999999999999997,\n",
       "              'CAAX amino terminal protease family protein': 0.76,\n",
       "              'CRISPR-associated RecB family exonuclease Cas4': 0.21999999999999997,\n",
       "              'CRISPR-associated helicase Cas3': 0.34,\n",
       "              'CRISPR-associated negative autoregulator Cas7/Cst2': 0.3,\n",
       "              'CRISPR-associated protein Cas1': 0.21999999999999997,\n",
       "              'CRISPR-associated protein, TM1814 family': 0.38000000000000006,\n",
       "              'Capsular polysaccharide biosynthesis protein capD': 0.18,\n",
       "              'Capsular polysaccharide transcription antitermination protein UpeY': 0.18,\n",
       "              'Capsular polysaccharide transcription antitermination protein, UpxY family': 1.2200000000000002,\n",
       "              'Carbon starvation protein A': 0.21999999999999997,\n",
       "              'Carboxy-terminal processing protease': 0.23999999999999996,\n",
       "              'Carboxyl-terminal protease': 0.18,\n",
       "              'Cation efflux system protein': 0.27999999999999997,\n",
       "              'Cell division ZapA family protein': 0.21999999999999997,\n",
       "              'Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage': 0.4999999999999999,\n",
       "              'Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family': 4.14,\n",
       "              'Cell surface protein': 0.34,\n",
       "              'Chaperone protein DnaJ': 0.46000000000000013,\n",
       "              'Chaperone protein HtpG': 0.52,\n",
       "              'Chromate transport protein': 0.23999999999999996,\n",
       "              'Clostripain-related protein': 0.14,\n",
       "              'Co-activator of prophage gene expression IbrB': 0.56,\n",
       "              'Cobalamin synthase (EC 2.7.8.26)': 0.27999999999999997,\n",
       "              'Cobalt-precorrin-2 C(20)-methyltransferase (EC 2.1.1.151)': 0.18,\n",
       "              'Cobalt-precorrin-7 (C5)-methyltransferase (EC 2.1.1.289) / Cobalt-precorrin-6B C15-methyltransferase [decarboxylating] (EC 2.1.1.196)': 0.21999999999999997,\n",
       "              'Cobyric acid synthase (EC 6.3.5.10)': 0.32,\n",
       "              'Conjugative transposon protein TraM': 0.27999999999999997,\n",
       "              'Conserved hypothetical protein 22': 0.41999999999999993,\n",
       "              'Conserved protein YuxK': 0.62,\n",
       "              'Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain': 0.4400000000000001,\n",
       "              'Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)': 0.21999999999999997,\n",
       "              'Cys-tRNA(Pro) deacylase YbaK': 0.39999999999999997,\n",
       "              'Cystathionine beta-lyase (EC 4.4.1.8)': 0.27999999999999997,\n",
       "              'Cytidylyltransferase domain / Phosphoenolpyruvate phosphomutase (EC 5.4.2.9)': 0.25999999999999995,\n",
       "              'Cytochrome d ubiquinol oxidase subunit II (EC 1.10.3.-)': 0.21999999999999997,\n",
       "              'CzcABC family efflux RND transporter, outer membrane protein': 0.5599999999999999,\n",
       "              'D-alanine--poly(phosphoribitol) ligase ACP subunit (EC 6.1.1.13)': 0.49999999999999994,\n",
       "              'D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)': 0.3,\n",
       "              'DNA base-flipping protein': 0.25999999999999995,\n",
       "              'DNA ligase (NAD(+)) (EC 6.5.1.2)': 0.56,\n",
       "              'DNA polymerase III delta subunit (EC 2.7.7.7)': 0.5399999999999999,\n",
       "              'DNA protection during starvation protein': 0.25999999999999995,\n",
       "              'DNA-binding protein': 0.21999999999999997,\n",
       "              'DNA-binding protein HU-beta': 0.5,\n",
       "              'DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)': 0.4799999999999999,\n",
       "              'DNA/RNA helicases, SNF2 family': 0.27999999999999997,\n",
       "              'Death on curing protein, Doc toxin': 0.19999999999999998,\n",
       "              'Dephospho-CoA kinase (EC 2.7.1.24)': 0.32,\n",
       "              'Dethiobiotin synthase BioD (EC 6.3.3.3)': 0.27999999999999997,\n",
       "              'Diacylglycerol kinase (EC 2.7.1.107)': 0.66,\n",
       "              'Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)': 0.41999999999999993,\n",
       "              'Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)': 0.44,\n",
       "              'Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)': 0.4999999999999999,\n",
       "              'DnaD domain protein': 0.25999999999999995,\n",
       "              'DnaK suppressor protein, putative': 0.18,\n",
       "              'DoxX family protein': 0.44,\n",
       "              'Efflux ABC transporter, permease protein': 0.54,\n",
       "              'Efflux transport system, outer membrane factor (OMF) lipoprotein': 0.1,\n",
       "              'Efflux transport system, outer membrane factor (OMF) lipoprotein BT_0954': 0.38000000000000006,\n",
       "              'Electron transport complex protein RnfA': 0.23999999999999996,\n",
       "              'Electron transport complex protein RnfB': 0.32,\n",
       "              'Electron transport complex protein RnfG': 0.19999999999999998,\n",
       "              'Endo-1,4-beta-xylanase D': 0.16,\n",
       "              'Endonuclease': 0.56,\n",
       "              'Energy-dependent translational throttle protein EttA': 0.23999999999999996,\n",
       "              'Error-prone, lesion bypass DNA polymerase V (UmuC)': 0.36000000000000004,\n",
       "              'Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)': 0.25999999999999995,\n",
       "              'FAD:protein FMN transferase (EC 2.7.1.180)': 0.23999999999999996,\n",
       "              'FHA domain protein': 0.5199999999999999,\n",
       "              'FIG00003370: Multicopper polyphenol oxidase': 0.16,\n",
       "              'FIG000605: protein co-occurring with transport systems (COG1739)': 0.5,\n",
       "              'FIG00414141: hypothetical protein': 0.25999999999999995,\n",
       "              'FIG00490011: hypothetical protein': 0.4999999999999999,\n",
       "              'FIG00491264: hypothetical protein': 0.32,\n",
       "              'FIG00936810: hypothetical protein': 0.39999999999999997,\n",
       "              'FIG01416087: hypothetical protein': 0.16,\n",
       "              'FIG015287: Zinc protease': 0.16,\n",
       "              'FIG016027: protein of unknown function YeaO': 0.58,\n",
       "              'FIG137478: Hypothetical protein': 0.56,\n",
       "              'FIG138056: a glutathione-dependent thiol reductase': 0.4200000000000001,\n",
       "              'FKBP-type peptidyl-prolyl cis-trans isomerase FkpA precursor (EC 5.2.1.8)': 0.4000000000000001,\n",
       "              'Ferredoxin': 0.23999999999999996,\n",
       "              'Ferredoxin domain containing protein': 0.32,\n",
       "              'Flavodoxin': 0.25999999999999995,\n",
       "              'Flavoredoxin': 0.41999999999999993,\n",
       "              'Fluoride ion transporter CrcB': 0.25999999999999995,\n",
       "              'Formate--tetrahydrofolate ligase (EC 6.3.4.3)': 0.25999999999999995,\n",
       "              'Formate-dependent phosphoribosylglycinamide formyltransferase': 0.34,\n",
       "              'Fosmidomycin resistance protein': 0.5800000000000001,\n",
       "              'Free methionine-(R)-sulfoxide reductase, contains GAF domain': 0.68,\n",
       "              'Fructose-1,6-bisphosphatase, Bacillus type (EC 3.1.3.11)': 0.21999999999999997,\n",
       "              'GTP cyclohydrolase 1 type 2 homolog YbgI @ Bsu YqfO NIF3/CutA domain': 0.43999999999999995,\n",
       "              'GTP-binding protein EngB': 0.18,\n",
       "              'GldB': 0.3,\n",
       "              'Gll2767 protein': 0.32,\n",
       "              'Glucosamine-1-phosphate guanylyltransferase': 0.23999999999999996,\n",
       "              'Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)': 0.34,\n",
       "              'Glycosyl transferase, family 2': 0.34,\n",
       "              'Glycosyl transferase, group 1': 0.32,\n",
       "              'Glycosyl transferase, group 1 family protein': 0.32,\n",
       "              'Glycosyl transferase, group 2 family': 0.27999999999999997,\n",
       "              'Glycosylase BF2339': 0.41999999999999993,\n",
       "              'Glycosyltransferase': 0.8999999999999999,\n",
       "              'Glycosyltransferase involved in cell wall biogenesis': 0.62,\n",
       "              'Glyoxalase family protein': 0.27999999999999997,\n",
       "              'Guanine deaminase (EC 3.5.4.3)': 0.32,\n",
       "              'Hcp transcriptional regulator HcpR (Crp/Fnr family)': 0.19999999999999998,\n",
       "              'Heat shock protein GrpE': 0.19999999999999998,\n",
       "              'Heat shock protein HtpX': 0.38000000000000006,\n",
       "              'Helicase': 0.36000000000000004,\n",
       "              'Helicase, C-terminal:Type III restriction enzyme, res subunit:DEAD/DEAH box helicase, N-terminal': 0.6599999999999999,\n",
       "              'Hemin receptor': 0.36000000000000004,\n",
       "              'Huntingtin interacting protein E-like protein': 0.32,\n",
       "              'Hydroxylamine reductase (EC 1.7.99.1)': 0.3,\n",
       "              'Hydroxymethylpyrimidine phosphate kinase ThiD (EC 2.7.4.7)': 0.19999999999999998,\n",
       "              'Hypothetical Nudix-like regulator': 0.39999999999999997,\n",
       "              'ISNCY family transposase': 1.2000000000000002,\n",
       "              'Imidazole glycerol phosphate synthase amidotransferase subunit HisH': 0.5199999999999999,\n",
       "              'Imidazole glycerol phosphate synthase cyclase subunit': 0.21999999999999997,\n",
       "              'Immunoreactive 53 kDa antigen PG123': 0.45999999999999996,\n",
       "              'Indole-3-glycerol phosphate synthase (EC 4.1.1.48)': 0.39999999999999997,\n",
       "              'Inner membrane protein RclC associated with response to reactive chlorine species': 0.36000000000000004,\n",
       "              'Inner membrane protein YhiM': 0.5399999999999999,\n",
       "              'Inner membrane protein YihY, formerly thought to be RNase BN': 0.21999999999999997,\n",
       "              'Inorganic pyrophospatase PpaX': 0.23999999999999996,\n",
       "              'Integral membrane protein': 0.6,\n",
       "              'Integrase': 1.66,\n",
       "              'Integrase IntN1': 0.19999999999999998,\n",
       "              'Integrase protein': 0.18,\n",
       "              'Integration host factor alpha subunit': 0.54,\n",
       "              'Integration host factor beta subunit': 0.25999999999999995,\n",
       "              'Intracellular protease': 0.21999999999999997,\n",
       "              'Iron-sulfur flavoprotein': 0.3,\n",
       "              'Isoleucyl-tRNA synthetase (EC 6.1.1.5)': 0.21999999999999997,\n",
       "              'Keratan-sulfate endo-1,4-beta-galactosidase (EC 3.2.1.103)': 0.3,\n",
       "              'L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)': 0.14,\n",
       "              'L-fucose kinase (EC 2.7.1.52)': 0.23999999999999996,\n",
       "              'L-fucose mutarotase (EC 5.1.3.29)': 0.27999999999999997,\n",
       "              'LSU ribosomal protein L24p (L26e)': 0.27999999999999997,\n",
       "              'LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent': 0.27999999999999997,\n",
       "              'LSU ribosomal protein L29p (L35e)': 0.12000000000000001,\n",
       "              'LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent': 0.25999999999999995,\n",
       "              'LSU ribosomal protein L34p': 0.12000000000000001,\n",
       "              'Lactaldehyde reductase (EC 1.1.1.77)': 0.4000000000000001,\n",
       "              'Large-conductance mechanosensitive channel': 0.7000000000000001,\n",
       "              'LemA protein': 0.38000000000000006,\n",
       "              'Lipid-A-disaccharide synthase (EC 2.4.1.182)': 0.19999999999999998,\n",
       "              'Lipopolysaccharide export system permease protein LptG': 0.45999999999999996,\n",
       "              'Lipoprotein signal peptidase (EC 3.4.23.36)': 0.4799999999999999,\n",
       "              'Long tail fiber protein p37': 0.18,\n",
       "              'Long-chain-fatty-acid--CoA ligase (EC 6.2.1.3)': 0.36,\n",
       "              'LysM-repeat proteins and domains': 0.27999999999999997,\n",
       "              'Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)': 0.27999999999999997,\n",
       "              'MBL-fold metallo-hydrolase superfamily': 0.49999999999999994,\n",
       "              'Malonyl CoA-acyl carrier protein transacylase (EC 2.3.1.39)': 0.39999999999999997,\n",
       "              'Mannan endo-1,4-beta-mannosidase': 0.18,\n",
       "              'Mannose-1-phosphate guanylyltransferase (EC 2.7.7.13)': 0.46000000000000013,\n",
       "              'Membrane protein': 0.66,\n",
       "              'Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids': 0.19999999999999998,\n",
       "              'Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly': 0.25999999999999995,\n",
       "              'Methylmalonyl-CoA epimerase (EC 5.1.99.1)': 0.32,\n",
       "              'Monofunctional biosynthetic peptidoglycan transglycosylase': 0.8800000000000001,\n",
       "              'Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps': 0.23999999999999996,\n",
       "              'Multidrug resistance protein': 0.4799999999999999,\n",
       "              'Murein endolytic transglycosylase MltG': 0.6,\n",
       "              'MutT/Nudix family protein': 0.27999999999999997,\n",
       "              'N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)': 0.5,\n",
       "              'N-acetylgalactosamine-6-sulfatase': 0.27999999999999997,\n",
       "              'N-acetylglucosaminyltransferase (EC 2.4.1.-)': 0.43999999999999995,\n",
       "              'N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)': 0.72,\n",
       "              'N-acetylneuraminate lyase (EC 4.1.3.3)': 0.16,\n",
       "              'N-acetylneuraminate synthase (EC 2.5.1.56)': 0.4200000000000001,\n",
       "              'N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)': 0.7799999999999999,\n",
       "              'N-carbamoylputrescine amidase (EC 3.5.1.53)': 0.18,\n",
       "              'N-carbamoylputrescine amidase (EC 3.5.1.53) @ Aliphatic amidase AmiE (EC 3.5.1.4)': 0.18,\n",
       "              'Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)': 0.25999999999999995,\n",
       "              'Na+-driven multidrug efflux pump': 0.21999999999999997,\n",
       "              'Na+/H+-dicarboxylate symporter': 0.56,\n",
       "              'Nitrogen regulation protein NtrY (EC 2.7.3.-)': 0.25999999999999995,\n",
       "              'Nitroreductase family protein': 0.5599999999999999,\n",
       "              'O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)': 0.6,\n",
       "              'Oligopeptide transporter, OPT family': 0.27999999999999997,\n",
       "              'OmpA': 0.25999999999999995,\n",
       "              'OmpA family outer membrane protein': 0.68,\n",
       "              'Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family': 2.3200000000000003,\n",
       "              'Outer membrane assembly lipoprotein YfiO': 0.5800000000000002,\n",
       "              'Outer membrane protein romA': 0.45999999999999996,\n",
       "              'Outer membrane receptor for Fe3+-dicitrate': 0.23999999999999996,\n",
       "              'Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)': 0.21999999999999997,\n",
       "              'Oxidoreductase': 0.5800000000000001,\n",
       "              'Oxidoreductase, aldo/keto reductase family': 0.8200000000000001,\n",
       "              'PAP2 superfamily protein': 0.52,\n",
       "              'Pantoate--beta-alanine ligase (EC 6.3.2.1)': 0.19999999999999998,\n",
       "              'Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)': 0.45999999999999996,\n",
       "              'Pentapeptide repeat family protein': 0.32,\n",
       "              'Peptidase, S41 family': 0.25999999999999995,\n",
       "              'Peptide deformylase (EC 3.5.1.88)': 0.34,\n",
       "              'Permease of the drug/metabolite transporter (DMT) superfamily': 0.96,\n",
       "              'Peroxide stress regulator; Ferric uptake regulation protein; Fe2+/Zn2+ uptake regulation proteins': 0.34,\n",
       "              'Phage integrase': 0.58,\n",
       "              'Phage protein': 0.27999999999999997,\n",
       "              'Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)': 0.16,\n",
       "              'Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO': 0.7200000000000001,\n",
       "              'Phosphonopyruvate decarboxylase (EC 4.1.1.82)': 0.19999999999999998,\n",
       "              'Phosphoribosylanthranilate isomerase (EC 5.3.1.24)': 0.18,\n",
       "              'Phosphoribosylformimino-5-aminoimidazole carboxamide ribotide isomerase (EC 5.3.1.16)': 0.6,\n",
       "              'Pirin': 0.18,\n",
       "              'Polysaccharide deacetylase': 0.23999999999999996,\n",
       "              'Polysaccharide deacetylase-like protein': 0.3,\n",
       "              'Possible exported heavy-metal binding protein': 0.6200000000000001,\n",
       "              'Possible glyoxylase family protein (Lactoylglutathione lyase) (EC 4.4.1.5)': 0.43999999999999995,\n",
       "              'Possible restriction /modification enzyme': 0.23999999999999996,\n",
       "              'Predicted cobalt transporter in Bacteroides_Porphyromonas': 0.16,\n",
       "              'Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup': 0.38000000000000006,\n",
       "              'Predicted transcriptional regulator': 0.21999999999999997,\n",
       "              'Probable integral membrane protein Cj1165c': 0.62,\n",
       "              'Probable integral membrane protein Cj1166c': 0.23999999999999996,\n",
       "              'Probable iron export ATP-binding protein FetA': 0.5399999999999999,\n",
       "              'Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)': 0.25999999999999995,\n",
       "              'Prolyl oligopeptidase family protein': 0.34,\n",
       "              'Protein BF2544, predicted transporter component': 0.56,\n",
       "              'Protein DUF2149, predicted transporter component': 0.18,\n",
       "              'Protein of unknown function UPF0060': 0.21999999999999997,\n",
       "              'Protein translocase membrane subunit SecG': 0.34,\n",
       "              'Protein translocase subunit YajC': 0.23999999999999996,\n",
       "              'Putative ATP-binding protein': 0.4200000000000001,\n",
       "              'Putative DNA-binding protein': 0.38,\n",
       "              'Putative Na+/H+ exchange protein': 0.21999999999999997,\n",
       "              'Putative alpha-1,6-mannanase': 0.25999999999999995,\n",
       "              'Putative aminotransferase BF2110': 0.32,\n",
       "              'Putative anti-sigma factor': 2.6999999999999997,\n",
       "              'Putative calcineurin superfamily phosphohydrolase': 0.34,\n",
       "              'Putative dipeptidase': 0.16,\n",
       "              'Putative glycosyltransferase': 0.9399999999999997,\n",
       "              'Putative helicase': 1.14,\n",
       "              'Putative lipopolysaccharide biosynthesis protein': 0.27999999999999997,\n",
       "              'Putative manganese efflux pump MntP': 0.16,\n",
       "              'Putative membrane-associated metal-dependent hydrolase': 0.19999999999999998,\n",
       "              'Putative molybdenum transport ATP-binding protein modF': 0.3,\n",
       "              'Putative outer membrane protein, probably involved in nutrient binding': 0.5800000000000001,\n",
       "              'Putative periplasmic protein': 0.39999999999999997,\n",
       "              'Putative permease': 0.36,\n",
       "              'Putative phage-encoded peptidoglycan binding protein': 0.25999999999999995,\n",
       "              'Putative phosphatidylinositol-4-phosphate 5-kinase': 0.19999999999999998,\n",
       "              'Putative pre-16S rRNA nuclease YqgF': 0.39999999999999997,\n",
       "              'Putative preQ0 transporter YhhQ': 0.36000000000000004,\n",
       "              'Putative secreted polysaccharide polymerase': 0.23999999999999996,\n",
       "              'Putative sodium-dependent transporter': 0.19999999999999998,\n",
       "              'Putative thioredoxin': 0.45999999999999996,\n",
       "              'Pyridoxal phosphate-containing protein YggS': 0.34,\n",
       "              'Pyruvate formate-lyase activating enzyme (EC 1.97.1.4)': 0.3,\n",
       "              'Pyruvate:Oxaloacetate transcarboxylase domain protein / Biotin carboxyl carrier protein': 0.21999999999999997,\n",
       "              'Quinolinate synthetase (EC 2.5.1.72)': 0.21999999999999997,\n",
       "              'RNA methyltransferase, TrmH family': 0.36,\n",
       "              'RNA polymerase ECF-type sigma factor': 9.079999999999998,\n",
       "              'RNA-binding protein': 1.6400000000000001,\n",
       "              'RND efflux system, inner membrane transporter': 0.3,\n",
       "              'ROK family sugar kinase or transcriptional regulator': 0.68,\n",
       "              'Recombination protein RecR': 0.6,\n",
       "              'Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor': 0.25999999999999995,\n",
       "              'Related to F420H2-dehydrogenase, beta subunit': 0.19999999999999998,\n",
       "              'Ribonuclease G': 0.21999999999999997,\n",
       "              'Ribonucleotide reductase of class Ia (aerobic), beta subunit (EC 1.17.4.1)': 0.23999999999999996,\n",
       "              'Ribose-5-phosphate isomerase B (EC 5.3.1.6)': 0.32,\n",
       "              'Ribose-phosphate pyrophosphokinase (EC 2.7.6.1)': 0.34,\n",
       "              'Ribosomal protein S12p Asp88 (E. coli) methylthiotransferase (EC 2.8.4.4)': 0.46,\n",
       "              'Ribosomal silencing factor RsfA': 0.62,\n",
       "              'Ribosomal-protein-S5p-alanine acetyltransferase (EC 2.3.1.128)': 0.23999999999999996,\n",
       "              'Ribosome hibernation promoting factor Hpf': 0.21999999999999997,\n",
       "              'Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)': 0.45999999999999996,\n",
       "              'Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria': 0.45999999999999996,\n",
       "              'RidA/YER057c/UK114 superfamily protein': 0.48,\n",
       "              'Rod shape-determining protein MreD': 0.34,\n",
       "              'Rrf2-linked NADH-flavin reductase': 0.45999999999999996,\n",
       "              'Rubredoxin': 0.27999999999999997,\n",
       "              'SAM-dependent methyltransferase': 0.25999999999999995,\n",
       "              'SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent': 0.16,\n",
       "              'SSU ribosomal protein S20p': 0.22,\n",
       "              'SSU ribosomal protein S6p': 0.21999999999999997,\n",
       "              'SanA protein': 0.43999999999999995,\n",
       "              'SclB protein': 0.94,\n",
       "              'Ser-tRNA(Ala) deacylase @ Gly-tRNA(Ala) deacylase': 0.18,\n",
       "              'Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog': 0.42,\n",
       "              'Serine hydroxymethyltransferase (EC 2.1.2.1)': 0.18,\n",
       "              'Serine palmitoyltransferase (EC 2.3.1.50)': 0.34,\n",
       "              'Serine protease': 0.54,\n",
       "              'Serine/threonine kinase': 0.14,\n",
       "              'Sigma factor RpoE regulatory protein RseC': 0.4999999999999999,\n",
       "              'Signal recognition particle receptor FtsY': 0.21999999999999997,\n",
       "              'Similar to conjugative transposon protein TraB': 0.4200000000000001,\n",
       "              'Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)': 0.5800000000000001,\n",
       "              'Slr1117 protein': 0.49999999999999994,\n",
       "              'Small heat shock protein': 0.68,\n",
       "              'Sodium/proline symporter': 0.27999999999999997,\n",
       "              'Stress responsive alpha-beta barrel domain protein Dabb': 0.6,\n",
       "              'Succinyl-CoA ligase [ADP-forming] alpha chain (EC 6.2.1.5)': 0.23999999999999996,\n",
       "              'Succinyl-CoA ligase [ADP-forming] beta chain (EC 6.2.1.5)': 0.23999999999999996,\n",
       "              'Sugar-phosphate nucleotidyl transferase': 0.18,\n",
       "              'Sulfur carrier protein ThiS': 0.21999999999999997,\n",
       "              'Sulfur carrier protein ThiS adenylyltransferase (EC 2.7.7.73)': 0.23999999999999996,\n",
       "              'Superfamily I DNA and RNA helicases and helicase subunits': 0.3,\n",
       "              'Superfamily I DNA/RNA helicase protein': 0.32,\n",
       "              'Superoxide dismutase [Mn/Fe] (EC 1.15.1.1)': 0.32,\n",
       "              'T6SS PAAR-repeat protein / RhaS protein': 0.14,\n",
       "              'TPR domain protein': 0.25999999999999995,\n",
       "              'TPR repeat precursor': 0.3,\n",
       "              'TPR repeat:TPR repeat precursor': 0.38000000000000006,\n",
       "              'TPR-repeat-containing protein': 0.5399999999999999,\n",
       "              'Tetracenomycin polyketide synthesis O-methyltransferase TcmP': 0.32,\n",
       "              'Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)': 0.34,\n",
       "              'Thiol peroxidase, Bcp-type (EC 1.11.1.15)': 0.68,\n",
       "              'Thiol peroxidase, Tpx-type (EC 1.11.1.15)': 0.36000000000000004,\n",
       "              'Thiol:disulfide interchange protein': 0.48,\n",
       "              'Thiol:disulfide interchange protein tlpA': 0.4400000000000001,\n",
       "              'Thioredoxin': 1.9799999999999998,\n",
       "              'Thioredoxin domain-containing protein BT1583': 0.21999999999999997,\n",
       "              'Tll0814 protein': 0.45999999999999996,\n",
       "              'Tn5520-like integrase': 0.27999999999999997,\n",
       "              'TonB family protein': 0.5,\n",
       "              'TonB-dependent receptor': 0.21999999999999997,\n",
       "              'Transcription accessory protein (S1 RNA-binding domain)': 0.19999999999999998,\n",
       "              'Transcription regulator [contains diacylglycerol kinase catalytic domain]': 0.32,\n",
       "              'Transcription termination protein NusB': 0.38,\n",
       "              'Transcriptional regulator': 0.38000000000000006,\n",
       "              'Transcriptional regulator YeiE, LysR family': 0.1,\n",
       "              'Transcriptional regulator of fucose utilization, GntR family': 0.18,\n",
       "              'Transcriptional regulator, AcrR family': 0.52,\n",
       "              'Transcriptional regulator, AraC family': 2.9800000000000004,\n",
       "              'Transcriptional regulator, AsnC family': 0.5,\n",
       "              'Transcriptional regulator, LuxR family': 0.14,\n",
       "              'Transcriptional regulator, MarR family': 0.34,\n",
       "              'Transcriptional regulator, PadR family': 0.21999999999999997,\n",
       "              'Transcriptional repressor, BlaI/MecI family': 0.21999999999999997,\n",
       "              'Transglycosylase-associated protein': 0.45999999999999996,\n",
       "              'Translation initiation factor 3': 0.54,\n",
       "              'Translation initiation factor SUI1-related protein': 0.36,\n",
       "              'Transmembrane protein': 0.5199999999999999,\n",
       "              'Transporter, LysE family': 0.3,\n",
       "              'Transposase': 0.44,\n",
       "              'Trk potassium uptake system protein TrkH': 0.14,\n",
       "              'Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)': 0.33999999999999997,\n",
       "              'Twin-arginine translocation protein TatA': 0.38000000000000006,\n",
       "              'Two-component system sensor histidine kinase': 2.44,\n",
       "              'Two-component system sensor histidine kinase/response regulator hybrid': 0.18,\n",
       "              'Two-component transcriptional response regulator, LuxR family': 1.1,\n",
       "              'Two-component transcriptional response regulator, LuxR family BT1217': 0.5800000000000001,\n",
       "              'Two-component transcriptional response regulator, OmpR family': 0.3,\n",
       "              'Tyrosine type site-specific recombinase': 0.56,\n",
       "              'Tyrosine-protein kinase EpsD (EC 2.7.10.2)': 0.21999999999999997,\n",
       "              'UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)': 0.32,\n",
       "              'UDP-galactose-lipid carrier transferase (EC 2.-.-.-)': 0.19999999999999998,\n",
       "              'UDP-glucose 4-epimerase (EC 5.1.3.2)': 0.6199999999999999,\n",
       "              'UDP-glucuronate decarboxylase (EC 4.1.1.35)': 0.41999999999999993,\n",
       "              'UPF0056 inner membrane protein MarC': 0.48,\n",
       "              'UPF0126 inner membrane protein BT_1201': 0.5,\n",
       "              'UPF0145 protein YbjQ': 0.3,\n",
       "              'UPF0246 protein YaaA': 0.25999999999999995,\n",
       "              'UPF0301 protein YqgE': 0.21999999999999997,\n",
       "              'UPF0324 inner membrane protein YeiH': 0.19999999999999998,\n",
       "              'Uncharacterized MFS-type transporter': 1.7400000000000002,\n",
       "              'Uncharacterized membrane protein PA1577': 0.62,\n",
       "              'Uncharacterized membrane protein YsdA': 0.32,\n",
       "              'Uncharacterized protein BT1741': 0.21999999999999997,\n",
       "              'Uncharacterized protein BT4380': 0.37999999999999995,\n",
       "              'Uncharacterized protein BT_4646': 0.19999999999999998,\n",
       "              'Uncharacterized protein CHU_1311': 0.18,\n",
       "              'Uncharacterized protein TTC1240': 0.21999999999999997,\n",
       "              'Uncharacterized protein YgbA': 0.68,\n",
       "              'Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)': 0.6799999999999999,\n",
       "              'Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)': 0.27999999999999997,\n",
       "              'UpdZ protein': 0.41999999999999993,\n",
       "              'Uracil permease @ Uracil:proton symporter UraA': 0.21999999999999997,\n",
       "              'Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)': 0.19999999999999998,\n",
       "              'Uronate isomerase (EC 5.3.1.12)': 0.43999999999999995,\n",
       "              'UspA': 0.5399999999999999,\n",
       "              'V-type ATP synthase subunit C (EC 3.6.3.14)': 0.32,\n",
       "              'V-type ATP synthase subunit D (EC 3.6.3.14)': 0.32,\n",
       "              'V-type ATP synthase subunit E (EC 3.6.3.14)': 0.52,\n",
       "              'V-type ATP synthase subunit K (EC 3.6.3.14)': 0.19999999999999998,\n",
       "              'Vitamin B12 ABC transporter, permease protein BtuC': 0.58,\n",
       "              'Vitamin B12 ABC transporter, substrate-binding protein BtuF': 1.1199999999999999,\n",
       "              'Voltage-gated H(+)/2Cl(-) exchange transporter ClcA': 0.3,\n",
       "              'Xanthine phosphoribosyltransferase (EC 2.4.2.22)': 0.19999999999999998,\n",
       "              'Xylanase': 0.56,\n",
       "              'Xylulose kinase (EC 2.7.1.17)': 0.18,\n",
       "              'Ync': 0.5599999999999999,\n",
       "              'Zn-ribbon-containing, possibly RNA-binding protein and truncated derivatives': 0.45999999999999996,\n",
       "              '[FeFe]-hydrogenase maturation protein HydF': 0.4799999999999999,\n",
       "              'alpha-L-fucosidase (EC 3.2.1.51)': 0.25999999999999995,\n",
       "              'archaeal ATPase, fused to C-terminal DUF234 domain': 0.52,\n",
       "              'beta-N-acetylglucosaminidase (EC 3.2.1.52)': 0.3,\n",
       "              'beta-galactosidase (EC 3.2.1.23)': 0.25999999999999995,\n",
       "              'beta-glucosidase (EC 3.2.1.21)': 0.25999999999999995,\n",
       "              'beta-glycosyl hydrolase': 0.56,\n",
       "              'cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases': 1.24,\n",
       "              'cytochrome c biogenesis protein': 0.52,\n",
       "              'dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)': 0.41999999999999993,\n",
       "              'dTDP-4-dehydrorhamnose reductase (EC 1.1.1.133)': 0.52,\n",
       "              'galactosamine-containing minor teichoic acid biosynthesis protein': 0.25999999999999995,\n",
       "              'hypothetical domain / Cold shock protein of CSP family': 0.21999999999999997,\n",
       "              'hypothetical protein': 360.3799999999995,\n",
       "              'hypothetical protein; Some similarities with the C-terminal region of hemagglutinin/hemolysin-related protein and with purine NTPase': 0.23999999999999996,\n",
       "              'low molecular weight phosphatase supefamily': 0.19999999999999998,\n",
       "              'mobilization protein B': 0.5599999999999999,\n",
       "              'probable lipoprotein YPO2292': 0.18,\n",
       "              'putative DNA mismatch repair protein': 0.25999999999999995,\n",
       "              'putative N-acetylmuramoyl-L-alanine amidase': 0.19999999999999998,\n",
       "              'putative TonB-dependent outer membrane protein': 0.16,\n",
       "              'putative UndPP-QuiNAc-P-transferase': 0.54,\n",
       "              'putative acetyltransferase': 0.27999999999999997,\n",
       "              'putative capsule polysaccharide export protein': 0.32,\n",
       "              'putative integration host factor IHF alpha subunit': 0.49999999999999994,\n",
       "              'putative lipoprotein': 0.7000000000000001,\n",
       "              'putative membrane protein': 1.72,\n",
       "              'putative patatin-like phospholipase': 0.45999999999999996,\n",
       "              'putative phosphatase/phosphohexomutase': 0.21999999999999997,\n",
       "              'putative protease I': 0.27999999999999997,\n",
       "              'putative protein involved in capsular polysaccharide biosynthesis': 0.21999999999999997,\n",
       "              'putative pyrogenic exotoxin B': 0.21999999999999997,\n",
       "              'sodium/iodide co-transporter': 0.32,\n",
       "              'tRNA (5-methylaminomethyl-2-thiouridylate)-methyltransferase (EC 2.1.1.61)': 0.43999999999999995,\n",
       "              'tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)': 0.68,\n",
       "              'tRNA dimethylallyltransferase (EC 2.5.1.75)': 0.12000000000000001,\n",
       "              'tRNA(1)(Val) (adenine(37)-N(6))-methyltransferase (EC 2.1.1.223)': 0.3,\n",
       "              'tmRNA-binding protein SmpB': 0.6}),\n",
       " 'II-premie': defaultdict(int,\n",
       "             {'Aminoglycoside 6-nucleotidyltransferase (EC 2.7.7.-) => ANT(6)-I': 0,\n",
       "              'Mobilization protein MobA @ Conjugal transfer protein TraA': 0,\n",
       "              'TnpY-like protein': 0,\n",
       "              'hypothetical protein': 0}),\n",
       " 'MI-nonpremie': defaultdict(int, {}),\n",
       " 'MI-premie': defaultdict(int, {}),\n",
       " 'MM': defaultdict(int, {})}"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gained_genes_counts_nulls_by_tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "tp_types = ['AA', 'II-premie', 'II-nonpremie','MM', 'MI-premie', 'MI-nonpremie']\n",
    "num_gene_losses_by_tp_type = {tp_type: 0 for tp_type in tp_types}\n",
    "lost_genes_by_tp_type = {tp_type: set() for tp_type in tp_types}\n",
    "lost_genes_counts_by_tp_type = {tp_type: defaultdict(int) for tp_type in tp_types}\n",
    "lost_genes_counts_nulls_by_tp_type = {tp_type: defaultdict(int) for tp_type in tp_types}\n",
    "\n",
    "# Enumerate genes\n",
    "for species in modification_gene_losses:\n",
    "    for cohort in modification_gene_losses[species]:\n",
    "        for subject in modification_gene_losses[species][cohort]:\n",
    "            subdict = modification_gene_losses[species][cohort][subject]\n",
    "            subdict_null = modification_gene_losses_present_null[species][cohort][subject]\n",
    "            # For each sample pair in this cohort/species,\n",
    "            # have a count of how many gains of a particular gene\n",
    "            for tp_pair in subdict:\n",
    "                tp_type = tp_pair_to_tp_type(tp_pair, cohort)\n",
    "                for gene_id, gene_name in subdict[tp_pair]:\n",
    "                    gene_loss_count = subdict[tp_pair][(gene_id, gene_name)]\n",
    "                    gene_loss_count_null = subdict_null[tp_pair][(gene_id, gene_name)]\n",
    "                    num_gene_losses_by_tp_type[tp_type] += gene_gain_count\n",
    "                    lost_genes_by_tp_type[tp_type].add(gene_name)\n",
    "                    lost_genes_counts_by_tp_type[tp_type][gene_name] += gene_gain_count\n",
    "                    lost_genes_counts_nulls_by_tp_type[tp_type][gene_name] += gene_gain_count_null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AA': 17,\n",
       " 'II-nonpremie': 60,\n",
       " 'II-premie': 6,\n",
       " 'MI-nonpremie': 1792,\n",
       " 'MI-premie': 0,\n",
       " 'MM': 0}"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_gene_losses_by_tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7 AA\n",
      "5 II-premie\n",
      "0 MI-premie\n",
      "0 MM\n",
      "407 MI-nonpremie\n",
      "24 II-nonpremie\n"
     ]
    }
   ],
   "source": [
    "for tp_type in lost_genes_by_tp_type:\n",
    "    count = len(lost_genes_by_tp_type[tp_type])\n",
    "    print('%i %s' % (count, tp_type))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA\n",
      "11 hypothetical protein\n",
      "\n",
      "II-premie\n",
      "2 hypothetical protein\n",
      "\n",
      "MI-premie\n",
      "\n",
      "MM\n",
      "\n",
      "MI-nonpremie\n",
      "8 Transcriptional regulator, AraC family\n",
      "7 putative membrane protein\n",
      "7 Two-component system sensor histidine kinase\n",
      "7 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "7 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "6 Glycosyltransferase\n",
      "5 cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases\n",
      "5 Thioredoxin\n",
      "5 Putative anti-sigma factor\n",
      "4 Uncharacterized MFS-type transporter\n",
      "4 RNA-binding protein\n",
      "4 Putative helicase\n",
      "3 putative TonB-dependent receptor\n",
      "3 UpdZ protein\n",
      "3 Two-component transcriptional response regulator, LuxR family\n",
      "3 Transcriptional regulator, AcrR family\n",
      "3 Permease of the drug/metabolite transporter (DMT) superfamily\n",
      "3 O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)\n",
      "3 N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)\n",
      "3 Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids\n",
      "3 Lipopolysaccharide biosynthesis protein\n",
      "3 Efflux ABC transporter, permease protein\n",
      "3 Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain\n",
      "3 Acidobacterial duplicated orphan permease (function unknown)\n",
      "3 Acetyltransferase, GNAT family\n",
      "3 ABC-type antimicrobial peptide transport system, ATPase component\n",
      "2 tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)\n",
      "2 dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)\n",
      "2 [FeFe]-hydrogenase maturation protein HydF\n",
      "2 UDP-glucose 4-epimerase (EC 5.1.3.2)\n",
      "2 UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)\n",
      "2 Transcriptional regulatory protein rprY\n",
      "2 Transamidase GatB domain protein\n",
      "2 Tll0814 protein\n",
      "2 SclB protein\n",
      "2 SAM-dependent methyltransferase\n",
      "2 RidA/YER057c/UK114 superfamily protein\n",
      "2 Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)\n",
      "2 Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor\n",
      "2 Putative pre-16S rRNA nuclease YqgF\n",
      "2 Putative phosphohydrolase, Icc family\n",
      "2 Putative permease\n",
      "2 Putative periplasmic protein\n",
      "2 Putative membrane protease YugP\n",
      "2 Protein DUF2149, predicted transporter component\n",
      "2 Prolipoprotein diacylglyceryl transferase\n",
      "2 Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator\n",
      "2 Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)\n",
      "2 Nitroreductase family protein\n",
      "2 N5-carboxyaminoimidazole ribonucleotide mutase (EC 5.4.99.18)\n",
      "2 N-acetylmuramoyl-L-alanine amidase\n",
      "2 Multidrug resistance protein\n",
      "2 Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps\n",
      "2 Mannan endo-1,4-beta-mannosidase\n",
      "2 Lipoprotein signal peptidase (EC 3.4.23.36)\n",
      "2 Large-conductance mechanosensitive channel\n",
      "2 Integrase\n",
      "2 Helicase conserved C-terminal domain protein\n",
      "2 Glycosyl transferase, group 2 family\n",
      "2 Glycosyl transferase, group 1\n",
      "2 Glycosyl transferase, family 2\n",
      "2 FIG000605: protein co-occurring with transport systems (COG1739)\n",
      "2 Electron transport complex protein RnfB\n",
      "2 DnaD domain protein\n",
      "2 Diacylglycerol kinase (EC 2.7.1.107)\n",
      "2 D-alanyl-D-alanine carboxypeptidase (EC 3.4.16.4)\n",
      "2 Cytidylate kinase (EC 2.7.4.25)\n",
      "2 Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage\n",
      "2 Cell division ZapA family protein\n",
      "2 Cation efflux system protein\n",
      "2 Bacterial non-heme ferritin (EC 1.16.3.2)\n",
      "2 Aminotransferase, DegT/DnrJ/EryC1/StrS family\n",
      "2 ATP synthase F0 sector subunit c (EC 3.6.3.14)\n",
      "2 2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)\n",
      "2 2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)\n",
      "15 RNA polymerase ECF-type sigma factor\n",
      "1238 hypothetical protein\n",
      "\n",
      "II-nonpremie\n",
      "34 hypothetical protein\n",
      "3 Mobile element protein\n",
      "2 Similar to carbon monoxide dehydrogenase CooS subunit\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for tp_type in lost_genes_counts_by_tp_type:\n",
    "    print(tp_type)\n",
    "    strs = []\n",
    "    for gene_name in lost_genes_counts_by_tp_type[tp_type]:\n",
    "        count = lost_genes_counts_by_tp_type[tp_type][gene_name]\n",
    "        if count > 1:\n",
    "            strs.append(\"%i %s\" % (count, gene_name))\n",
    "    for string in sorted(strs, reverse=True):\n",
    "        print(string)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "gene_change_counts_by_tp_type = {}\n",
    "gene_change_null_counts_by_tp_type = {}\n",
    "\n",
    "for tp_type in lost_genes_counts_by_tp_type:\n",
    "    gene_change_counts_by_tp_type[tp_type] = {}\n",
    "    gene_change_null_counts_by_tp_type[tp_type] = {}\n",
    "    \n",
    "    for gene in set(lost_genes_counts_by_tp_type[tp_type].keys() + gained_genes_counts_by_tp_type[tp_type].keys()):\n",
    "        num_losses = 0 if gene not in lost_genes_counts_by_tp_type[tp_type] else lost_genes_counts_by_tp_type[tp_type][gene]        \n",
    "        num_gains = 0 if gene not in gained_genes_counts_by_tp_type[tp_type] else gained_genes_counts_by_tp_type[tp_type][gene]\n",
    "        gene_change_counts_by_tp_type[tp_type][gene] = num_losses + num_gains\n",
    "        \n",
    "        num_losses_null = 0 if gene not in lost_genes_counts_nulls_by_tp_type[tp_type] else lost_genes_counts_nulls_by_tp_type[tp_type][gene]        \n",
    "        num_gains_null = 0 if gene not in gained_genes_counts_nulls_by_tp_type[tp_type] else gained_genes_counts_nulls_by_tp_type[tp_type][gene]        \n",
    "        gene_change_null_counts_by_tp_type[tp_type][gene] = num_losses_null + num_gains_null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('hypothetical protein', 15), ('helicase, UvrD/Rep family', 1), ('Glycosyl transferase, family 2', 1), ('Glycosyltransferase', 1), ('LtrC-like protein', 1), ('FIG00897069: hypothetical protein', 1), ('Holin', 1)]\n",
      "[('hypothetical protein', 63), ('Phage protein', 3), ('Doubtful CDS. No database matches', 2), ('N-acetylglutamate kinase (EC 2.7.2.8)', 1), ('UPF0358 protein YlaN', 1), ('Biotin carboxyl carrier protein associated with SAR1687-8', 1), ('Lambdoid phage Rac integrase', 1), ('Uncharacterized protein YebG', 1), ('hypothetical protein within a prophage', 1), ('[Genomic island nu Sa beta2]', 1), ('Mobile element protein', 1), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 1), ('EsaC protein within ESAT-6 gene cluster (S.aureus type)', 1), ('Membrane protein insertion efficiency factor YidD', 1), ('Choline dehydrogenase (EC 1.1.99.1)', 1), ('ORF058', 1), ('Hypothetical protein SAV2259', 1), ('Copper/silver efflux RND transporter, transmembrane protein CusA', 1), ('Copper/silver efflux RND transporter, membrane fusion protein CusB', 1), ('putative membrane protein', 1), ('Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)', 1), ('Repetitive hypothetical protein in ESAT cluster, COG4495', 1), ('Aminoglycoside 6-nucleotidyltransferase (EC 2.7.7.-) => ANT(6)-I', 1), ('Uncharacterized sodium-dependent transporter YocS', 1), ('Mobilization protein MobA @ Conjugal transfer protein TraA', 1), ('FIG00630611: hypothetical protein', 1), ('LSU ribosomal protein L31p @ LSU ribosomal protein L31p, zinc-independent', 1), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 1), ('Extracellular adherence protein of broad specificity Eap/Map', 1), ('TnpY-like protein', 1), ('FIG010063: hypothetical protein', 1), ('Hypothetical protein SAV2286', 1), ('PTS system, galactitol-specific IIB component (EC 2.7.1.200)', 1), ('hypothetical protein transposon-related', 1)]\n",
      "[]\n",
      "[]\n",
      "[('hypothetical protein', 1239), ('RNA polymerase ECF-type sigma factor', 15), ('Transcriptional regulator, AraC family', 8), ('Two-component system sensor histidine kinase', 7), ('putative membrane protein', 7), ('Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family', 7), ('Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family', 7), ('Glycosyltransferase', 6), ('Putative anti-sigma factor', 5), ('cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases', 5), ('Thioredoxin', 5), ('RNA-binding protein', 4), ('Uncharacterized MFS-type transporter', 4), ('Putative helicase', 4), ('Lipopolysaccharide biosynthesis protein', 3), ('Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain', 3), ('N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)', 3), ('Acidobacterial duplicated orphan permease (function unknown)', 3), ('Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids', 3), ('Two-component transcriptional response regulator, LuxR family', 3), ('UpdZ protein', 3), ('Permease of the drug/metabolite transporter (DMT) superfamily', 3), ('Acetyltransferase, GNAT family', 3), ('ABC-type antimicrobial peptide transport system, ATPase component', 3), ('Transcriptional regulator, AcrR family', 3), ('O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)', 3), ('Efflux ABC transporter, permease protein', 3), ('putative TonB-dependent receptor', 3), ('tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)', 2), ('Protein DUF2149, predicted transporter component', 2), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 2), ('D-alanyl-D-alanine carboxypeptidase (EC 3.4.16.4)', 2), ('Diacylglycerol kinase (EC 2.7.1.107)', 2), ('N-acetylmuramoyl-L-alanine amidase', 2), ('ATP synthase F0 sector subunit c (EC 3.6.3.14)', 2), ('Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor', 2), ('Nitroreductase family protein', 2), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)', 2), ('N5-carboxyaminoimidazole ribonucleotide mutase (EC 5.4.99.18)', 2), ('Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps', 2), ('Putative membrane protease YugP', 2), ('Large-conductance mechanosensitive channel', 2), ('Multidrug resistance protein', 2), ('Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)', 2), ('RidA/YER057c/UK114 superfamily protein', 2), ('Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)', 2), ('Prolipoprotein diacylglyceryl transferase', 2), ('dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)', 2), ('Electron transport complex protein RnfB', 2), ('Putative permease', 2), ('UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)', 2), ('SclB protein', 2), ('2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)', 2), ('Lipoprotein signal peptidase (EC 3.4.23.36)', 2), ('SAM-dependent methyltransferase', 2), ('Putative periplasmic protein', 2), ('Integrase', 2), ('Tll0814 protein', 2), ('Glycosyl transferase, group 1', 2), ('DnaD domain protein', 2), ('Putative pre-16S rRNA nuclease YqgF', 2), ('Aminotransferase, DegT/DnrJ/EryC1/StrS family', 2), ('Glycosyl transferase, family 2', 2), ('Cell division ZapA family protein', 2), ('Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage', 2), ('Transcriptional regulatory protein rprY', 2), ('Cytidylate kinase (EC 2.7.4.25)', 2), ('Helicase conserved C-terminal domain protein', 2), ('[FeFe]-hydrogenase maturation protein HydF', 2), ('Mannan endo-1,4-beta-mannosidase', 2), ('Cation efflux system protein', 2), ('Putative phosphohydrolase, Icc family', 2), ('UDP-glucose 4-epimerase (EC 5.1.3.2)', 2), ('FIG000605: protein co-occurring with transport systems (COG1739)', 2), ('Glycosyl transferase, group 2 family', 2), ('Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator', 2), ('Transamidase GatB domain protein', 2), (\"5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\", 1), ('Na+-driven multidrug efflux pump', 1), ('Phospholipid ABC transporter ATP-binding protein MlaF', 1), ('Xanthine phosphoribosyltransferase (EC 2.4.2.22)', 1), ('Xylanase', 1), ('Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog', 1), ('ATP synthase epsilon chain (EC 3.6.3.14)', 1), ('CzcABC family efflux RND transporter, outer membrane protein', 1), ('dTDP-glucose 4,6-dehydratase (EC 4.2.1.46)', 1), ('probable DNA alkylation repair enzyme', 1), ('Phosphomethylpyrimidine synthase ThiC (EC 4.1.99.17)', 1), ('Beta-lactamase class C-like and penicillin binding proteins (PBPs) superfamily', 1), ('Low molecular weight protein tyrosine phosphatase (EC 3.1.3.48)', 1), ('Transglycosylase-associated protein', 1), ('Carboxynorspermidine synthase (EC 1.5.1.43)', 1), ('Na+/H+-dicarboxylate symporter', 1), ('Type I restriction-modification system, DNA-methyltransferase subunit M (EC 2.1.1.72)', 1), ('4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)', 1), ('PAP2 superfamily protein', 1), ('FIG015287: Zinc protease', 1), ('Arginyl-tRNA synthetase (EC 6.1.1.19)', 1), ('Exodeoxyribonuclease III (EC 3.1.11.2)', 1), ('Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)', 1), ('LemA protein', 1), ('Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)', 1), ('23S rRNA (pseudouridine(1915)-N(3))-methyltransferase (EC 2.1.1.177)', 1), ('alpha-L-fucosidase (EC 3.2.1.51)', 1), ('Endonuclease', 1), ('4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase', 1), ('Imidazole glycerol phosphate synthase amidotransferase subunit HisH', 1), ('hypothetical transporter PduT for various metalloporphyrins', 1), ('LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent', 1), ('beta-glycosyl hydrolase', 1), ('FIG016027: protein of unknown function YeaO', 1), ('Uncharacterized protease YdcP', 1), ('ATP synthase gamma chain (EC 3.6.3.14)', 1), ('Translation initiation factor 3', 1), ('FIG00936810: hypothetical protein', 1), ('N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)', 1), ('Phage integrase', 1), ('SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent', 1), ('2-dehydro-3-deoxy-D-gluconate 5-dehydrogenase (EC 1.1.1.127) @ 2-deoxy-D-gluconate 3-dehydrogenase (EC 1.1.1.125)', 1), ('FKBP-type peptidyl-prolyl cis-trans isomerase FkpA precursor (EC 5.2.1.8)', 1), ('Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)', 1), ('Fluoride ion transporter CrcB', 1), ('Trk potassium uptake system protein TrkH', 1), ('Probable integral membrane protein Cj1165c', 1), ('FIG01197961: hypothetical protein', 1), ('Nucleotidyltransferase domain protein, Caur_2868 group', 1), ('Signal recognition particle receptor FtsY', 1), ('LSU ribosomal protein L32p @ LSU ribosomal protein L32p, zinc-dependent', 1), ('Inner membrane protein YhiM', 1), ('Thioredoxin domain-containing protein BT1583', 1), ('tmRNA-binding protein SmpB', 1), ('Sensor histidine kinase involved in nitrogen fixation and metabolism regulation', 1), ('FHA domain protein', 1), ('Recombination protein RecR', 1), ('Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase', 1), ('Aminodeoxychorismate lyase (EC 4.1.3.38)', 1), ('dTDP-4-amino-4,6-dideoxygalactose transaminase (EC 2.6.1.59)', 1), ('Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)', 1), ('Ribosome-binding factor A', 1), ('Carbonic anhydrase, beta class (EC 4.2.1.1)', 1), ('putative acetyltransferase', 1), ('Carboxynorspermidine decarboxylase (EC 4.1.1.96)', 1), ('UPF0126 inner membrane protein BT_1201', 1), ('Cobalt/zinc/cadmium resistance protein CzcD', 1), ('putative acetyl transferase', 1), ('Alpha-glucosidase (EC 3.2.1.20)', 1), ('ABC export system, membrane fusion protein BT1213', 1), ('Small heat shock protein', 1), ('YgfD: protein that forms a complex with the methylmalonyl-CoA mutase in a pathway for conversion of succinyl-CoA to propionyl-CoA', 1), ('Lipoprotein spr precursor', 1), ('1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)', 1), ('Capsular polysaccharide transcription antitermination protein UphY', 1), ('Agmatine deiminase (EC 3.5.3.12)', 1), ('FIG00410332: hypothetical protein', 1), ('Fe-S oxidoreductase', 1), ('Glucose-1-phosphate thymidylyltransferase (EC 2.7.7.24)', 1), ('AMP nucleosidase (EC 3.2.2.4)', 1), ('Aminomethyltransferase (glycine cleavage system T protein) (EC 2.1.2.10)', 1), ('Monofunctional biosynthetic peptidoglycan transglycosylase', 1), ('HlpA protein', 1), ('N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)', 1), ('Polysaccharide export outer membrane protein', 1), ('Uncharacterized membrane protein PA1577', 1), ('putative integration host factor IHF alpha subunit', 1), ('UPF0309 protein SCO4393', 1), ('2-dehydro-3-deoxygluconokinase (EC 2.7.1.45)', 1), ('UspA', 1), ('RNA methyltransferase, TrmH family', 1), ('Beta-1,3-glucosyltransferase', 1), ('3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)', 1), ('LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent', 1), ('5-amino-6-(5-phosphoribosylamino)uracil reductase', 1), ('Transcriptional regulator', 1), ('LSU ribosomal protein L7p/L12p (P1/P2)', 1), ('Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)', 1), ('Type II restriction enzyme DpnII (EC 3.1.21.4) (Endonuclease DpnII) (R.DpnII)', 1), ('Conserved protein YuxK', 1), ('OmpA family outer membrane protein', 1), ('Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)', 1), ('Electron transport complex protein RnfG', 1), ('Electron transport complex protein RnfA', 1), ('SSU ribosomal protein S6p', 1), ('Acetyltransferase', 1), ('Oxidoreductase', 1), ('Thiol:disulfide interchange protein', 1), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 1), ('Tetratricopeptide repeat family protein', 1), ('putative glucose transferase', 1), ('Dethiobiotin synthase BioD (EC 6.3.3.3)', 1), ('Major outer membrane protein OmpA', 1), ('Isoleucyl-tRNA synthetase (EC 6.1.1.5)', 1), ('Transketolase (EC 2.2.1.1)', 1), ('Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria', 1), ('TPR-repeat-containing protein', 1), ('ATP synthase delta chain (EC 3.6.3.14)', 1), ('Capsular polysaccharide transcription antitermination protein, UpxY family', 1), ('Chromate transport protein', 1), ('Transcription accessory protein (S1 RNA-binding domain)', 1), ('Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)', 1), ('FIG00490011: hypothetical protein', 1), ('Transcriptional regulator, LacI family', 1), ('Repair of Iron Centers di-iron protein', 1), ('Polysaccharide deacetylase-like protein', 1), ('FIG01964566: Predicted membrane protein, hemolysin III homolog', 1), ('Aminobenzoyl-glutamate transport protein', 1), ('Two-component transcriptional response regulator, LuxR family BT1217', 1), ('Diadenylate cyclase spyDAC; Bacterial checkpoint controller DisA with nucleotide-binding domain', 1), ('Lipid carrier : UDP-N-acetylgalactosaminyltransferase (EC 2.4.1.-)', 1), ('L-threonine 3-dehydrogenase (EC 1.1.1.103)', 1), ('Inner membrane protein YihY, formerly thought to be RNase BN', 1), ('DnaK suppressor protein, putative', 1), ('L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)', 1), ('4-hydroxybenzoyl-CoA thioesterase (EC 3.1.2.23)', 1), ('Serine protease, DegP/HtrA, do-like (EC 3.4.21.-)', 1), ('Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)', 1), ('Pyridoxal phosphate-containing protein YggS', 1), ('Ribosomal silencing factor RsfA', 1), ('probable integral membrane protein Cj0341c', 1), ('putative alpha-1,6-mannanase', 1), ('Transcriptional regulator, AsnC family', 1), ('ATPase RavA', 1), ('predicted nucleotidyltransferase', 1), ('Iron-containing alcohol dehydrogenase', 1), ('Phosphoribosylanthranilate isomerase (EC 5.3.1.24)', 1), ('Clostripain-related protein', 1), ('Protein translocase membrane subunit SecG', 1), ('putative xylanase', 1), ('DnaJ-class molecular chaperone CbpA', 1), ('SanA protein', 1), ('Huntingtin interacting protein E-like protein', 1), ('Uncharacterized corrinoid protein', 1), ('Regulatory protein RecX', 1), ('Protoporphyrinogen IX oxidase, aerobic, HemY (EC 1.3.3.4)', 1), ('putative patatin-like phospholipase', 1), ('UDP-N-acetyl-D-mannosamine dehydrogenase (EC 1.1.1.336)', 1), ('LSU ribosomal protein L24p (L26e)', 1), ('Uridine kinase (EC 2.7.1.48)', 1), ('Phosphonopyruvate decarboxylase (EC 4.1.1.82)', 1), ('Toxin HigB / Protein kinase domain of HipA', 1), ('Phosphopantothenoylcysteine decarboxylase (EC 4.1.1.36) / Phosphopantothenoylcysteine synthetase (EC 6.3.2.5)', 1), ('putative pyrogenic exotoxin B', 1), ('ROK family sugar kinase or transcriptional regulator', 1), ('Phosphoglycerate mutase (EC 5.4.2.11)', 1), ('Integral membrane protein', 1), ('Adenylylsulfate kinase (EC 2.7.1.25)', 1), ('4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)', 1), ('NADH dehydrogenase (EC 1.6.99.3)', 1), ('Aspartate carbamoyltransferase regulatory chain (PyrI)', 1), ('Hemin receptor', 1), ('Fucose permease', 1), ('Pantothenate kinase type III, CoaX-like (EC 2.7.1.33)', 1), ('Zinc ABC transporter, ATP-binding protein ZnuC', 1), ('Glycosyltransferases involved in cell wall biogenesis', 1), ('LSU ribosomal protein L35p', 1), ('OmpA', 1), ('Conserved hypothetical protein 22', 1), ('Uncharacterized protein YgbA', 1), ('Chaperone protein HtpG', 1), ('Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)', 1), (\"UDP-glucuronate 5'-epimerase (EC 5.1.3.12)\", 1), ('Capsular polysaccharide transcription antitermination protein UpfY', 1), ('Fosmidomycin resistance protein', 1), ('Branched-chain phosphotransacylase (EC 2.3.1.- )', 1), ('Uncharacterized protein TTC1240', 1), ('N-acetylneuraminate lyase (EC 4.1.3.3)', 1), ('RND efflux system, inner membrane transporter', 1), ('Zinc finger domain-containing protein', 1), ('Sigma factor RpoE regulatory protein RseC', 1), ('Tyrosine-protein kinase (EC 2.7.10.2)', 1), ('Coproporphyrinogen III oxidase, oxygen-independent (EC 1.3.99.22)', 1), ('FIG00414141: hypothetical protein', 1), ('UPF0056 inner membrane protein MarC', 1), ('archaeal ATPase, fused to C-terminal DUF234 domain', 1), ('Possible exported heavy-metal binding protein', 1), ('LysM-repeat proteins and domains', 1), ('Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)', 1), ('Integrase protein', 1), ('Fic family protein', 1), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)', 1), ('4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)', 1), ('Putative phosphate ABC transporter, phosphate-binding component', 1), ('Colicin V production protein', 1), ('Amino acid-binding ACT', 1), ('2-isopropylmalate synthase (EC 2.3.3.13)', 1), ('Oxidoreductase, short-chain dehydrogenase/reductase family', 1), ('ATP synthase F0 sector subunit a (EC 3.6.3.14)', 1), ('L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)', 1), ('beta-glucosidase (EC 3.2.1.21)', 1), ('VgrG protein', 1), ('Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)', 1), ('Propionyl-CoA carboxylase biotin-containing subunit (EC 6.4.1.3)', 1), ('Efflux transport system, outer membrane factor (OMF) lipoprotein', 1), ('Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup', 1), ('FIG00003370: Multicopper polyphenol oxidase', 1), ('Twin-arginine translocation protein TatA', 1), ('LSU ribosomal protein L10p (P0)', 1), ('RNA polymerase sigma-70 factor', 1), ('DNA polymerase III delta subunit (EC 2.7.7.7)', 1), ('Uncharacterized di-Fe4-S4 cluster-containing protein AF_1153', 1), ('Phage terminase, large subunit', 1), ('Lipopolysaccharide export system protein LptC', 1), ('Integration host factor alpha subunit', 1), ('Uncharacterized protein BT1741', 1), ('Capsular polysaccharide transcription antitermination protein UpgY', 1), ('beta-galactosidase (EC 3.2.1.23)', 1), ('Uronate isomerase (EC 5.3.1.12)', 1), ('ATP-dependent Clp protease proteolytic subunit ClpP (EC 3.4.21.92)', 1), ('Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly', 1), ('Putative glycosyltransferase', 1), ('Hemophore HmuY, recruits heme from host protein carriers', 1), ('Nitrogen regulation protein NtrY (EC 2.7.3.-)', 1), ('Alpha-1,2-mannosidase', 1), ('DoxX family protein', 1), ('Arginine pathway regulatory protein ArgR, repressor of arg regulon', 1), ('Competence protein F homolog, phosphoribosyltransferase domain; protein YhgH required for utilization of DNA as sole source of carbon and energy', 1), ('16S rRNA (guanine(527)-N(7))-methyltransferase (EC 2.1.1.170)', 1), ('D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)', 1), ('DNA-binding protein, histone-like, putative', 1), ('Alkaline phosphatase (EC 3.1.3.1)', 1), ('Transmembrane protein', 1), ('TonB family protein', 1), ('LSU ribosomal protein L18p (L5e)', 1), ('LSU ribosomal protein L21p', 1), ('Vitamin B12 ABC transporter, substrate-binding protein BtuF', 1), ('Hcp transcriptional regulator HcpR (Crp/Fnr family)', 1), ('Oligopeptide transporter, OPT family', 1), ('Uncharacterized transcriptional regulator YozG, Cro/CI family', 1), ('Ribosome hibernation promoting factor Hpf', 1), ('Aspartate carbamoyltransferase (EC 2.1.3.2)', 1), ('Putative isomerase', 1), ('RNA pseudouridylate synthase, group 1', 1), ('Replication-associated recombination protein RarA', 1), ('FIG01257981: hypothetical protein', 1), ('HIT family protein', 1), ('FIG138056: a glutathione-dependent thiol reductase', 1), ('6-carboxy-5,6,7,8-tetrahydropterin synthase (EC 4.1.2.50)', 1), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis', 1), ('Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)', 1), ('Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)', 1), ('Uncharacterized integral membrane, glycosyltransferase', 1), ('Two-component system sensor histidine kinase/response regulator hybrid', 1), ('LSU ribosomal protein L22p (L17e)', 1), ('Stress responsive alpha-beta barrel domain protein Dabb', 1), ('Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)', 1), ('Aliphatic amidase AmiE (EC 3.5.1.4)', 1), ('Flavoredoxin', 1), ('Inner membrane protein RclC associated with response to reactive chlorine species', 1), ('putative N-acetylmuramoyl-L-alanine amidase', 1), ('Spermidine/putrescine import ABC transporter ATP-binding protein PotA (TC 3.A.1.11.1)', 1), ('Uncharacterized protein BT4380', 1), ('Pyruvate carboxylase (EC 6.4.1.1) / Biotin carboxyl carrier protein', 1), ('Serine protease', 1), ('Gll3207 protein', 1), ('Putative membrane-associated metal-dependent hydrolase', 1), ('ATP synthase protein I', 1), ('FIG034863: hypothetical protein', 1), ('Indole-3-glycerol phosphate synthase (EC 4.1.1.48)', 1), ('Transposase, IS4', 1), ('Putative alpha-1,6-mannanase', 1), ('UDP-glucose 6-dehydrogenase (EC 1.1.1.22)', 1), ('ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)', 1), ('DNA base-flipping protein', 1), ('MoxR-like ATPase in aerotolerance operon', 1), ('Gll2767 protein', 1), ('Quinolinate synthetase (EC 2.5.1.72)', 1), ('Methyl-directed repair DNA adenine methylase (EC 2.1.1.72)', 1), ('Peptide deformylase (EC 3.5.1.88)', 1), ('alpha-L-arabinofuranosidase (EC 3.2.1.55)', 1), ('Heat shock protein 10 kDa family chaperone GroES', 1), ('Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)', 1), ('Aldose 1-epimerase (EC 5.1.3.3)', 1), ('Aerotolerance protein BatB', 1), ('Aerotolerance protein BatC', 1), ('Aerotolerance protein BatA', 1), ('Aerotolerance protein BatD', 1), ('Aerotolerance protein BatE', 1), ('Tetrapyrrole methylase family protein', 1), ('LSU ribosomal protein L25p', 1), ('SSU ribosomal protein S4p (S9e) @ SSU ribosomal protein S4p (S9e), zinc-independent', 1), ('ABC transporter, permease protein', 1), ('Polymer-forming bactofilin', 1), ('3,4-dihydroxy-2-butanone 4-phosphate synthase (EC 4.1.99.12) / GTP cyclohydrolase II (EC 3.5.4.25)', 1), ('Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)', 1), ('Endo-1,4-beta-xylanase D', 1), ('putative general stress protein', 1), ('Vitamin B12 ABC transporter, permease protein BtuC', 1), ('probable lipoprotein YPO2292', 1), ('Exonuclease', 1), ('Uncharacterized protein Fjoh_4981, DUF4295-containing', 1), ('Holliday junction ATP-dependent DNA helicase RuvA (EC 3.6.4.12)', 1), ('Sugar-phosphate isomerase', 1), ('Glycosyl transferase, group 1 family protein', 1), ('Sugar-phosphate nucleotidyl transferase', 1), ('Slr1117 protein', 1), ('Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)', 1), ('DedA protein', 1), ('Peptide chain release factor homolog', 1), ('NifU-related domain containing protein', 1), ('LSU ribosomal protein L17p', 1), ('Anaerobic C4-dicarboxylate transporter DcuC', 1), ('L-alanyl-gamma-D-glutamyl-L-diamino acid endopeptidase', 1), ('Na(+)-translocating NADH-quinone reductase subunit E (EC 1.6.5.8)', 1), ('Uncharacterized protein CHU_1311', 1), ('SSU ribosomal protein S20p', 1), ('Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)', 1), ('Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO', 1), ('Zinc-regulated zinc transporting ATPase ZntA', 1)]\n",
      "[('hypothetical protein', 2296), ('RNA polymerase ECF-type sigma factor', 46), ('Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family', 16), ('Transcriptional regulator, AraC family', 15), ('Two-component system sensor histidine kinase', 14), ('Thioredoxin', 13), ('Putative anti-sigma factor', 12), ('putative membrane protein', 12), ('Acidobacterial duplicated orphan permease (function unknown)', 11), ('RNA-binding protein', 9), ('Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family', 9), ('Integrase', 8), ('Putative helicase', 8), ('Uncharacterized MFS-type transporter', 7), ('Two-component transcriptional response regulator, LuxR family', 6), ('Glycosyltransferase', 6), ('UpdZ protein', 6), ('cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases', 6), ('Mobile element protein', 6), ('N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)', 5), ('Capsular polysaccharide transcription antitermination protein, UpxY family', 5), ('SclB protein', 5), ('O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)', 5), ('Putative glycosyltransferase', 5), ('ABC-type antimicrobial peptide transport system, ATPase component', 5), ('UDP-glucose 4-epimerase (EC 5.1.3.2)', 4), ('Endonuclease', 4), ('Nitroreductase family protein', 4), ('MBL-fold metallo-hydrolase superfamily', 4), ('beta-glycosyl hydrolase', 4), ('Phage integrase', 4), ('Monofunctional biosynthetic peptidoglycan transglycosylase', 4), ('3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)', 4), ('CAAX amino terminal protease family protein', 4), ('TPR-repeat-containing protein', 4), ('N-acetylglucosaminyltransferase (EC 2.4.1.-)', 4), ('Putative periplasmic protein', 4), ('Oxidoreductase, aldo/keto reductase family', 4), ('Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids', 4), ('ISNCY family transposase', 4), ('DNA ligase (NAD(+)) (EC 6.5.1.2)', 4), ('Vitamin B12 ABC transporter, substrate-binding protein BtuF', 4), ('[FeFe]-hydrogenase maturation protein HydF', 4), ('Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)', 4), ('Flavoredoxin', 4), ('Efflux ABC transporter, permease protein', 4), ('Malonyl CoA-acyl carrier protein transacylase (EC 2.3.1.39)', 4), ('Tyrosine type site-specific recombinase', 4), ('Vitamin B12 ABC transporter, permease protein BtuC', 4), ('Zn-ribbon-containing, possibly RNA-binding protein and truncated derivatives', 3), ('tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)', 3), ('Protein DUF2149, predicted transporter component', 3), ('Xylanase', 3), ('ATP synthase epsilon chain (EC 3.6.3.14)', 3), ('Immunoreactive 53 kDa antigen PG123', 3), ('Na+/H+-dicarboxylate symporter', 3), ('Co-activator of prophage gene expression IbrB', 3), ('Diacylglycerol kinase (EC 2.7.1.107)', 3), ('4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)', 3), ('PAP2 superfamily protein', 3), ('Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)', 3), ('ATP synthase F0 sector subunit c (EC 3.6.3.14)', 3), ('FIG016027: protein of unknown function YeaO', 3), ('ATP synthase gamma chain (EC 3.6.3.14)', 3), ('N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)', 3), ('Probable integral membrane protein Cj1165c', 3), ('Inner membrane protein YhiM', 3), ('tmRNA-binding protein SmpB', 3), ('Aminodeoxychorismate lyase (EC 4.1.3.38)', 3), ('Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)', 3), ('UPF0126 inner membrane protein BT_1201', 3), ('Small heat shock protein', 3), ('Large-conductance mechanosensitive channel', 3), ('Conserved hypothetical protein 22', 3), ('Possible glyoxylase family protein (Lactoylglutathione lyase) (EC 4.4.1.5)', 3), ('Multidrug resistance protein', 3), ('cytochrome c biogenesis protein', 3), ('RidA/YER057c/UK114 superfamily protein', 3), ('N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)', 3), ('putative integration host factor IHF alpha subunit', 3), ('dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)', 3), ('Conserved protein YuxK', 3), ('Glycosyl transferase, family 2', 3), ('Oxidoreductase', 3), ('Thiol:disulfide interchange protein', 3), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 3), ('GTP cyclohydrolase 1 type 2 homolog YbgI @ Bsu YqfO NIF3/CutA domain', 3), ('Hypothetical Nudix-like regulator', 3), ('Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)', 3), ('FIG00490011: hypothetical protein', 3), ('Integrase IntN1', 3), ('Putative permease', 3), ('Two-component transcriptional response regulator, LuxR family BT1217', 3), ('RNA methyltransferase, TrmH family', 3), ('Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)', 3), ('Ribosomal silencing factor RsfA', 3), ('Transcriptional regulator, AsnC family', 3), ('Helicase, C-terminal:Type III restriction enzyme, res subunit:DEAD/DEAH box helicase, N-terminal', 3), ('Permease of the drug/metabolite transporter (DMT) superfamily', 3), ('2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)', 3), ('Putative outer membrane protein, probably involved in nutrient binding', 3), ('Lipoprotein signal peptidase (EC 3.4.23.36)', 3), ('Outer membrane protein romA', 3), ('Free methionine-(R)-sulfoxide reductase, contains GAF domain', 3), ('putative patatin-like phospholipase', 3), ('Xylulose kinase (EC 2.7.1.17)', 3), ('ROK family sugar kinase or transcriptional regulator', 3), ('Integral membrane protein', 3), ('Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria', 3), ('Uncharacterized protein YgbA', 3), ('Tll0814 protein', 3), ('Fosmidomycin resistance protein', 3), ('Rrf2-linked NADH-flavin reductase', 3), ('RND efflux system, inner membrane transporter', 3), ('Sigma factor RpoE regulatory protein RseC', 3), ('Aminotransferase, DegT/DnrJ/EryC1/StrS family', 3), ('Transcriptional regulator, AcrR family', 3), ('UPF0056 inner membrane protein MarC', 3), ('Possible exported heavy-metal binding protein', 3), ('Membrane protein', 3), ('4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)', 3), ('FIG137478: Hypothetical protein', 3), ('DNA polymerase III delta subunit (EC 2.7.7.7)', 3), ('CzcABC family efflux RND transporter, outer membrane protein', 3), ('Integration host factor alpha subunit', 3), ('Uronate isomerase (EC 5.3.1.12)', 3), ('OmpA family outer membrane protein', 3), ('Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage', 3), ('DoxX family protein', 3), ('Acetyltransferase, GNAT family', 3), ('Transmembrane protein', 3), ('TonB family protein', 3), ('ATP-dependent RNA helicase', 3), ('D-alanine--poly(phosphoribitol) ligase ACP subunit (EC 6.1.1.13)', 3), ('Stress responsive alpha-beta barrel domain protein Dabb', 3), ('Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)', 3), ('Uncharacterized protein BT4380', 3), ('ATP synthase protein I', 3), ('Indole-3-glycerol phosphate synthase (EC 4.1.1.48)', 3), ('Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)', 3), ('Aerotolerance protein BatB', 3), ('Aerotolerance protein BatC', 3), ('Aerotolerance protein BatD', 3), ('Aerotolerance protein BatE', 3), ('Endo-1,4-beta-xylanase D', 3), ('Phosphoribosylformimino-5-aminoimidazole carboxamide ribotide isomerase (EC 5.3.1.16)', 3), ('Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)', 3), ('Slr1117 protein', 3), ('Cys-tRNA(Pro) deacylase YbaK', 3), ('FIG000605: protein co-occurring with transport systems (COG1739)', 3), ('Probable iron export ATP-binding protein FetA', 3), ('Putative thioredoxin', 3), ('Serine protease', 3), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)', 3), ('Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO', 3), ('Uncharacterized membrane protein PA1577', 3), ('Translation initiation factor SUI1-related protein', 2), (\"5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\", 2), ('Cobalt-precorrin-7 (C5)-methyltransferase (EC 2.1.1.289) / Cobalt-precorrin-6B C15-methyltransferase [decarboxylating] (EC 2.1.1.196)', 2), ('Na+-driven multidrug efflux pump', 2), ('putative lipoprotein', 2), ('Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog', 2), ('Glycosyltransferase involved in cell wall biogenesis', 2), ('Beta-lactamase class C-like and penicillin binding proteins (PBPs) superfamily', 2), ('Ync', 2), ('6-phosphogluconate dehydrogenase, decarboxylating (EC 1.1.1.44)', 2), ('Long-chain-fatty-acid--CoA ligase (EC 6.2.1.3)', 2), ('Putative pre-16S rRNA nuclease YqgF', 2), ('Protein BF2544, predicted transporter component', 2), ('Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)', 2), ('Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor', 2), ('4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase', 2), ('Imidazole glycerol phosphate synthase amidotransferase subunit HisH', 2), (\"Alpha-ribazole-5'-phosphate phosphatase (EC 3.1.3.73)\", 2), ('Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain', 2), ('Translation initiation factor 3', 2), ('FIG00936810: hypothetical protein', 2), ('Sortase A, LPXTG specific', 2), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)', 2), ('tRNA (5-methylaminomethyl-2-thiouridylate)-methyltransferase (EC 2.1.1.61)', 2), ('4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)', 2), ('Putative DNA-binding protein', 2), ('SSU ribosomal protein S20p', 2), ('DNA-binding protein HU-beta', 2), ('Fluoride ion transporter CrcB', 2), ('Trk potassium uptake system protein TrkH', 2), ('Signal recognition particle receptor FtsY', 2), ('LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent', 2), ('LSU ribosomal protein L34p', 2), ('FHA domain protein', 2), ('Aspartate carbamoyltransferase regulatory chain (PyrI)', 2), ('Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase', 2), ('Polyribonucleotide nucleotidyltransferase (EC 2.7.7.8)', 2), ('Isoleucyl-tRNA synthetase (EC 6.1.1.5)', 2), ('Ribose-5-phosphate isomerase B (EC 5.3.1.6)', 2), ('Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps', 2), ('Serine/threonine kinase', 2), ('Uracil permease @ Uracil:proton symporter UraA', 2), ('Agmatine deiminase (EC 3.5.3.12)', 2), ('Putative dipeptidase', 2), ('putative acetyltransferase', 2), ('putative UndPP-QuiNAc-P-transferase', 2), ('FIG00491264: hypothetical protein', 2), ('UspA', 2), ('ATP synthase alpha chain (EC 3.6.3.14)', 2), ('6-phosphogluconolactonase (EC 3.1.1.31), eukaryotic type', 2), ('UDP-glucuronate decarboxylase (EC 4.1.1.35)', 2), ('Cobalt-precorrin-2 C(20)-methyltransferase (EC 2.1.1.151)', 2), ('LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent', 2), ('5-amino-6-(5-phosphoribosylamino)uracil reductase', 2), ('Transcriptional regulator', 2), ('Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)', 2), ('sodium/iodide co-transporter', 2), ('Electron transport complex protein RnfA', 2), ('Electron transport complex protein RnfB', 2), ('SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent', 2), ('Dethiobiotin synthase BioD (EC 6.3.3.3)', 2), ('Chromate transport protein', 2), ('Transcription accessory protein (S1 RNA-binding domain)', 2), ('Polysaccharide deacetylase-like protein', 2), ('Cytochrome d ubiquinol oxidase subunit II (EC 1.10.3.-)', 2), ('Aminobenzoyl-glutamate transport protein', 2), ('Inner membrane protein YihY, formerly thought to be RNase BN', 2), ('low molecular weight phosphatase supefamily', 2), ('DnaK suppressor protein, putative', 2), ('Methylmalonyl-CoA epimerase (EC 5.1.99.1)', 2), ('Pyridoxal phosphate-containing protein YggS', 2), ('Formate-dependent phosphoribosylglycinamide formyltransferase', 2), ('Arylsulfatase (EC 3.1.6.1)', 2), ('tRNA(1)(Val) (adenine(37)-N(6))-methyltransferase (EC 2.1.1.223)', 2), ('DNA/RNA helicases, SNF2 family', 2), ('ATPase RavA', 2), ('Transmembrane component YkoC of energizing module of thiamin-regulated ECF transporter for HydroxyMethylPyrimidine', 2), ('Adenosylcobinamide-phosphate synthase (EC 6.3.1.10)', 2), ('dTDP-4-dehydrorhamnose reductase (EC 1.1.1.133)', 2), ('Transcription regulator [contains diacylglycerol kinase catalytic domain]', 2), ('FIG138056: a glutathione-dependent thiol reductase', 2), ('Phosphoribosylanthranilate isomerase (EC 5.3.1.24)', 2), ('Clostripain-related protein', 2), ('SanA protein', 2), ('SAM-dependent methyltransferase', 2), ('Sulfur carrier protein ThiS adenylyltransferase (EC 2.7.7.73)', 2), ('Phosphonopyruvate decarboxylase (EC 4.1.1.82)', 2), ('Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)', 2), ('putative pyrogenic exotoxin B', 2), ('Adenylylsulfate kinase (EC 2.7.1.25)', 2), ('Recombination protein RecR', 2), ('Hemin receptor', 2), ('OmpA', 2), ('Glyoxalase family protein', 2), ('UPF0324 inner membrane protein YeiH', 2), ('Chaperone protein HtpG', 2), ('Rubredoxin', 2), ('Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)', 2), ('Carboxyl-terminal protease', 2), ('Glycosylase BF2339', 2), ('FIG00414141: hypothetical protein', 2), ('archaeal ATPase, fused to C-terminal DUF234 domain', 2), ('Hcp transcriptional regulator HcpR (Crp/Fnr family)', 2), ('Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)', 2), ('Integrase protein', 2), ('AMP nucleosidase (EC 3.2.2.4)', 2), ('probable lipoprotein YPO2292', 2), ('ATP synthase F0 sector subunit a (EC 3.6.3.14)', 2), ('L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)', 2), ('Efflux transport system, outer membrane factor (OMF) lipoprotein', 2), ('FIG00003370: Multicopper polyphenol oxidase', 2), ('Adenosine deaminase (EC 3.5.4.4)', 2), ('Pentapeptide repeat family protein', 2), ('Phage terminase, large subunit', 2), ('Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)', 2), ('Superoxide dismutase [Mn/Fe] (EC 1.15.1.1)', 2), ('Similar to carbon monoxide dehydrogenase CooS subunit', 2), ('Outer membrane assembly lipoprotein YfiO', 2), ('LemA protein', 2), ('Putative phage-encoded peptidoglycan binding protein', 2), ('Peroxide stress regulator; Ferric uptake regulation protein; Fe2+/Zn2+ uptake regulation proteins', 2), ('Murein endolytic transglycosylase MltG', 2), ('D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)', 2), ('Alkaline phosphatase (EC 3.1.3.1)', 2), ('Ribosomal protein S12p Asp88 (E. coli) methylthiotransferase (EC 2.8.4.4)', 2), ('Lipopolysaccharide export system permease protein LptG', 2), ('Oligopeptide transporter, OPT family', 2), ('Amino acid-binding ACT', 2), ('Ribose-phosphate pyrophosphokinase (EC 2.7.6.1)', 2), ('ATP-dependent DNA helicase', 2), ('Transcription termination protein NusB', 2), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis', 2), ('Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)', 2), ('Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)', 2), ('Two-component system sensor histidine kinase/response regulator hybrid', 2), ('Cation efflux system protein', 2), ('putative N-acetylmuramoyl-L-alanine amidase', 2), ('V-type ATP synthase subunit E (EC 3.6.3.14)', 2), ('Transglycosylase-associated protein', 2), ('3-deoxy-manno-octulosonate cytidylyltransferase (EC 2.7.7.38)', 2), ('Putative membrane-associated metal-dependent hydrolase', 2), ('Putative alpha-1,6-mannanase', 2), ('ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)', 2), ('Transcriptional regulator, PadR family', 2), ('DNA base-flipping protein', 2), ('Thiol peroxidase, Bcp-type (EC 1.11.1.15)', 2), ('V-type ATP synthase subunit K (EC 3.6.3.14)', 2), ('Aerotolerance protein BatA', 2), ('Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup', 2), ('Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)', 2), ('Putative calcineurin superfamily phosphohydrolase', 2), ('LysM-repeat proteins and domains', 2), ('Sugar-phosphate nucleotidyl transferase', 2), ('mobilization protein B', 2), ('Transposase', 2), ('Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)', 2), ('Glycosyl transferase, group 2 family', 2), ('Cytidylyltransferase domain / Phosphoenolpyruvate phosphomutase (EC 5.4.2.9)', 2), ('ATP synthase beta chain (EC 3.6.3.14)', 2), ('Similar to conjugative transposon protein TraB', 2), ('Uncharacterized protein CHU_1311', 2), ('Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)', 2), ('Glutamate formiminotransferase (EC 2.1.2.5) @ Glutamate formyltransferase', 1), ('Uncharacterized membrane protein YsdA', 1), ('putative capsule polysaccharide export protein', 1), ('UPF0301 protein YqgE', 1), ('3-hydroxyacyl-[acyl-carrier-protein] dehydratase, FabZ form (EC 4.2.1.59)', 1), ('Diaminohydroxyphosphoribosylaminopyrimidine deaminase (EC 3.5.4.26)', 1), ('Acyl carrier protein', 1), ('Riboflavin synthase eubacterial/eukaryotic (EC 2.5.1.9)', 1), ('Phosphomethylpyrimidine synthase ThiC (EC 4.1.99.17)', 1), ('Transcriptional regulator, MerR family', 1), ('Cobyric acid synthase (EC 6.3.5.10)', 1), ('2-amino-3-ketobutyrate coenzyme A ligase (EC 2.3.1.29)', 1), ('Anthranilate synthase, aminase component (EC 4.1.3.27)', 1), ('Substrate-specific component RibU of riboflavin ECF transporter', 1), ('Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly', 1), ('Type I restriction-modification system, DNA-methyltransferase subunit M (EC 2.1.1.72)', 1), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 1), ('Carboxy-terminal processing protease', 1), ('ABC-type antimicrobial peptide transport system, ATPase component BT1214', 1), ('Heat shock protein GrpE', 1), ('GTP-binding protein EngB', 1), ('Ribonuclease PH (EC 2.7.7.56)', 1), ('Manganese-dependent protein-tyrosine phosphatase (EC 3.1.3.48)', 1), ('Translation elongation factor Ts', 1), ('T6SS PAAR-repeat protein / RhaS protein', 1), ('UPF0246 protein YaaA', 1), ('AAA+ superfamily protein', 1), ('Ser-tRNA(Ala) deacylase @ Gly-tRNA(Ala) deacylase', 1), ('2-iminoacetate synthase (ThiH) (EC 4.1.99.19)', 1), ('Bacterial ribosome SSU maturation protein RimP', 1), ('Lipopolysaccharide biosynthesis protein', 1), ('Arabinan endo-1,5-alpha-L-arabinosidase A', 1), ('Uncharacterized UPF0721 integral membrane protein', 1), ('alpha-L-fucosidase (EC 3.2.1.51)', 1), ('Alpha-1,2-fucosyltransferase', 1), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, alpha subunit (EC 1.2.7.-)', 1), ('Error-prone, lesion bypass DNA polymerase V (UmuC)', 1), ('putative phosphatase/phosphohexomutase', 1), ('hypothetical transporter PduT for various metalloporphyrins', 1), ('Oligopeptide ABC transporter, ATP-binding protein OppF (TC 3.A.1.5.1)', 1), ('UDP-galactose-lipid carrier transferase (EC 2.-.-.-)', 1), ('Acetyl esterase', 1), ('Transcriptional regulator, LuxR family', 1), ('Pantoate--beta-alanine ligase (EC 6.3.2.1)', 1), ('Biotin carboxylase of acetyl-CoA carboxylase (EC 6.3.4.14)', 1), ('ABC transporter, permease protein 1 (cluster 1, maltose/g3p/polyamine/iron)', 1), ('Predicted transcriptional regulator', 1), ('Methylated-DNA--protein-cysteine methyltransferase', 1), ('Voltage-gated H(+)/2Cl(-) exchange transporter ClcA', 1), ('L-fucose kinase (EC 2.7.1.52)', 1), ('Polysaccharide deacetylase', 1), ('Efflux transport system, outer membrane factor (OMF) lipoprotein BT_0954', 1), ('Succinyl-CoA ligase [ADP-forming] beta chain (EC 6.2.1.5)', 1), ('3-oxo-5-alpha-steroid 4-dehydrogenase', 1), ('PKD domain-containing protein', 1), ('Anthranilate phosphoribosyltransferase (EC 2.4.2.18)', 1), ('Prolyl oligopeptidase family protein', 1), ('FKBP-type peptidyl-prolyl cis-trans isomerase FkpA precursor (EC 5.2.1.8)', 1), ('RNA pseudouridylate synthase BT0642', 1), ('hypothetical domain / Cold shock protein of CSP family', 1), ('Guanine deaminase (EC 3.5.4.3)', 1), ('Putative manganese efflux pump MntP', 1), ('Cystathionine beta-lyase (EC 4.4.1.8)', 1), ('hypothetical protein; Some similarities with the C-terminal region of hemagglutinin/hemolysin-related protein and with purine NTPase', 1), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, gamma subunit (EC 1.2.7.-)', 1), ('RNA methyltransferase, TrmA family', 1), ('Sodium/proline symporter', 1), ('Thioredoxin domain-containing protein BT1583', 1), ('Sensor histidine kinase involved in nitrogen fixation and metabolism regulation', 1), ('TPR repeat precursor', 1), ('2-methylfumaryl-CoA hydratase (EC 4.2.1.148)', 1), ('Phage lysin, 1,4-beta-N-acetylmuramidase (EC 3.2.1.17) or lysozyme', 1), ('Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)', 1), ('Capsular polysaccharide biosynthesis protein capD', 1), ('OsmC/Ohr family protein', 1), ('Quinolinate phosphoribosyltransferase [decarboxylating] (EC 2.4.2.19)', 1), ('Queuine tRNA-ribosyltransferase (EC 2.4.2.29)', 1), ('DNA-binding protein', 1), ('Lysozyme (EC 3.2.1.17)', 1), ('Transcriptional repressor, BlaI/MecI family', 1), ('Carboxynorspermidine decarboxylase (EC 4.1.1.96)', 1), ('Lipopolysaccharide cholinephosphotransferase LicD1 (EC 2.7.8.-)', 1), ('putative acetyl transferase', 1), ('Alpha-glucosidase (EC 3.2.1.20)', 1), (\"Inosine-5'-monophosphate dehydrogenase (EC 1.1.1.205) / CBS domain\", 1), ('ABC export system, membrane fusion protein BT1213', 1), ('Uncharacterized protein BT_4646', 1), ('Peptidase, S41 family', 1), ('Serine palmitoyltransferase (EC 2.3.1.50)', 1), ('Oxaloacetate decarboxylase Na(+) pump, alpha chain (EC 4.1.1.3)', 1), ('Integration host factor beta subunit', 1), ('Helicase', 1), ('1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)', 1), ('Intracellular protease', 1), ('Substrate-specific component YkoE of thiamin-regulated ECF transporter for HydroxyMethylPyrimidine', 1), ('FIG01416087: hypothetical protein', 1), ('Phage protein', 1), ('putative TonB-dependent outer membrane protein', 1), ('Probable integral membrane protein Cj1166c', 1), ('Septum formation protein Maf', 1), ('Glucose-1-phosphate thymidylyltransferase (EC 2.7.7.24)', 1), ('Superfamily I DNA and RNA helicases and helicase subunits', 1), ('Formate--tetrahydrofolate ligase (EC 6.3.4.3)', 1), ('Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)', 1), ('Arylsulfatase regulator', 1), ('putative glycosyl transferase domain protein', 1), ('Polysaccharide export outer membrane protein', 1), ('Lead, cadmium, zinc and mercury transporting ATPase (EC 3.6.3.3) (EC 3.6.3.5); Copper-translocating P-type ATPase (EC 3.6.3.4)', 1), ('Acetyl-coenzyme A carboxyl transferase beta chain (EC 6.4.1.2)', 1), ('ATP-dependent Clp protease proteolytic subunit ClpP (EC 3.4.21.92)', 1), ('ADP-ribose pyrophosphatase of COG1058 family (EC 3.6.1.13) / Nicotinamide-nucleotide amidase (EC 3.5.1.42)', 1), ('Conjugative transposon protein TraM', 1), ('Transcriptional regulator of fucose utilization, GntR family', 1), ('Conjugative transposon protein TraN', 1), ('Hydrolase, alpha/beta fold family', 1), ('DNA topoisomerase IV subunit B (EC 5.99.1.3)', 1), ('Conjugative transposon protein TraE', 1), ('Beta-1,3-glucosyltransferase', 1), ('ATP synthase F0 sector subunit b (EC 3.6.3.14)', 1), ('N-acetylmannosaminyltransferase (EC 2.4.1.187)', 1), ('ABC transporter, ATP-binding protein', 1), ('Tetracenomycin polyketide synthesis O-methyltransferase TcmP', 1), ('V-type ATP synthase subunit D (EC 3.6.3.14)', 1), ('putative beta-glucosidase', 1), ('Flavodoxin', 1), ('Electron transport complex protein RnfD', 1), ('Electron transport complex protein RnfG', 1), ('SSU ribosomal protein S6p', 1), ('Acetyltransferase', 1), ('AmpG protein', 1), ('UPF0145 protein YbjQ', 1), ('Cell envelope-associated transcriptional attenuator LytR-CpsA-Psr, subfamily F2', 1), ('Succinyl-CoA ligase [ADP-forming] alpha chain (EC 6.2.1.5)', 1), ('Pyruvate:Oxaloacetate transcarboxylase domain protein / Biotin carboxyl carrier protein', 1), ('N-carbamoylputrescine amidase (EC 3.5.1.53)', 1), ('Superfamily I DNA/RNA helicase protein', 1), ('Capsular polysaccharide transcription antitermination protein UphY', 1), ('Energy-dependent translational throttle protein EttA', 1), (\"16S rRNA (cytidine(1402)-2'-O)-methyltransferase (EC 2.1.1.198)\", 1), ('Putative ATP-binding protein', 1), ('Cell surface protein', 1), ('Transporter, LysE family', 1), ('Long tail fiber protein p37', 1), ('Type II restriction enzyme HpaII (EC 3.1.21.4)', 1), ('CRISPR-associated protein, TM1814 family', 1), ('Diadenylate cyclase spyDAC; Bacterial checkpoint controller DisA with nucleotide-binding domain', 1), ('Putative secreted polysaccharide polymerase', 1), ('TadB-like protein BAD_0130', 1), ('Exopolyphosphatase (EC 3.6.1.11)', 1), ('Oligo-1,6-glucosidase (EC 3.2.1.10)', 1), ('Choloylglycine hydrolase family protein', 1), ('Heat shock protein HtpX', 1), ('UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)', 1), ('LemA family protein', 1), ('Na(+)-translocating NADH-quinone reductase subunit E (EC 1.6.5.8)', 1), ('Imidazole glycerol phosphate synthase cyclase subunit', 1), ('Alpha-1,2-mannosidase', 1), ('Type I restriction-modification system, specificity subunit S', 1), ('NADH-ubiquinone oxidoreductase chain K (EC 1.6.5.3)', 1), ('Protein translocase membrane subunit SecG', 1), ('putative xylanase', 1), ('FIG015287: Zinc protease', 1), ('DnaJ-class molecular chaperone CbpA', 1), ('Huntingtin interacting protein E-like protein', 1), ('Integrase, catalytic region', 1), ('SSU ribosomal protein S12p (S23e)', 1), ('Uncharacterized corrinoid protein', 1), ('Mlc, transcriptional repressor of MalT (the transcriptional activator of maltose regulon) and manXYZ operon', 1), ('DNA helicase (EC 3.6.4.12), mobile-element-associated', 1), ('4-carboxymuconolactone decarboxylase (EC 4.1.1.44)', 1), ('Rod shape-determining protein MreD', 1), ('CRISPR-associated protein Cas1', 1), ('LSU ribosomal protein L24p (L26e)', 1), ('Toxin HigB / Protein kinase domain of HipA', 1), ('Related to F420H2-dehydrogenase, beta subunit', 1), ('Putative lipopolysaccharide biosynthesis protein', 1), ('Protein translocase subunit YajC', 1), ('Thioredoxin family protein', 1), ('Serine protease, DegP/HtrA, do-like (EC 3.4.21.-)', 1), ('Endonuclease III (EC 4.2.99.18)', 1), ('Tryptophan synthase beta chain (EC 4.2.1.20)', 1), ('Zinc ABC transporter, ATP-binding protein ZnuC', 1), ('Xylose isomerase (EC 5.3.1.5)', 1), ('Acetylhydrolase', 1), ('Sulfate adenylyltransferase subunit 1 (EC 2.7.7.4)', 1), ('Xanthine phosphoribosyltransferase (EC 2.4.2.22)', 1), ('Glycosyl transferase, group 1', 1), ('Indolepyruvate oxidoreductase subunit IorB (EC 1.2.7.8)', 1), ('Protein of unknown function UPF0060', 1), ('LtrC-like protein', 1), ('Uncharacterized protein TTC1240', 1), ('V-type ATP synthase subunit C (EC 3.6.3.14)', 1), ('Thiazole synthase (EC 2.8.1.10)', 1), ('N-acetylneuraminate lyase (EC 4.1.3.3)', 1), ('Putative molybdenum transport ATP-binding protein modF', 1), ('Lipid-A-disaccharide synthase (EC 2.4.1.182)', 1), ('DNA protection during starvation protein', 1), ('Ribosomal protein L11 methyltransferase', 1), ('Antirestriction protein ArdA', 1), ('Ferredoxin', 1), ('Capsular polysaccharide transcription antitermination protein UpeY', 1), ('NADH-dependent glycerate dehydrogenase', 1), ('FIG00672316: hypothetical protein', 1), ('Tyrosine-protein kinase EpsD (EC 2.7.10.2)', 1), ('Diaminopimelate epimerase (EC 5.1.1.7)', 1), ('Putative phosphatidylinositol-4-phosphate 5-kinase', 1), ('beta-galactosidase (EC 3.2.1.23)', 1), ('Keratan-sulfate endo-1,4-beta-galactosidase (EC 3.2.1.103)', 1), ('CRISPR-associated negative autoregulator Cas7/Cst2', 1), ('Putative sodium-dependent transporter', 1), ('Lactaldehyde reductase (EC 1.1.1.77)', 1), ('Sulfur carrier protein ThiS', 1), ('Pyrrolidone-carboxylate peptidase (EC 3.4.19.3)', 1), ('L,L-diaminopimelate aminotransferase (EC 2.6.1.83)', 1), ('D-xylose proton-symporter XylE', 1), ('Oxidoreductase, short-chain dehydrogenase/reductase family', 1), (\"Inosine-5'-monophosphate dehydrogenase, catalytic domain (EC 1.1.1.205)\", 1), ('Tn5520-like integrase', 1), ('beta-glucosidase (EC 3.2.1.21)', 1), ('VgrG protein', 1), ('Ferredoxin domain containing protein', 1), ('Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)', 1), ('Thiol:disulfide interchange protein tlpA', 1), ('Propionyl-CoA carboxylase biotin-containing subunit (EC 6.4.1.3)', 1), ('NADPH-dependent 7-cyano-7-deazaguanine reductase (EC 1.7.1.13)', 1), ('Tryptophan synthase alpha chain (EC 4.2.1.20)', 1), ('DnaD domain protein', 1), ('Sugar-phosphate isomerase', 1), ('Chaperone protein DnaJ', 1), ('Methionine transporter MetT', 1), ('Efflux transport system, outer membrane factor (OMF) lipoprotein BT1212', 1), ('Two-component transcriptional response regulator, OmpR family', 1), ('Fructose-1,6-bisphosphatase, Bacillus type (EC 3.1.3.11)', 1), ('Ribosomal-protein-S5p-alanine acetyltransferase (EC 2.3.1.128)', 1), ('RND efflux system, membrane fusion protein', 1), ('Alpha-N-acetylglucosaminidase (EC 3.2.1.50)', 1), ('TPR repeat:TPR repeat precursor', 1), ('beta-N-acetylglucosaminidase (EC 3.2.1.52)', 1), ('putative oxidoreductase, nitronate monooxygenase family', 1), ('alpha-galactosidase (EC 3.2.1.22)', 1), ('Uncharacterized protein BT1741', 1), ('Phospho-N-acetylmuramoyl-pentapeptide-transferase (EC 2.7.8.13)', 1), ('Acyltransferase', 1), ('Cobalamin synthase (EC 2.7.8.26)', 1), ('Uncharacterized DUF349-containing protein SCO1511', 1), ('Cytochrome c551 peroxidase (EC 1.11.1.5)', 1), ('TonB-dependent receptor', 1), ('Dephospho-CoA kinase (EC 2.7.1.24)', 1), ('Nicotinate-nucleotide--dimethylbenzimidazole phosphoribosyltransferase (EC 2.4.2.21)', 1), ('Hemophore HmuY, recruits heme from host protein carriers', 1), ('Nitrogen regulation protein NtrY (EC 2.7.3.-)', 1), ('tRNA dimethylallyltransferase (EC 2.5.1.75)', 1), ('Inorganic pyrophospatase PpaX', 1), ('Cell division ZapA family protein', 1), ('Ribonuclease G', 1), ('TPR domain protein', 1), ('Tyrosine-protein kinase (EC 2.7.10.2)', 1), ('putative protease I', 1), ('putative protein involved in capsular polysaccharide biosynthesis', 1), ('Putative aminotransferase BF2110', 1), ('Ribosome hibernation promoting factor Hpf', 1), ('Exonuclease SbcD', 1), ('Putative preQ0 transporter YhhQ', 1), ('Lipoate-protein ligase A', 1), ('Putative Na+/H+ exchange protein', 1), ('Short-chain alcohol dehydrogenase', 1), ('Carbon starvation protein A', 1), ('Iron-containing alcohol dehydrogenase', 1), ('6-carboxy-5,6,7,8-tetrahydropterin synthase (EC 4.1.2.50)', 1), ('N-acetylneuraminate synthase (EC 2.5.1.56)', 1), ('PTS system, IIB component', 1), ('Hydroxymethylpyrimidine phosphate kinase ThiD (EC 2.7.4.7)', 1), ('Uncharacterized integral membrane, glycosyltransferase', 1), ('Possible restriction /modification enzyme', 1), ('Aliphatic amidase AmiE (EC 3.5.1.4)', 1), ('Transcriptional regulator, MarR family', 1), ('Mannan endo-1,4-beta-mannosidase', 1), ('galactosamine-containing minor teichoic acid biosynthesis protein', 1), ('Spermidine/putrescine import ABC transporter ATP-binding protein PotA (TC 3.A.1.11.1)', 1), ('Putative phosphohydrolase, Icc family', 1), ('NAD(P)H-hydrate epimerase (EC 5.1.99.6) / ADP-dependent (S)-NAD(P)H-hydrate dehydratase (EC 4.2.1.136)', 1), ('L-alanyl-gamma-D-glutamyl-L-diamino acid endopeptidase', 1), ('Gll3207 protein', 1), ('Transcriptional regulator YeiE, LysR family', 1), ('ATP phosphoribosyltransferase (EC 2.4.2.17) => HisGl', 1), ('Alpha-1,4-N-acetylgalactosamine transferase PglH (EC 2.4.1.-)', 1), ('FIG034863: hypothetical protein', 1), ('Iron-sulfur flavoprotein', 1), ('Uncharacterized ABC1 family protein Bbr_0249', 1), ('Ribonucleotide reductase of class Ia (aerobic), beta subunit (EC 1.17.4.1)', 1), ('L-fucose mutarotase (EC 5.1.3.29)', 1), ('Death on curing protein, Doc toxin', 1), ('Mannose-1-phosphate guanylyltransferase (EC 2.7.7.13)', 1), ('ABC transporter, permease protein 2 (cluster 1, maltose/g3p/polyamine/iron)', 1), ('3-isopropylmalate dehydratase small subunit (EC 4.2.1.33)', 1), ('Gll2767 protein', 1), ('Quinolinate synthetase (EC 2.5.1.72)', 1), ('LSU ribosomal protein L29p (L35e)', 1), ('Peptide deformylase (EC 3.5.1.88)', 1), ('Hydroxylamine reductase (EC 1.7.99.1)', 1), ('Phosphoglycerate mutase family 5', 1), ('Biotin carboxyl carrier protein of acetyl-CoA carboxylase', 1), ('putative DNA mismatch repair protein', 1), ('ATP-dependent RNA helicase BT1154', 1), ('Outer membrane efflux protein precursor', 1), ('Mannose-6-phosphate isomerase (EC 5.3.1.8)', 1), ('Maltose O-acetyltransferase (EC 2.3.1.79)', 1), ('ABC transporter, permease protein', 1), ('putative TonB-dependent receptor', 1), ('Hypothetical radical SAM family enzyme in interesting gene cluster', 1), ('CRISPR-associated RecB family exonuclease Cas4', 1), ('GldB', 1), ('Probable iron export permease protein FetB', 1), ('CobN-like chelatase BtuS for metalloporphyrine salvage', 1), ('Serine hydroxymethyltransferase (EC 2.1.2.1)', 1), ('tRNA-Arg-CCT', 1), ('CRISPR-associated helicase Cas3', 1), ('Glycosyl transferase, group 1 family protein', 1), ('Flavodoxin 1', 1), ('Inner membrane protein RclC associated with response to reactive chlorine species', 1), ('3-oxoacyl-[acyl-carrier protein] reductase (EC 1.1.1.100), FadG', 1), ('Glucosamine-1-phosphate guanylyltransferase', 1), ('Pyruvate formate-lyase activating enzyme (EC 1.97.1.4)', 1), ('Peptide chain release factor homolog', 1), ('Capsular polysaccharide transcription antitermination protein UpfY', 1), ('Pirin', 1), ('Aspartate carbamoyltransferase (EC 2.1.3.2)', 1), ('8-amino-7-oxononanoate synthase (EC 2.3.1.47)', 1), ('Twin-arginine translocation protein TatA', 1), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, beta subunit (EC 1.2.7.-)', 1), ('FAD:protein FMN transferase (EC 2.7.1.180)', 1), ('Acetoin dehydrogenase E1 component alpha-subunit (EC 2.3.1.190) / Acetoin dehydrogenase E1 component beta-subunit (EC 2.3.1.190)', 1), ('Outer membrane receptor for Fe3+-dicitrate', 1), ('N-carbamoylputrescine amidase (EC 3.5.1.53) @ Aliphatic amidase AmiE (EC 3.5.1.4)', 1), ('Predicted cobalt transporter in Bacteroides_Porphyromonas', 1), ('Thiol peroxidase, Tpx-type (EC 1.11.1.15)', 1), ('N-acetylgalactosamine-6-sulfatase', 1), ('MutT/Nudix family protein', 1)]\n"
     ]
    }
   ],
   "source": [
    "for tp_type in gene_change_counts_by_tp_type:\n",
    "    print(sorted(gene_change_counts_by_tp_type[tp_type].items(), key=lambda x: x[1], reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('helicase, UvrD/Rep family', 0), ('Glycosyl transferase, family 2', 0), ('Glycosyltransferase', 0), ('LtrC-like protein', 0), ('FIG00897069: hypothetical protein', 0), ('Holin', 0), ('hypothetical protein', 0)]\n",
      "[('hypothetical protein', 0.9400000000000004), ('Phage protein', 0.08), ('hypothetical protein transposon-related', 0.08), ('LSU ribosomal protein L31p @ LSU ribosomal protein L31p, zinc-independent', 0.06), ('UPF0358 protein YlaN', 0.04), ('EsaC protein within ESAT-6 gene cluster (S.aureus type)', 0.04), ('Hypothetical protein SAV2259', 0.04), ('Hypothetical protein SAV2286', 0.04), ('PTS system, galactitol-specific IIB component (EC 2.7.1.200)', 0.04), ('N-acetylglutamate kinase (EC 2.7.2.8)', 0.02), ('Biotin carboxyl carrier protein associated with SAR1687-8', 0.02), ('Uncharacterized protein YebG', 0.02), ('hypothetical protein within a prophage', 0.02), ('Choline dehydrogenase (EC 1.1.99.1)', 0.02), ('ORF058', 0.02), ('Repetitive hypothetical protein in ESAT cluster, COG4495', 0.02), ('Uncharacterized sodium-dependent transporter YocS', 0.02), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 0.02), ('FIG010063: hypothetical protein', 0.02), ('Lambdoid phage Rac integrase', 0), ('[Genomic island nu Sa beta2]', 0), ('Mobile element protein', 0), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 0), ('Membrane protein insertion efficiency factor YidD', 0), ('Copper/silver efflux RND transporter, transmembrane protein CusA', 0), ('Copper/silver efflux RND transporter, membrane fusion protein CusB', 0), ('putative membrane protein', 0), ('Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)', 0), ('Aminoglycoside 6-nucleotidyltransferase (EC 2.7.7.-) => ANT(6)-I', 0), ('Doubtful CDS. No database matches', 0), ('Mobilization protein MobA @ Conjugal transfer protein TraA', 0), ('FIG00630611: hypothetical protein', 0), ('Extracellular adherence protein of broad specificity Eap/Map', 0), ('TnpY-like protein', 0)]\n",
      "[]\n",
      "[]\n",
      "[('tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)', 0), (\"5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\", 0), ('Protein DUF2149, predicted transporter component', 0), ('Na+-driven multidrug efflux pump', 0), ('Phospholipid ABC transporter ATP-binding protein MlaF', 0), ('Xanthine phosphoribosyltransferase (EC 2.4.2.22)', 0), ('Xylanase', 0), ('Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog', 0), ('ATP synthase epsilon chain (EC 3.6.3.14)', 0), ('CzcABC family efflux RND transporter, outer membrane protein', 0), ('dTDP-glucose 4,6-dehydratase (EC 4.2.1.46)', 0), ('probable DNA alkylation repair enzyme', 0), ('Phosphomethylpyrimidine synthase ThiC (EC 4.1.99.17)', 0), ('Beta-lactamase class C-like and penicillin binding proteins (PBPs) superfamily', 0), ('Low molecular weight protein tyrosine phosphatase (EC 3.1.3.48)', 0), ('Transglycosylase-associated protein', 0), ('RNA-binding protein', 0), ('Carboxynorspermidine synthase (EC 1.5.1.43)', 0), ('Na+/H+-dicarboxylate symporter', 0), ('Type I restriction-modification system, DNA-methyltransferase subunit M (EC 2.1.1.72)', 0), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 0), ('D-alanyl-D-alanine carboxypeptidase (EC 3.4.16.4)', 0), ('Diacylglycerol kinase (EC 2.7.1.107)', 0), ('4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)', 0), ('PAP2 superfamily protein', 0), ('FIG015287: Zinc protease', 0), ('N-acetylmuramoyl-L-alanine amidase', 0), ('Arginyl-tRNA synthetase (EC 6.1.1.19)', 0), ('Exodeoxyribonuclease III (EC 3.1.11.2)', 0), ('Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)', 0), ('LemA protein', 0), ('Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)', 0), ('Lipopolysaccharide biosynthesis protein', 0), ('23S rRNA (pseudouridine(1915)-N(3))-methyltransferase (EC 2.1.1.177)', 0), ('alpha-L-fucosidase (EC 3.2.1.51)', 0), ('ATP synthase F0 sector subunit c (EC 3.6.3.14)', 0), ('Endonuclease', 0), ('Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor', 0), ('4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase', 0), ('Imidazole glycerol phosphate synthase amidotransferase subunit HisH', 0), ('Nitroreductase family protein', 0), ('hypothetical transporter PduT for various metalloporphyrins', 0), ('LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent', 0), ('beta-glycosyl hydrolase', 0), ('Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain', 0), ('FIG016027: protein of unknown function YeaO', 0), ('Uncharacterized protease YdcP', 0), ('ATP synthase gamma chain (EC 3.6.3.14)', 0), ('Translation initiation factor 3', 0), ('FIG00936810: hypothetical protein', 0), ('N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)', 0), ('Phage integrase', 0), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)', 0), ('SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent', 0), ('2-dehydro-3-deoxy-D-gluconate 5-dehydrogenase (EC 1.1.1.127) @ 2-deoxy-D-gluconate 3-dehydrogenase (EC 1.1.1.125)', 0), ('FKBP-type peptidyl-prolyl cis-trans isomerase FkpA precursor (EC 5.2.1.8)', 0), ('Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)', 0), ('Fluoride ion transporter CrcB', 0), ('Trk potassium uptake system protein TrkH', 0), ('N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)', 0), ('Probable integral membrane protein Cj1165c', 0), ('FIG01197961: hypothetical protein', 0), ('Nucleotidyltransferase domain protein, Caur_2868 group', 0), ('Signal recognition particle receptor FtsY', 0), ('LSU ribosomal protein L32p @ LSU ribosomal protein L32p, zinc-dependent', 0), ('Inner membrane protein YhiM', 0), ('Acidobacterial duplicated orphan permease (function unknown)', 0), ('Thioredoxin domain-containing protein BT1583', 0), ('N5-carboxyaminoimidazole ribonucleotide mutase (EC 5.4.99.18)', 0), ('tmRNA-binding protein SmpB', 0), ('Sensor histidine kinase involved in nitrogen fixation and metabolism regulation', 0), ('FHA domain protein', 0), ('Recombination protein RecR', 0), ('Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase', 0), ('Aminodeoxychorismate lyase (EC 4.1.3.38)', 0), ('dTDP-4-amino-4,6-dideoxygalactose transaminase (EC 2.6.1.59)', 0), ('Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)', 0), ('Ribosome-binding factor A', 0), ('Carbonic anhydrase, beta class (EC 4.2.1.1)', 0), ('putative acetyltransferase', 0), ('Carboxynorspermidine decarboxylase (EC 4.1.1.96)', 0), ('UPF0126 inner membrane protein BT_1201', 0), ('Cobalt/zinc/cadmium resistance protein CzcD', 0), ('putative acetyl transferase', 0), ('Alpha-glucosidase (EC 3.2.1.20)', 0), ('ABC export system, membrane fusion protein BT1213', 0), ('Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps', 0), ('Small heat shock protein', 0), ('YgfD: protein that forms a complex with the methylmalonyl-CoA mutase in a pathway for conversion of succinyl-CoA to propionyl-CoA', 0), ('Lipoprotein spr precursor', 0), ('Putative membrane protease YugP', 0), ('1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)', 0), ('Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids', 0), ('Capsular polysaccharide transcription antitermination protein UphY', 0), ('Large-conductance mechanosensitive channel', 0), ('Agmatine deiminase (EC 3.5.3.12)', 0), ('FIG00410332: hypothetical protein', 0), ('Fe-S oxidoreductase', 0), ('Two-component system sensor histidine kinase', 0), ('Multidrug resistance protein', 0), ('Glucose-1-phosphate thymidylyltransferase (EC 2.7.7.24)', 0), ('Two-component transcriptional response regulator, LuxR family', 0), ('AMP nucleosidase (EC 3.2.2.4)', 0), ('Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)', 0), ('Aminomethyltransferase (glycine cleavage system T protein) (EC 2.1.2.10)', 0), ('RidA/YER057c/UK114 superfamily protein', 0), ('Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)', 0), ('Monofunctional biosynthetic peptidoglycan transglycosylase', 0), ('HlpA protein', 0), ('N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)', 0), ('Polysaccharide export outer membrane protein', 0), ('Uncharacterized membrane protein PA1577', 0), ('Glycosyltransferase', 0), ('putative integration host factor IHF alpha subunit', 0), ('UPF0309 protein SCO4393', 0), ('2-dehydro-3-deoxygluconokinase (EC 2.7.1.45)', 0), ('UspA', 0), ('Prolipoprotein diacylglyceryl transferase', 0), ('RNA methyltransferase, TrmH family', 0), ('Beta-1,3-glucosyltransferase', 0), ('3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)', 0), ('LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent', 0), ('dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)', 0), ('5-amino-6-(5-phosphoribosylamino)uracil reductase', 0), ('Transcriptional regulator', 0), ('LSU ribosomal protein L7p/L12p (P1/P2)', 0), ('Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)', 0), ('Type II restriction enzyme DpnII (EC 3.1.21.4) (Endonuclease DpnII) (R.DpnII)', 0), ('Conserved protein YuxK', 0), ('OmpA family outer membrane protein', 0), ('Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)', 0), ('Electron transport complex protein RnfG', 0), ('Electron transport complex protein RnfA', 0), ('Electron transport complex protein RnfB', 0), ('SSU ribosomal protein S6p', 0), ('Acetyltransferase', 0), ('Oxidoreductase', 0), ('Thiol:disulfide interchange protein', 0), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 0), ('Tetratricopeptide repeat family protein', 0), ('putative glucose transferase', 0), ('Dethiobiotin synthase BioD (EC 6.3.3.3)', 0), ('Major outer membrane protein OmpA', 0), ('Isoleucyl-tRNA synthetase (EC 6.1.1.5)', 0), ('Putative anti-sigma factor', 0), ('Transketolase (EC 2.2.1.1)', 0), ('Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria', 0), ('TPR-repeat-containing protein', 0), ('ATP synthase delta chain (EC 3.6.3.14)', 0), ('Capsular polysaccharide transcription antitermination protein, UpxY family', 0), ('Chromate transport protein', 0), ('Transcription accessory protein (S1 RNA-binding domain)', 0), ('Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)', 0), ('FIG00490011: hypothetical protein', 0), ('Transcriptional regulator, LacI family', 0), ('Repair of Iron Centers di-iron protein', 0), ('Putative permease', 0), ('Polysaccharide deacetylase-like protein', 0), ('FIG01964566: Predicted membrane protein, hemolysin III homolog', 0), ('Aminobenzoyl-glutamate transport protein', 0), ('Two-component transcriptional response regulator, LuxR family BT1217', 0), ('Diadenylate cyclase spyDAC; Bacterial checkpoint controller DisA with nucleotide-binding domain', 0), ('Lipid carrier : UDP-N-acetylgalactosaminyltransferase (EC 2.4.1.-)', 0), ('UpdZ protein', 0), ('L-threonine 3-dehydrogenase (EC 1.1.1.103)', 0), ('Inner membrane protein YihY, formerly thought to be RNase BN', 0), ('DnaK suppressor protein, putative', 0), ('L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)', 0), ('4-hydroxybenzoyl-CoA thioesterase (EC 3.1.2.23)', 0), ('Serine protease, DegP/HtrA, do-like (EC 3.4.21.-)', 0), ('Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)', 0), ('Pyridoxal phosphate-containing protein YggS', 0), ('Ribosomal silencing factor RsfA', 0), ('probable integral membrane protein Cj0341c', 0), ('putative alpha-1,6-mannanase', 0), ('UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)', 0), ('Transcriptional regulator, AsnC family', 0), ('ATPase RavA', 0), ('SclB protein', 0), ('Uncharacterized MFS-type transporter', 0), ('Permease of the drug/metabolite transporter (DMT) superfamily', 0), ('2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)', 0), ('Acetyltransferase, GNAT family', 0), ('predicted nucleotidyltransferase', 0), ('Iron-containing alcohol dehydrogenase', 0), ('Lipoprotein signal peptidase (EC 3.4.23.36)', 0), ('ABC-type antimicrobial peptide transport system, ATPase component', 0), ('Phosphoribosylanthranilate isomerase (EC 5.3.1.24)', 0), ('Clostripain-related protein', 0), ('Protein translocase membrane subunit SecG', 0), ('putative xylanase', 0), ('DnaJ-class molecular chaperone CbpA', 0), ('SanA protein', 0), ('Huntingtin interacting protein E-like protein', 0), ('SAM-dependent methyltransferase', 0), ('Uncharacterized corrinoid protein', 0), ('RNA polymerase ECF-type sigma factor', 0), ('Regulatory protein RecX', 0), ('Protoporphyrinogen IX oxidase, aerobic, HemY (EC 1.3.3.4)', 0), ('putative membrane protein', 0), ('putative patatin-like phospholipase', 0), ('UDP-N-acetyl-D-mannosamine dehydrogenase (EC 1.1.1.336)', 0), ('Transcriptional regulator, AraC family', 0), ('LSU ribosomal protein L24p (L26e)', 0), ('Uridine kinase (EC 2.7.1.48)', 0), ('Phosphonopyruvate decarboxylase (EC 4.1.1.82)', 0), ('Toxin HigB / Protein kinase domain of HipA', 0), ('Phosphopantothenoylcysteine decarboxylase (EC 4.1.1.36) / Phosphopantothenoylcysteine synthetase (EC 6.3.2.5)', 0), ('Putative periplasmic protein', 0), ('putative pyrogenic exotoxin B', 0), ('ROK family sugar kinase or transcriptional regulator', 0), ('Phosphoglycerate mutase (EC 5.4.2.11)', 0), ('Integral membrane protein', 0), ('Adenylylsulfate kinase (EC 2.7.1.25)', 0), ('4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)', 0), ('NADH dehydrogenase (EC 1.6.99.3)', 0), ('Aspartate carbamoyltransferase regulatory chain (PyrI)', 0), ('Hemin receptor', 0), ('Fucose permease', 0), ('Pantothenate kinase type III, CoaX-like (EC 2.7.1.33)', 0), ('Zinc ABC transporter, ATP-binding protein ZnuC', 0), ('Glycosyltransferases involved in cell wall biogenesis', 0), ('LSU ribosomal protein L35p', 0), ('OmpA', 0), ('Conserved hypothetical protein 22', 0), ('Integrase', 0), ('Uncharacterized protein YgbA', 0), ('Chaperone protein HtpG', 0), ('Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)', 0), (\"UDP-glucuronate 5'-epimerase (EC 5.1.3.12)\", 0), ('Capsular polysaccharide transcription antitermination protein UpfY', 0), ('Tll0814 protein', 0), ('Glycosyl transferase, group 1', 0), ('Fosmidomycin resistance protein', 0), ('Branched-chain phosphotransacylase (EC 2.3.1.- )', 0), ('Uncharacterized protein TTC1240', 0), ('N-acetylneuraminate lyase (EC 4.1.3.3)', 0), ('RND efflux system, inner membrane transporter', 0), ('Zinc finger domain-containing protein', 0), ('Sigma factor RpoE regulatory protein RseC', 0), ('Tyrosine-protein kinase (EC 2.7.10.2)', 0), ('Transcriptional regulator, AcrR family', 0), ('Coproporphyrinogen III oxidase, oxygen-independent (EC 1.3.99.22)', 0), ('FIG00414141: hypothetical protein', 0), ('UPF0056 inner membrane protein MarC', 0), ('archaeal ATPase, fused to C-terminal DUF234 domain', 0), ('Possible exported heavy-metal binding protein', 0), ('LysM-repeat proteins and domains', 0), ('Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)', 0), ('Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family', 0), ('O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)', 0), ('Integrase protein', 0), ('Fic family protein', 0), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)', 0), ('4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)', 0), ('Putative phosphate ABC transporter, phosphate-binding component', 0), ('Colicin V production protein', 0), ('Amino acid-binding ACT', 0), ('2-isopropylmalate synthase (EC 2.3.3.13)', 0), ('Oxidoreductase, short-chain dehydrogenase/reductase family', 0), ('ATP synthase F0 sector subunit a (EC 3.6.3.14)', 0), ('L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)', 0), ('beta-glucosidase (EC 3.2.1.21)', 0), ('VgrG protein', 0), ('Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)', 0), ('Propionyl-CoA carboxylase biotin-containing subunit (EC 6.4.1.3)', 0), ('Efflux transport system, outer membrane factor (OMF) lipoprotein', 0), ('Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup', 0), ('FIG00003370: Multicopper polyphenol oxidase', 0), ('DnaD domain protein', 0), ('Twin-arginine translocation protein TatA', 0), ('LSU ribosomal protein L10p (P0)', 0), ('RNA polymerase sigma-70 factor', 0), ('DNA polymerase III delta subunit (EC 2.7.7.7)', 0), ('Uncharacterized di-Fe4-S4 cluster-containing protein AF_1153', 0), ('Phage terminase, large subunit', 0), ('Putative pre-16S rRNA nuclease YqgF', 0), ('Aminotransferase, DegT/DnrJ/EryC1/StrS family', 0), ('Lipopolysaccharide export system protein LptC', 0), ('Integration host factor alpha subunit', 0), ('Uncharacterized protein BT1741', 0), ('Capsular polysaccharide transcription antitermination protein UpgY', 0), ('beta-galactosidase (EC 3.2.1.23)', 0), ('Uronate isomerase (EC 5.3.1.12)', 0), ('ATP-dependent Clp protease proteolytic subunit ClpP (EC 3.4.21.92)', 0), ('Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly', 0), ('Putative glycosyltransferase', 0), ('Glycosyl transferase, family 2', 0), ('Hemophore HmuY, recruits heme from host protein carriers', 0), ('Nitrogen regulation protein NtrY (EC 2.7.3.-)', 0), ('Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family', 0), ('Alpha-1,2-mannosidase', 0), ('hypothetical protein', 0), ('Cell division ZapA family protein', 0), ('Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage', 0), ('DoxX family protein', 0), ('Arginine pathway regulatory protein ArgR, repressor of arg regulon', 0), ('Competence protein F homolog, phosphoribosyltransferase domain; protein YhgH required for utilization of DNA as sole source of carbon and energy', 0), ('16S rRNA (guanine(527)-N(7))-methyltransferase (EC 2.1.1.170)', 0), ('D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)', 0), ('DNA-binding protein, histone-like, putative', 0), ('Alkaline phosphatase (EC 3.1.3.1)', 0), ('Transmembrane protein', 0), ('TonB family protein', 0), ('Transcriptional regulatory protein rprY', 0), ('LSU ribosomal protein L18p (L5e)', 0), ('LSU ribosomal protein L21p', 0), ('Vitamin B12 ABC transporter, substrate-binding protein BtuF', 0), ('Hcp transcriptional regulator HcpR (Crp/Fnr family)', 0), ('Oligopeptide transporter, OPT family', 0), ('Uncharacterized transcriptional regulator YozG, Cro/CI family', 0), ('Ribosome hibernation promoting factor Hpf', 0), ('Aspartate carbamoyltransferase (EC 2.1.3.2)', 0), ('Cytidylate kinase (EC 2.7.4.25)', 0), ('Helicase conserved C-terminal domain protein', 0), ('Putative isomerase', 0), ('RNA pseudouridylate synthase, group 1', 0), ('Replication-associated recombination protein RarA', 0), ('[FeFe]-hydrogenase maturation protein HydF', 0), ('FIG01257981: hypothetical protein', 0), ('HIT family protein', 0), ('FIG138056: a glutathione-dependent thiol reductase', 0), ('6-carboxy-5,6,7,8-tetrahydropterin synthase (EC 4.1.2.50)', 0), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis', 0), ('Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)', 0), ('Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)', 0), ('Uncharacterized integral membrane, glycosyltransferase', 0), ('Two-component system sensor histidine kinase/response regulator hybrid', 0), ('LSU ribosomal protein L22p (L17e)', 0), ('Stress responsive alpha-beta barrel domain protein Dabb', 0), ('Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)', 0), ('Aliphatic amidase AmiE (EC 3.5.1.4)', 0), ('Flavoredoxin', 0), ('cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases', 0), ('Inner membrane protein RclC associated with response to reactive chlorine species', 0), ('Mannan endo-1,4-beta-mannosidase', 0), ('Cation efflux system protein', 0), ('putative N-acetylmuramoyl-L-alanine amidase', 0), ('Spermidine/putrescine import ABC transporter ATP-binding protein PotA (TC 3.A.1.11.1)', 0), ('Uncharacterized protein BT4380', 0), ('Pyruvate carboxylase (EC 6.4.1.1) / Biotin carboxyl carrier protein', 0), ('Putative phosphohydrolase, Icc family', 0), ('UDP-glucose 4-epimerase (EC 5.1.3.2)', 0), ('Serine protease', 0), ('Efflux ABC transporter, permease protein', 0), ('Gll3207 protein', 0), ('Putative membrane-associated metal-dependent hydrolase', 0), ('ATP synthase protein I', 0), ('FIG034863: hypothetical protein', 0), ('Indole-3-glycerol phosphate synthase (EC 4.1.1.48)', 0), ('Transposase, IS4', 0), ('Putative alpha-1,6-mannanase', 0), ('Putative helicase', 0), ('UDP-glucose 6-dehydrogenase (EC 1.1.1.22)', 0), ('ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)', 0), ('Thioredoxin', 0), ('DNA base-flipping protein', 0), ('MoxR-like ATPase in aerotolerance operon', 0), ('Gll2767 protein', 0), ('Quinolinate synthetase (EC 2.5.1.72)', 0), ('Methyl-directed repair DNA adenine methylase (EC 2.1.1.72)', 0), ('Peptide deformylase (EC 3.5.1.88)', 0), ('alpha-L-arabinofuranosidase (EC 3.2.1.55)', 0), ('Heat shock protein 10 kDa family chaperone GroES', 0), ('Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)', 0), ('Aldose 1-epimerase (EC 5.1.3.3)', 0), ('Aerotolerance protein BatB', 0), ('Aerotolerance protein BatC', 0), ('Aerotolerance protein BatA', 0), ('Aerotolerance protein BatD', 0), ('Aerotolerance protein BatE', 0), ('Tetrapyrrole methylase family protein', 0), ('LSU ribosomal protein L25p', 0), ('SSU ribosomal protein S4p (S9e) @ SSU ribosomal protein S4p (S9e), zinc-independent', 0), ('ABC transporter, permease protein', 0), ('Polymer-forming bactofilin', 0), ('3,4-dihydroxy-2-butanone 4-phosphate synthase (EC 4.1.99.12) / GTP cyclohydrolase II (EC 3.5.4.25)', 0), ('Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)', 0), ('putative TonB-dependent receptor', 0), ('Endo-1,4-beta-xylanase D', 0), ('putative general stress protein', 0), ('Vitamin B12 ABC transporter, permease protein BtuC', 0), ('probable lipoprotein YPO2292', 0), ('Exonuclease', 0), ('Uncharacterized protein Fjoh_4981, DUF4295-containing', 0), ('Holliday junction ATP-dependent DNA helicase RuvA (EC 3.6.4.12)', 0), ('Sugar-phosphate isomerase', 0), ('Glycosyl transferase, group 1 family protein', 0), ('Sugar-phosphate nucleotidyl transferase', 0), ('Slr1117 protein', 0), ('Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)', 0), ('DedA protein', 0), ('Peptide chain release factor homolog', 0), ('NifU-related domain containing protein', 0), ('LSU ribosomal protein L17p', 0), ('FIG000605: protein co-occurring with transport systems (COG1739)', 0), ('Anaerobic C4-dicarboxylate transporter DcuC', 0), ('Glycosyl transferase, group 2 family', 0), ('L-alanyl-gamma-D-glutamyl-L-diamino acid endopeptidase', 0), ('Na(+)-translocating NADH-quinone reductase subunit E (EC 1.6.5.8)', 0), ('Uncharacterized protein CHU_1311', 0), ('SSU ribosomal protein S20p', 0), ('Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)', 0), ('Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator', 0), ('Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO', 0), ('Zinc-regulated zinc transporting ATPase ZntA', 0), ('Transamidase GatB domain protein', 0)]\n",
      "[('hypothetical protein', 555.1799999999995), ('RNA polymerase ECF-type sigma factor', 11.68), ('Two-component system sensor histidine kinase', 4.079999999999999), ('Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family', 3.88), ('Transcriptional regulator, AraC family', 3.84), ('Thioredoxin', 3.3799999999999994), ('putative membrane protein', 3.1999999999999997), ('Putative anti-sigma factor', 2.78), ('RNA-binding protein', 2.54), ('Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family', 2.5), ('Acidobacterial duplicated orphan permease (function unknown)', 2.4), ('Putative helicase', 1.96), ('cAMP-binding proteins - catabolite gene activator and regulatory subunit of cAMP-dependent protein kinases', 1.6800000000000002), ('Two-component transcriptional response regulator, LuxR family', 1.6400000000000001), ('Integrase', 1.64), ('UpdZ protein', 1.62), ('Uncharacterized MFS-type transporter', 1.5399999999999998), ('Glycosyltransferase', 1.46), ('O-acetylhomoserine sulfhydrylase (EC 2.5.1.49) @ O-succinylhomoserine sulfhydrylase (EC 2.5.1.48)', 1.44), ('Capsular polysaccharide transcription antitermination protein, UpxY family', 1.34), ('N-acetylmuramoyl-L-alanine amidase (EC 3.5.1.28)', 1.2799999999999998), ('Putative glycosyltransferase', 1.2599999999999998), ('ABC-type antimicrobial peptide transport system, ATPase component', 1.24), ('Flavoredoxin', 1.1800000000000002), ('Possible exported heavy-metal binding protein', 1.1400000000000001), ('ISNCY family transposase', 1.1400000000000001), ('Oxidoreductase, aldo/keto reductase family', 1.12), ('UDP-glucose 4-epimerase (EC 5.1.3.2)', 1.1), ('Putative periplasmic protein', 1.0999999999999999), ('Undecaprenyl-phosphate alpha-N-acetylglucosaminyl 1-phosphate transferase (EC 2.7.8.33)', 1.0799999999999998), ('OmpA family outer membrane protein', 1.0600000000000003), ('TPR-repeat-containing protein', 1.06), ('Two-component transcriptional response regulator, LuxR family BT1217', 1.04), ('Monofunctional biosynthetic peptidoglycan transglycosylase', 1.0399999999999998), ('MBL-fold metallo-hydrolase superfamily', 1.02), ('SclB protein', 1.02), ('Membrane protein involved in the export of O-antigen, teichoic acid lipoteichoic acids', 1.02), ('Vitamin B12 ABC transporter, permease protein BtuC', 1.02), ('3-oxoacyl-[acyl-carrier protein] reductase paralog (EC 1.1.1.100)', 1.0199999999999998), ('beta-glycosyl hydrolase', 1.0), ('Zn-ribbon-containing, possibly RNA-binding protein and truncated derivatives', 0.98), ('Conserved protein YuxK', 0.98), ('Cys-tRNA(Pro) deacylase YbaK', 0.9600000000000001), ('Aerotolerance protein BatC', 0.96), ('Endonuclease', 0.94), ('Nitroreductase family protein', 0.9199999999999999), ('Large-conductance mechanosensitive channel', 0.9199999999999999), ('Vitamin B12 ABC transporter, substrate-binding protein BtuF', 0.9199999999999999), ('Efflux ABC transporter, permease protein', 0.9199999999999999), ('Phosphoribosylformimino-5-aminoimidazole carboxamide ribotide isomerase (EC 5.3.1.16)', 0.9199999999999999), ('DoxX family protein', 0.9000000000000001), ('Sigma factor RpoE regulatory protein RseC', 0.8999999999999999), ('UPF0056 inner membrane protein MarC', 0.8999999999999999), ('FIG00490011: hypothetical protein', 0.8800000000000001), ('N-acetylglucosaminyltransferase (EC 2.4.1.-)', 0.88), ('Dihydroorotate dehydrogenase (quinone) (EC 1.3.5.2)', 0.88), ('Transmembrane protein', 0.8600000000000001), ('Malonyl CoA-acyl carrier protein transacylase (EC 2.3.1.39)', 0.8600000000000001), ('Adenosylcobinamide kinase (EC 2.7.1.156) / Adenosylcobinamide-phosphate guanylyltransferase (EC 2.7.7.62)', 0.86), ('Uncharacterized protein BT4380', 0.86), ('putative integration host factor IHF alpha subunit', 0.8400000000000001), ('Rrf2-linked NADH-flavin reductase', 0.8400000000000001), ('Conserved hypothetical protein 22', 0.84), ('2-amino-4-hydroxy-6-hydroxymethyldihydropteridine pyrophosphokinase (EC 2.7.6.3)', 0.84), ('Hypothetical Nudix-like regulator', 0.8399999999999999), ('FIG137478: Hypothetical protein', 0.8399999999999999), ('Probable iron export ATP-binding protein FetA', 0.8399999999999999), ('Phage integrase', 0.8200000000000001), ('N(6)-L-threonylcarbamoyladenine synthase (EC 2.3.1.234)', 0.8200000000000001), ('Putative outer membrane protein, probably involved in nutrient binding', 0.8200000000000001), ('Outer membrane protein romA', 0.8200000000000001), ('Indole-3-glycerol phosphate synthase (EC 4.1.1.48)', 0.8200000000000001), ('Multidrug resistance protein', 0.82), ('RidA/YER057c/UK114 superfamily protein', 0.82), ('GTP cyclohydrolase 1 type 2 homolog YbgI @ Bsu YqfO NIF3/CutA domain', 0.82), ('Cell division protein DivIC (FtsB), stabilizes FtsL against RasP cleavage', 0.82), ('Ribosomal silencing factor RsfA', 0.8), ('Integral membrane protein', 0.8), ('Tyrosine type site-specific recombinase', 0.8), ('Slr1117 protein', 0.8), ('N-acylneuraminate cytidylyltransferase (EC 2.7.7.43)', 0.7999999999999999), ('cytochrome c biogenesis protein', 0.7999999999999999), ('[FeFe]-hydrogenase maturation protein HydF', 0.7999999999999999), ('Uncharacterized membrane protein PA1577', 0.7999999999999999), ('D-alanine--poly(phosphoribitol) ligase ACP subunit (EC 6.1.1.13)', 0.7999999999999998), ('UPF0126 inner membrane protein BT_1201', 0.78), ('Oxidoreductase', 0.78), ('Thiol:disulfide interchange protein', 0.78), ('Ribosome-associated heat shock protein implicated in the recycling of the 50S subunit (S4 paralog)', 0.76), ('4-diphosphocytidyl-2-C-methyl-D-erythritol kinase (EC 2.7.1.148)', 0.76), ('FIG000605: protein co-occurring with transport systems (COG1739)', 0.76), ('Fosmidomycin resistance protein', 0.7599999999999999), ('Membrane protein', 0.7599999999999999), ('CzcABC family efflux RND transporter, outer membrane protein', 0.7599999999999998), ('Permease of the drug/metabolite transporter (DMT) superfamily', 0.7400000000000001), ('Dihydroorotate dehydrogenase (NAD(+)), electron transfer subunit (EC 1.3.1.14)', 0.74), ('tmRNA-binding protein SmpB', 0.74), ('DNA-directed RNA polymerase omega subunit (EC 2.7.7.6)', 0.74), ('Free methionine-(R)-sulfoxide reductase, contains GAF domain', 0.74), ('Aerotolerance protein BatB', 0.74), ('Na+/H+-dicarboxylate symporter', 0.7399999999999999), ('Transcriptional regulator', 0.7200000000000001), ('Electron transport complex protein RnfA', 0.7200000000000001), ('Transcriptional regulator, AcrR family', 0.7200000000000001), ('V-type ATP synthase subunit K (EC 3.6.3.14)', 0.7200000000000001), ('Possible glyoxylase family protein (Lactoylglutathione lyase) (EC 4.4.1.5)', 0.72), ('Glycosyl transferase, family 2', 0.72), ('putative patatin-like phospholipase', 0.72), ('Tll0814 protein', 0.72), ('ATP synthase protein I', 0.72), ('Aerotolerance protein BatE', 0.72), ('Serine protease', 0.72), ('TonB family protein', 0.7199999999999999), ('Aminodeoxychorismate lyase (EC 4.1.3.38)', 0.7000000000000001), ('Co-activator of prophage gene expression IbrB', 0.7), ('Putative permease', 0.7), ('Transcriptional regulator, AsnC family', 0.7), ('Putative thioredoxin', 0.7), ('Lipoprotein signal peptidase (EC 3.4.23.36)', 0.6999999999999998), ('Glycosyltransferase involved in cell wall biogenesis', 0.68), ('LSU ribosomal protein L34p', 0.68), ('6-phosphogluconolactonase (EC 3.1.1.31), eukaryotic type', 0.68), ('Helicase, C-terminal:Type III restriction enzyme, res subunit:DEAD/DEAH box helicase, N-terminal', 0.68), ('Uncharacterized protein YgbA', 0.68), ('Sugar-phosphate nucleotidyl transferase', 0.68), ('ATP synthase beta chain (EC 3.6.3.14)', 0.68), ('Probable integral membrane protein Cj1165c', 0.6799999999999999), ('CAAX amino terminal protease family protein', 0.6799999999999999), ('Uronate isomerase (EC 5.3.1.12)', 0.6799999999999999), ('ATP-dependent RNA helicase', 0.6799999999999999), ('Aerotolerance protein BatD', 0.6799999999999999), ('Putative DNA-binding protein', 0.6600000000000001), ('DNA-binding protein HU-beta', 0.66), ('Trk potassium uptake system protein TrkH', 0.66), ('Chromate transport protein', 0.66), ('Hemin receptor', 0.66), ('Putative phage-encoded peptidoglycan binding protein', 0.66), ('Cytidylyltransferase domain / Phosphoenolpyruvate phosphomutase (EC 5.4.2.9)', 0.66), ('Immunoreactive 53 kDa antigen PG123', 0.6599999999999999), ('4-hydroxy-2-oxoglutarate aldolase (EC 4.1.3.16) @ 2-dehydro-3-deoxyphosphogluconate aldolase (EC 4.1.2.14)', 0.6599999999999999), ('Dihydroorotate dehydrogenase (NAD(+)), catalytic subunit (EC 1.3.1.14)', 0.6599999999999999), ('DNA polymerase III delta subunit (EC 2.7.7.7)', 0.6599999999999999), ('tRNA (guanine(37)-N(1))-methyltransferase (EC 2.1.1.228)', 0.64), ('Protein DUF2149, predicted transporter component', 0.64), ('Beta-lactamase class C-like and penicillin binding proteins (PBPs) superfamily', 0.64), ('FIG00491264: hypothetical protein', 0.64), ('D-glycero-alpha-D-manno-heptose 7-phosphate kinase (EC 2.7.1.168)', 0.64), ('Amino acid-binding ACT', 0.64), ('ATP-dependent DNA helicase UvrD/PcrA (EC 3.6.4.12)', 0.64), ('Xylanase', 0.6399999999999999), ('ATP synthase gamma chain (EC 3.6.3.14)', 0.6399999999999999), ('Small heat shock protein', 0.6399999999999999), ('dTDP-4-dehydrorhamnose 3,5-epimerase (EC 5.1.3.13)', 0.6399999999999999), ('Stress responsive alpha-beta barrel domain protein Dabb', 0.6399999999999999), ('sodium/iodide co-transporter', 0.6200000000000001), ('Chaperone protein HtpG', 0.6200000000000001), ('ATP synthase F0 sector subunit a (EC 3.6.3.14)', 0.6200000000000001), ('Translation initiation factor 3', 0.62), ('Cytochrome d ubiquinol oxidase subunit II (EC 1.10.3.-)', 0.62), ('Aminotransferase, DegT/DnrJ/EryC1/StrS family', 0.62), ('AMP nucleosidase (EC 3.2.2.4)', 0.62), ('L-alanine-DL-glutamate epimerase (EC 5.1.1.n1)', 0.62), ('Efflux transport system, outer membrane factor (OMF) lipoprotein', 0.62), ('Diacylglycerol kinase (EC 2.7.1.107)', 0.6199999999999999), ('FIG016027: protein of unknown function YeaO', 0.6199999999999999), ('Phosphate/pyrophosphate-specific outer membrane porin OprP/OprO', 0.6199999999999999), ('Na+-driven multidrug efflux pump', 0.6), ('Long-chain-fatty-acid--CoA ligase (EC 6.2.1.3)', 0.6), ('SSU ribosomal protein S20p', 0.6), ('Electron transport complex protein RnfB', 0.6), ('Integrase IntN1', 0.6), ('DnaK suppressor protein, putative', 0.6), ('Recombination protein RecR', 0.6), ('3-deoxy-manno-octulosonate cytidylyltransferase (EC 2.7.7.38)', 0.6), ('Sirohydrochlorin cobaltochelatase CbiK (EC 4.99.1.3) @ Sirohydrochlorin ferrochelatase activity of CbiK (EC 4.99.1.4)', 0.6), ('Aerotolerance protein BatA', 0.6), ('Exodeoxyribonuclease VII small subunit (EC 3.1.11.6)', 0.5800000000000001), ('DNA ligase (NAD(+)) (EC 6.5.1.2)', 0.5800000000000001), ('Thiamin-phosphate pyrophosphorylase (EC 2.5.1.3)', 0.5800000000000001), ('Cobalt-precorrin-7 (C5)-methyltransferase (EC 2.1.1.289) / Cobalt-precorrin-6B C15-methyltransferase [decarboxylating] (EC 2.1.1.196)', 0.58), ('PAP2 superfamily protein', 0.58), ('4-keto-6-deoxy-N-Acetyl-D-hexosaminyl-(Lipid carrier) aminotransferase', 0.58), ('FIG00936810: hypothetical protein', 0.58), ('Rubredoxin', 0.58), ('Integration host factor alpha subunit', 0.58), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17) @ ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17), cobalamin synthesis', 0.58), ('Adenosylcobinamide-phosphate synthase (EC 6.3.1.10)', 0.56), ('Proline dehydrogenase (EC 1.5.5.2) / Delta-1-pyrroline-5-carboxylate dehydrogenase (EC 1.2.1.88)', 0.56), ('Pentapeptide repeat family protein', 0.56), ('Uncharacterized protein CHU_1311', 0.56), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, delta subunit, ferredoxin-like 4Fe-4S binding protein (EC 1.2.7.-)', 0.5599999999999999), ('putative UndPP-QuiNAc-P-transferase', 0.5599999999999999), ('UDP-glucuronate decarboxylase (EC 4.1.1.35)', 0.5599999999999999), ('RNA methyltransferase, TrmH family', 0.5599999999999999), ('dTDP-4-dehydrorhamnose reductase (EC 1.1.1.133)', 0.5599999999999999), ('Ribosylnicotinamide kinase (EC 2.7.1.22) homolog / Unknown conserved in Flavobacteria', 0.5599999999999999), ('Transcription termination protein NusB', 0.5599999999999999), ('Para-aminobenzoate synthase, aminase component (EC 2.6.1.85)', 0.5599999999999999), ('Cation efflux system protein', 0.5599999999999999), ('ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)', 0.5599999999999999), ('Inner membrane protein YhiM', 0.54), ('Biotin carboxyl carrier protein => methylmalonyl-CoA decarboxylase', 0.54), ('Carboxyl-terminal protease', 0.54), ('probable lipoprotein YPO2292', 0.54), ('Murein endolytic transglycosylase MltG', 0.54), ('Imidazole glycerol phosphate synthase amidotransferase subunit HisH', 0.5399999999999999), ('Crossover junction endodeoxyribonuclease RuvC (EC 3.1.22.4)', 0.5399999999999999), ('Phosphonopyruvate decarboxylase (EC 4.1.1.82)', 0.5399999999999999), ('Thiol peroxidase, Bcp-type (EC 1.11.1.15)', 0.5399999999999999), ('Putative pre-16S rRNA nuclease YqgF', 0.52), ('Uracil permease @ Uracil:proton symporter UraA', 0.52), ('Transcription accessory protein (S1 RNA-binding domain)', 0.52), ('Arylsulfatase (EC 3.1.6.1)', 0.52), ('ROK family sugar kinase or transcriptional regulator', 0.52), ('Acetyltransferase, GNAT family', 0.52), ('Predicted dye-decolorizing peroxidase (DyP), encapsulated subgroup', 0.52), ('LysM-repeat proteins and domains', 0.52), ('Isoleucyl-tRNA synthetase (EC 6.1.1.5)', 0.5199999999999999), ('Glyoxalase family protein', 0.5199999999999999), ('UPF0324 inner membrane protein YeiH', 0.5199999999999999), ('Oligopeptide transporter, OPT family', 0.5199999999999999), ('Transcriptional regulator, PadR family', 0.5199999999999999), ('Glycosyl transferase, group 2 family', 0.5199999999999999), (\"Inosine-5'-monophosphate dehydrogenase (EC 1.1.1.205) / CBS domain\", 0.5000000000000001), ('tRNA(1)(Val) (adenine(37)-N(6))-methyltransferase (EC 2.1.1.223)', 0.5), ('Adenylylsulfate kinase (EC 2.7.1.25)', 0.5), ('putative N-acetylmuramoyl-L-alanine amidase', 0.5), ('Endo-1,4-beta-xylanase D', 0.5), (\"Alpha-ribazole-5'-phosphate phosphatase (EC 3.1.3.73)\", 0.49999999999999994), ('tRNA (5-methylaminomethyl-2-thiouridylate)-methyltransferase (EC 2.1.1.61)', 0.49999999999999994), ('putative pyrogenic exotoxin B', 0.49999999999999994), ('ATP synthase alpha chain (EC 3.6.3.14)', 0.4999999999999999), ('Ribose-phosphate pyrophosphokinase (EC 2.7.6.1)', 0.4999999999999999), ('Alkaline phosphatase (EC 3.1.3.1)', 0.48000000000000004), ('Cobalt-precorrin-2 C(20)-methyltransferase (EC 2.1.1.151)', 0.48), ('5-amino-6-(5-phosphoribosylamino)uracil reductase', 0.48), ('SAM-dependent methyltransferase', 0.48), ('RND efflux system, inner membrane transporter', 0.48), ('Outer membrane assembly lipoprotein YfiO', 0.48), ('LemA protein', 0.48), ('Putative membrane-associated metal-dependent hydrolase', 0.48), ('Putative alpha-1,6-mannanase', 0.48), ('Na(+)-translocating NADH-quinone reductase subunit B (EC 1.6.5.8)', 0.48), (\"5'-nucleotidase (EC 3.1.3.5); 2',3'-cyclic-nucleotide 2'-phosphodiesterase (EC 3.1.4.16); Putative UDP-sugar hydrolase (EC 3.6.1.45)\", 0.4799999999999999), ('6-phosphogluconate dehydrogenase, decarboxylating (EC 1.1.1.44)', 0.4799999999999999), ('Agmatine deiminase (EC 3.5.3.12)', 0.4799999999999999), ('SSU ribosomal protein S14p (S29e) @ SSU ribosomal protein S14p (S29e), zinc-independent', 0.4799999999999999), ('FIG01416087: hypothetical protein', 0.46000000000000013), ('DnaJ-class molecular chaperone CbpA', 0.46000000000000013), ('Undecaprenyl-phosphate galactosephosphotransferase (EC 2.7.8.6)', 0.46), ('FHA domain protein', 0.45999999999999996), ('Aspartate carbamoyltransferase regulatory chain (PyrI)', 0.45999999999999996), ('Ribose-5-phosphate isomerase B (EC 5.3.1.6)', 0.45999999999999996), ('Pyridoxal phosphate-containing protein YggS', 0.45999999999999996), ('Clostripain-related protein', 0.45999999999999996), ('Xylulose kinase (EC 2.7.1.17)', 0.45999999999999996), ('FIG00003370: Multicopper polyphenol oxidase', 0.45999999999999996), ('Superoxide dismutase [Mn/Fe] (EC 1.15.1.1)', 0.45999999999999996), ('V-type ATP synthase subunit E (EC 3.6.3.14)', 0.45999999999999996), ('Transglycosylase-associated protein', 0.45999999999999996), ('Putative calcineurin superfamily phosphohydrolase', 0.45999999999999996), ('mobilization protein B', 0.45999999999999996), ('SSU ribosomal protein S6p', 0.4400000000000001), ('ATP synthase F0 sector subunit c (EC 3.6.3.14)', 0.44), ('Ribosomal protein S12p Asp88 (E. coli) methylthiotransferase (EC 2.8.4.4)', 0.44), ('Ser/Thr protein phosphatase family protein, UDP-2,3-diacylglucosamine hydrolase (EC 3.6.1.-) homolog', 0.43999999999999995), ('low molecular weight phosphatase supefamily', 0.43999999999999995), ('SanA protein', 0.43999999999999995), ('Phosphate ABC transporter, substrate-binding protein PstS (TC 3.A.1.7.1)', 0.43999999999999995), ('Hcp transcriptional regulator HcpR (Crp/Fnr family)', 0.43999999999999995), ('Integrase protein', 0.43999999999999995), ('Two-component system sensor histidine kinase/response regulator hybrid', 0.42), ('Translation initiation factor SUI1-related protein', 0.41999999999999993), ('putative lipoprotein', 0.41999999999999993), ('Regulatory sensor-transducer, BlaR1/MecR1 family / TonB-dependent receptor', 0.41999999999999993), ('4-hydroxy-tetrahydrodipicolinate synthase (EC 4.3.3.7)', 0.41999999999999993), ('Signal recognition particle receptor FtsY', 0.41999999999999993), ('LSU ribosomal protein L33p @ LSU ribosomal protein L33p, zinc-independent', 0.41999999999999993), ('UspA', 0.41999999999999993), ('Dethiobiotin synthase BioD (EC 6.3.3.3)', 0.41999999999999993), ('Methylmalonyl-CoA epimerase (EC 5.1.99.1)', 0.41999999999999993), ('OmpA', 0.41999999999999993), ('archaeal ATPase, fused to C-terminal DUF234 domain', 0.41999999999999993), ('Transposase', 0.41999999999999993), ('Succinyl-CoA ligase [ADP-forming] beta chain (EC 6.2.1.5)', 0.4000000000000001), ('Capsular polysaccharide transcription antitermination protein UpeY', 0.4000000000000001), ('ATP synthase epsilon chain (EC 3.6.3.14)', 0.39999999999999997), ('Ync', 0.39999999999999997), ('Putative dipeptidase', 0.39999999999999997), ('Aminobenzoyl-glutamate transport protein', 0.39999999999999997), ('Inner membrane protein YihY, formerly thought to be RNase BN', 0.39999999999999997), ('FIG00414141: hypothetical protein', 0.39999999999999997), ('Uracil-DNA glycosylase, family 1 (EC 3.2.2.27)', 0.39999999999999997), ('Lipopolysaccharide export system permease protein LptG', 0.39999999999999997), ('Tryptophan synthase (indole-salvaging) (EC 4.2.1.122)', 0.39999999999999997), ('AAA+ superfamily protein', 0.38000000000000006), ('Arylsulfatase regulator', 0.38000000000000006), ('Electron transport complex protein RnfD', 0.38000000000000006), ('putative xylanase', 0.38000000000000006), ('Methionine transporter MetT', 0.38000000000000006), ('Acyltransferase', 0.38000000000000006), ('Glucosamine-1-phosphate guanylyltransferase', 0.38000000000000006), ('Conserved protein, with a weak D-galactarate dehydratase/altronate hydrolase domain', 0.38), ('Fluoride ion transporter CrcB', 0.38), ('Serine/threonine kinase', 0.38), ('ATPase RavA', 0.38), ('Sulfur carrier protein ThiS adenylyltransferase (EC 2.7.7.73)', 0.38), ('Adenosine deaminase (EC 3.5.4.4)', 0.37999999999999995), ('Bacterial ribosome SSU maturation protein RimP', 0.36000000000000004), ('Prolyl oligopeptidase family protein', 0.36000000000000004), ('Cystathionine beta-lyase (EC 4.4.1.8)', 0.36000000000000004), ('Queuine tRNA-ribosyltransferase (EC 2.4.2.29)', 0.36000000000000004), ('Oxaloacetate decarboxylase Na(+) pump, gamma chain (EC 4.1.1.3)', 0.36000000000000004), ('Endonuclease III (EC 4.2.99.18)', 0.36000000000000004), ('Tyrosine-protein kinase EpsD (EC 2.7.10.2)', 0.36000000000000004), ('Two-component transcriptional response regulator, OmpR family', 0.36000000000000004), ('Ribosomal-protein-S5p-alanine acetyltransferase (EC 2.3.1.128)', 0.36000000000000004), ('Lipoate-protein ligase A', 0.36000000000000004), ('Gll2767 protein', 0.36000000000000004), ('LSU ribosomal protein L28p @ LSU ribosomal protein L28p, zinc-independent', 0.36), ('Polysaccharide deacetylase-like protein', 0.36), ('FIG138056: a glutathione-dependent thiol reductase', 0.36), ('Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4)', 0.36), ('DNA base-flipping protein', 0.36), ('putative phosphatase/phosphohexomutase', 0.34), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, gamma subunit (EC 1.2.7.-)', 0.34), ('DNA-binding protein', 0.34), ('Helicase', 0.34), ('Conjugative transposon protein TraM', 0.34), ('Tn5520-like integrase', 0.34), ('beta-glucosidase (EC 3.2.1.21)', 0.34), ('Ribosome hibernation promoting factor Hpf', 0.34), ('Hypothetical radical SAM family enzyme in interesting gene cluster', 0.34), ('putative acetyltransferase', 0.33999999999999997), ('Glycosylase BF2339', 0.33999999999999997), ('Pantoate--beta-alanine ligase (EC 6.3.2.1)', 0.32), ('Anthranilate phosphoribosyltransferase (EC 2.4.2.18)', 0.32), ('FKBP-type peptidyl-prolyl cis-trans isomerase FkpA precursor (EC 5.2.1.8)', 0.32), ('Capsular polysaccharide biosynthesis protein capD', 0.32), ('Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps', 0.32), ('ATP-dependent Clp protease proteolytic subunit ClpP (EC 3.4.21.92)', 0.32), ('Transcriptional regulator of fucose utilization, GntR family', 0.32), ('N-acetylmannosaminyltransferase (EC 2.4.1.187)', 0.32), ('Flavodoxin', 0.32), ('Pyruvate:Oxaloacetate transcarboxylase domain protein / Biotin carboxyl carrier protein', 0.32), ('N-carbamoylputrescine amidase (EC 3.5.1.53)', 0.32), ('Type II restriction enzyme HpaII (EC 3.1.21.4)', 0.32), ('CRISPR-associated protein, TM1814 family', 0.32), ('Type I restriction-modification system, specificity subunit S', 0.32), ('Phosphoribosylanthranilate isomerase (EC 5.3.1.24)', 0.32), ('Rod shape-determining protein MreD', 0.32), ('CRISPR-associated protein Cas1', 0.32), ('Diaminopimelate epimerase (EC 5.1.1.7)', 0.32), ('Lactaldehyde reductase (EC 1.1.1.77)', 0.32), ('L,L-diaminopimelate aminotransferase (EC 2.6.1.83)', 0.32), ('Ferredoxin domain containing protein', 0.32), ('Sugar-phosphate isomerase', 0.32), ('Putative preQ0 transporter YhhQ', 0.32), ('Lyzozyme M1 (1,4-beta-N-acetylmuramidase) (EC 3.2.1.17)', 0.32), ('tRNA-Arg-CCT', 0.32), ('Protein BF2544, predicted transporter component', 0.31999999999999995), ('Ser-tRNA(Ala) deacylase @ Gly-tRNA(Ala) deacylase', 0.3), ('Polysaccharide deacetylase', 0.3), ('TPR repeat precursor', 0.3), ('Transcriptional repressor, BlaI/MecI family', 0.3), ('Lipopolysaccharide cholinephosphotransferase LicD1 (EC 2.7.8.-)', 0.3), ('Peptidase, S41 family', 0.3), ('Intracellular protease', 0.3), ('UPF0145 protein YbjQ', 0.3), ('Heat shock protein HtpX', 0.3), ('Protein translocase membrane subunit SecG', 0.3), ('FIG015287: Zinc protease', 0.3), ('Uncharacterized protein TTC1240', 0.3), ('Alpha-N-acetylglucosaminidase (EC 3.2.1.50)', 0.3), ('Nitrogen regulation protein NtrY (EC 2.7.3.-)', 0.3), ('TPR domain protein', 0.3), ('Putative Na+/H+ exchange protein', 0.3), ('6-carboxy-5,6,7,8-tetrahydropterin synthase (EC 4.1.2.50)', 0.3), ('Transcriptional regulator, MarR family', 0.3), ('galactosamine-containing minor teichoic acid biosynthesis protein', 0.3), ('Alpha-1,4-N-acetylgalactosamine transferase PglH (EC 2.4.1.-)', 0.3), ('Iron-sulfur flavoprotein', 0.3), ('Hydroxylamine reductase (EC 1.7.99.1)', 0.3), ('Serine hydroxymethyltransferase (EC 2.1.2.1)', 0.3), ('Inner membrane protein RclC associated with response to reactive chlorine species', 0.3), ('Transcription regulator [contains diacylglycerol kinase catalytic domain]', 0.28), ('Peroxide stress regulator; Ferric uptake regulation protein; Fe2+/Zn2+ uptake regulation proteins', 0.28), ('Metal-dependent hydrolase YbeY, involved in rRNA and/or ribosome maturation and assembly', 0.27999999999999997), ('Carboxy-terminal processing protease', 0.27999999999999997), ('GTP-binding protein EngB', 0.27999999999999997), ('UPF0246 protein YaaA', 0.27999999999999997), ('alpha-L-fucosidase (EC 3.2.1.51)', 0.27999999999999997), ('PKD domain-containing protein', 0.27999999999999997), ('Guanine deaminase (EC 3.5.4.3)', 0.27999999999999997), ('Uncharacterized protein BT_4646', 0.27999999999999997), ('ADP-ribose pyrophosphatase of COG1058 family (EC 3.6.1.13) / Nicotinamide-nucleotide amidase (EC 3.5.1.42)', 0.27999999999999997), ('V-type ATP synthase subunit D (EC 3.6.3.14)', 0.27999999999999997), ('Cell surface protein', 0.27999999999999997), ('Huntingtin interacting protein E-like protein', 0.27999999999999997), ('Xanthine phosphoribosyltransferase (EC 2.4.2.22)', 0.27999999999999997), ('Glycosyl transferase, group 1', 0.27999999999999997), ('Putative phosphatidylinositol-4-phosphate 5-kinase', 0.27999999999999997), ('D-xylose proton-symporter XylE', 0.27999999999999997), ('tRNA dimethylallyltransferase (EC 2.5.1.75)', 0.27999999999999997), ('N-acetylneuraminate synthase (EC 2.5.1.56)', 0.27999999999999997), ('Hydroxymethylpyrimidine phosphate kinase ThiD (EC 2.7.4.7)', 0.27999999999999997), ('Gll3207 protein', 0.27999999999999997), ('putative DNA mismatch repair protein', 0.27999999999999997), ('ATP-dependent RNA helicase BT1154', 0.27999999999999997), ('Mannose-6-phosphate isomerase (EC 5.3.1.8)', 0.27999999999999997), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, beta subunit (EC 1.2.7.-)', 0.27999999999999997), ('N-carbamoylputrescine amidase (EC 3.5.1.53) @ Aliphatic amidase AmiE (EC 3.5.1.4)', 0.27999999999999997), ('Glutamate formiminotransferase (EC 2.1.2.5) @ Glutamate formyltransferase', 0.25999999999999995), ('UPF0301 protein YqgE', 0.25999999999999995), ('2-amino-3-ketobutyrate coenzyme A ligase (EC 2.3.1.29)', 0.25999999999999995), ('Bacterial non-heme ferritin (EC 1.16.3.2)', 0.25999999999999995), ('T6SS PAAR-repeat protein / RhaS protein', 0.25999999999999995), ('2-iminoacetate synthase (ThiH) (EC 4.1.99.19)', 0.25999999999999995), ('Arabinan endo-1,5-alpha-L-arabinosidase A', 0.25999999999999995), ('UDP-galactose-lipid carrier transferase (EC 2.-.-.-)', 0.25999999999999995), ('Transcriptional regulator, LuxR family', 0.25999999999999995), ('L-fucose kinase (EC 2.7.1.52)', 0.25999999999999995), ('Serine palmitoyltransferase (EC 2.3.1.50)', 0.25999999999999995), ('1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)', 0.25999999999999995), ('Acetyltransferase', 0.25999999999999995), ('Superfamily I DNA/RNA helicase protein', 0.25999999999999995), ('Transporter, LysE family', 0.25999999999999995), ('Diadenylate cyclase spyDAC; Bacterial checkpoint controller DisA with nucleotide-binding domain', 0.25999999999999995), ('DNA/RNA helicases, SNF2 family', 0.25999999999999995), ('Imidazole glycerol phosphate synthase cyclase subunit', 0.25999999999999995), ('Putative lipopolysaccharide biosynthesis protein', 0.25999999999999995), ('Xylose isomerase (EC 5.3.1.5)', 0.25999999999999995), ('Indolepyruvate oxidoreductase subunit IorB (EC 1.2.7.8)', 0.25999999999999995), ('N-acetylneuraminate lyase (EC 4.1.3.3)', 0.25999999999999995), ('DNA protection during starvation protein', 0.25999999999999995), ('Glycerophosphoryl diester phosphodiesterase (EC 3.1.4.46)', 0.25999999999999995), ('DnaD domain protein', 0.25999999999999995), ('putative protease I', 0.25999999999999995), ('Putative phosphohydrolase, Icc family', 0.25999999999999995), ('ATP phosphoribosyltransferase (EC 2.4.2.17) => HisGl', 0.25999999999999995), ('LSU ribosomal protein L29p (L35e)', 0.25999999999999995), ('Peptide deformylase (EC 3.5.1.88)', 0.25999999999999995), ('putative TonB-dependent receptor', 0.25999999999999995), ('GldB', 0.25999999999999995), ('CRISPR-associated helicase Cas3', 0.25999999999999995), ('Acetoin dehydrogenase E1 component alpha-subunit (EC 2.3.1.190) / Acetoin dehydrogenase E1 component beta-subunit (EC 2.3.1.190)', 0.25999999999999995), ('MutT/Nudix family protein', 0.25999999999999995), ('putative capsule polysaccharide export protein', 0.23999999999999996), ('Riboflavin synthase eubacterial/eukaryotic (EC 2.5.1.9)', 0.23999999999999996), ('Cobyric acid synthase (EC 6.3.5.10)', 0.23999999999999996), ('Error-prone, lesion bypass DNA polymerase V (UmuC)', 0.23999999999999996), ('Acetyl esterase', 0.23999999999999996), ('Predicted transcriptional regulator', 0.23999999999999996), ('Alpha-glucosidase (EC 3.2.1.20)', 0.23999999999999996), ('ABC export system, membrane fusion protein BT1213', 0.23999999999999996), ('Exopolyphosphatase (EC 3.6.1.11)', 0.23999999999999996), ('4-carboxymuconolactone decarboxylase (EC 4.1.1.44)', 0.23999999999999996), ('Protein translocase subunit YajC', 0.23999999999999996), ('V-type ATP synthase subunit C (EC 3.6.3.14)', 0.23999999999999996), ('NADH-dependent glycerate dehydrogenase', 0.23999999999999996), ('Oxidoreductase, short-chain dehydrogenase/reductase family', 0.23999999999999996), ('Thiol:disulfide interchange protein tlpA', 0.23999999999999996), ('Propionyl-CoA carboxylase biotin-containing subunit (EC 6.4.1.3)', 0.23999999999999996), ('NADPH-dependent 7-cyano-7-deazaguanine reductase (EC 1.7.1.13)', 0.23999999999999996), ('TPR repeat:TPR repeat precursor', 0.23999999999999996), ('Cobalamin synthase (EC 2.7.8.26)', 0.23999999999999996), ('Cytochrome c551 peroxidase (EC 1.11.1.5)', 0.23999999999999996), ('Dephospho-CoA kinase (EC 2.7.1.24)', 0.23999999999999996), ('Ribonuclease G', 0.23999999999999996), ('Possible restriction /modification enzyme', 0.23999999999999996), ('Ribonucleotide reductase of class Ia (aerobic), beta subunit (EC 1.17.4.1)', 0.23999999999999996), ('Quinolinate synthetase (EC 2.5.1.72)', 0.23999999999999996), ('Aspartate carbamoyltransferase (EC 2.1.3.2)', 0.23999999999999996), ('FAD:protein FMN transferase (EC 2.7.1.180)', 0.23999999999999996), ('Formate-dependent phosphoribosylglycinamide formyltransferase', 0.22000000000000003), ('Alpha-1,2-fucosyltransferase', 0.21999999999999997), ('Voltage-gated H(+)/2Cl(-) exchange transporter ClcA', 0.21999999999999997), ('Putative manganese efflux pump MntP', 0.21999999999999997), ('hypothetical protein; Some similarities with the C-terminal region of hemagglutinin/hemolysin-related protein and with purine NTPase', 0.21999999999999997), ('2-methylfumaryl-CoA hydratase (EC 4.2.1.148)', 0.21999999999999997), ('Oxaloacetate decarboxylase Na(+) pump, alpha chain (EC 4.1.1.3)', 0.21999999999999997), ('Phage protein', 0.21999999999999997), ('Superfamily I DNA and RNA helicases and helicase subunits', 0.21999999999999997), ('Lead, cadmium, zinc and mercury transporting ATPase (EC 3.6.3.3) (EC 3.6.3.5); Copper-translocating P-type ATPase (EC 3.6.3.4)', 0.21999999999999997), ('Beta-1,3-glucosyltransferase', 0.21999999999999997), ('AmpG protein', 0.21999999999999997), ('Capsular polysaccharide transcription antitermination protein UphY', 0.21999999999999997), ('Putative secreted polysaccharide polymerase', 0.21999999999999997), ('NADH-ubiquinone oxidoreductase chain K (EC 1.6.5.3)', 0.21999999999999997), ('Tryptophan synthase beta chain (EC 4.2.1.20)', 0.21999999999999997), ('Keratan-sulfate endo-1,4-beta-galactosidase (EC 3.2.1.103)', 0.21999999999999997), ('CRISPR-associated negative autoregulator Cas7/Cst2', 0.21999999999999997), ('Fructose-1,6-bisphosphatase, Bacillus type (EC 3.1.3.11)', 0.21999999999999997), ('putative oxidoreductase, nitronate monooxygenase family', 0.21999999999999997), ('TonB-dependent receptor', 0.21999999999999997), ('Inorganic pyrophospatase PpaX', 0.21999999999999997), ('Mannan endo-1,4-beta-mannosidase', 0.21999999999999997), ('3-isopropylmalate dehydratase small subunit (EC 4.2.1.33)', 0.21999999999999997), ('Probable iron export permease protein FetB', 0.21999999999999997), ('Flavodoxin 1', 0.21999999999999997), ('Peptide chain release factor homolog', 0.21999999999999997), ('Pirin', 0.21999999999999997), ('Twin-arginine translocation protein TatA', 0.21999999999999997), ('Similar to conjugative transposon protein TraB', 0.21999999999999997), ('Phosphomethylpyrimidine synthase ThiC (EC 4.1.99.17)', 0.19999999999999998), ('hypothetical domain / Cold shock protein of CSP family', 0.19999999999999998), ('Sensor histidine kinase involved in nitrogen fixation and metabolism regulation', 0.19999999999999998), ('Phosphoribosylaminoimidazolecarboxamide formyltransferase (EC 2.1.2.3)', 0.19999999999999998), ('Probable integral membrane protein Cj1166c', 0.19999999999999998), ('Formate--tetrahydrofolate ligase (EC 6.3.4.3)', 0.19999999999999998), ('Energy-dependent translational throttle protein EttA', 0.19999999999999998), (\"16S rRNA (cytidine(1402)-2'-O)-methyltransferase (EC 2.1.1.198)\", 0.19999999999999998), ('LemA family protein', 0.19999999999999998), ('Na(+)-translocating NADH-quinone reductase subunit E (EC 1.6.5.8)', 0.19999999999999998), ('Related to F420H2-dehydrogenase, beta subunit', 0.19999999999999998), ('Serine protease, DegP/HtrA, do-like (EC 3.4.21.-)', 0.19999999999999998), ('Zinc ABC transporter, ATP-binding protein ZnuC', 0.19999999999999998), ('Sulfate adenylyltransferase subunit 1 (EC 2.7.7.4)', 0.19999999999999998), ('Protein of unknown function UPF0060', 0.19999999999999998), ('Putative molybdenum transport ATP-binding protein modF', 0.19999999999999998), ('Nicotinate-nucleotide--dimethylbenzimidazole phosphoribosyltransferase (EC 2.4.2.21)', 0.19999999999999998), ('putative protein involved in capsular polysaccharide biosynthesis', 0.19999999999999998), ('ATP-dependent DNA helicase', 0.19999999999999998), ('Carbon starvation protein A', 0.19999999999999998), ('Iron-containing alcohol dehydrogenase', 0.19999999999999998), ('Transcriptional regulator YeiE, LysR family', 0.19999999999999998), ('Mannose-1-phosphate guanylyltransferase (EC 2.7.7.13)', 0.19999999999999998), ('Outer membrane efflux protein precursor', 0.19999999999999998), ('CRISPR-associated RecB family exonuclease Cas4', 0.19999999999999998), ('Glycosyl transferase, group 1 family protein', 0.19999999999999998), ('Thiol peroxidase, Tpx-type (EC 1.11.1.15)', 0.19999999999999998), ('Type I restriction-modification system, DNA-methyltransferase subunit M (EC 2.1.1.72)', 0.18), ('ABC-type antimicrobial peptide transport system, ATPase component BT1214', 0.18), ('Heat shock protein GrpE', 0.18), ('hypothetical transporter PduT for various metalloporphyrins', 0.18), ('RNA pseudouridylate synthase BT0642', 0.18), ('Sodium/proline symporter', 0.18), ('Thioredoxin domain-containing protein BT1583', 0.18), ('Quinolinate phosphoribosyltransferase [decarboxylating] (EC 2.4.2.19)', 0.18), ('Integration host factor beta subunit', 0.18), ('putative TonB-dependent outer membrane protein', 0.18), ('Septum formation protein Maf', 0.18), ('Glucose-1-phosphate thymidylyltransferase (EC 2.7.7.24)', 0.18), ('Polysaccharide export outer membrane protein', 0.18), ('ATP synthase F0 sector subunit b (EC 3.6.3.14)', 0.18), ('Transmembrane component YkoC of energizing module of thiamin-regulated ECF transporter for HydroxyMethylPyrimidine', 0.18), ('DNA helicase (EC 3.6.4.12), mobile-element-associated', 0.18), ('LSU ribosomal protein L24p (L26e)', 0.18), ('beta-galactosidase (EC 3.2.1.23)', 0.18), ('Putative sodium-dependent transporter', 0.18), ('Chaperone protein DnaJ', 0.18), ('Efflux transport system, outer membrane factor (OMF) lipoprotein BT1212', 0.18), ('Phage terminase, large subunit', 0.18), ('Hemophore HmuY, recruits heme from host protein carriers', 0.18), ('Tyrosine-protein kinase (EC 2.7.10.2)', 0.18), ('Uncharacterized integral membrane, glycosyltransferase', 0.18), ('Pyruvate formate-lyase activating enzyme (EC 1.97.1.4)', 0.18), ('Capsular polysaccharide transcription antitermination protein UpfY', 0.18), ('Predicted cobalt transporter in Bacteroides_Porphyromonas', 0.18), ('N-acetylgalactosamine-6-sulfatase', 0.18), ('Lipopolysaccharide biosynthesis protein', 0.16), ('2-oxoglutarate/2-oxoacid ferredoxin oxidoreductase, alpha subunit (EC 1.2.7.-)', 0.16), ('Polyribonucleotide nucleotidyltransferase (EC 2.7.7.8)', 0.16), ('Carboxynorspermidine decarboxylase (EC 4.1.1.96)', 0.16), ('Electron transport complex protein RnfG', 0.16), ('Succinyl-CoA ligase [ADP-forming] alpha chain (EC 6.2.1.5)', 0.16), ('UDP-N-acetylglucosamine 2-epimerase (EC 5.1.3.14)', 0.16), ('Alpha-1,2-mannosidase', 0.16), ('Uncharacterized corrinoid protein', 0.16), ('Ribosomal protein L11 methyltransferase', 0.16), ('VgrG protein', 0.16), ('Tryptophan synthase alpha chain (EC 4.2.1.20)', 0.16), ('Uncharacterized protein BT1741', 0.16), ('Putative aminotransferase BF2110', 0.16), ('Aliphatic amidase AmiE (EC 3.5.1.4)', 0.16), ('Death on curing protein, Doc toxin', 0.16), ('CobN-like chelatase BtuS for metalloporphyrine salvage', 0.16), ('8-amino-7-oxononanoate synthase (EC 2.3.1.47)', 0.16), ('Uncharacterized membrane protein YsdA', 0.14), ('Anthranilate synthase, aminase component (EC 4.1.3.27)', 0.14), ('Ribonuclease PH (EC 2.7.7.56)', 0.14), ('3-oxo-5-alpha-steroid 4-dehydrogenase', 0.14), ('ABC transporter, ATP-binding protein', 0.14), ('Tetracenomycin polyketide synthesis O-methyltransferase TcmP', 0.14), ('Toxin HigB / Protein kinase domain of HipA', 0.14), ('Acetylhydrolase', 0.14), ('Lipid-A-disaccharide synthase (EC 2.4.1.182)', 0.14), ('Ferredoxin', 0.14), ('RND efflux system, membrane fusion protein', 0.14), ('beta-N-acetylglucosaminidase (EC 3.2.1.52)', 0.14), ('L-alanyl-gamma-D-glutamyl-L-diamino acid endopeptidase', 0.14), ('L-fucose mutarotase (EC 5.1.3.29)', 0.14), ('Outer membrane receptor for Fe3+-dicitrate', 0.14), ('Efflux transport system, outer membrane factor (OMF) lipoprotein BT_0954', 0.12000000000000001), ('putative acetyl transferase', 0.12000000000000001), ('Hydrolase, alpha/beta fold family', 0.12000000000000001), ('Putative ATP-binding protein', 0.12000000000000001), ('Long tail fiber protein p37', 0.12000000000000001), ('Thiazole synthase (EC 2.8.1.10)', 0.12000000000000001), ('Cell division ZapA family protein', 0.12000000000000001), ('Exonuclease SbcD', 0.12000000000000001), ('Spermidine/putrescine import ABC transporter ATP-binding protein PotA (TC 3.A.1.11.1)', 0.12000000000000001), ('FIG034863: hypothetical protein', 0.12000000000000001), ('Sortase A, LPXTG specific', 0.1), ('SSU ribosomal protein S12p (S23e)', 0.1), ('FIG00672316: hypothetical protein', 0.1), ('Substrate-specific component RibU of riboflavin ECF transporter', 0.08), ('ABC transporter, permease protein 1 (cluster 1, maltose/g3p/polyamine/iron)', 0.08), ('Substrate-specific component YkoE of thiamin-regulated ECF transporter for HydroxyMethylPyrimidine', 0.08), ('TadB-like protein BAD_0130', 0.08), ('Sulfur carrier protein ThiS', 0.08), ('Pyrrolidone-carboxylate peptidase (EC 3.4.19.3)', 0.08), (\"Inosine-5'-monophosphate dehydrogenase, catalytic domain (EC 1.1.1.205)\", 0.08), ('Phospho-N-acetylmuramoyl-pentapeptide-transferase (EC 2.7.8.13)', 0.08), ('Maltose O-acetyltransferase (EC 2.3.1.79)', 0.08), ('Diaminohydroxyphosphoribosylaminopyrimidine deaminase (EC 3.5.4.26)', 0.06), ('Translation elongation factor Ts', 0.06), ('Uncharacterized UPF0721 integral membrane protein', 0.06), ('Methylated-DNA--protein-cysteine methyltransferase', 0.06), ('putative glycosyl transferase domain protein', 0.06), ('putative beta-glucosidase', 0.06), ('Integrase, catalytic region', 0.06), ('Mlc, transcriptional repressor of MalT (the transcriptional activator of maltose regulon) and manXYZ operon', 0.06), ('alpha-galactosidase (EC 3.2.1.22)', 0.06), ('Uncharacterized DUF349-containing protein SCO1511', 0.06), ('ABC transporter, permease protein 2 (cluster 1, maltose/g3p/polyamine/iron)', 0.06), ('Oligopeptide ABC transporter, ATP-binding protein OppF (TC 3.A.1.5.1)', 0.04), ('DNA topoisomerase IV subunit B (EC 5.99.1.3)', 0.04), ('NAD(P)H-hydrate epimerase (EC 5.1.99.6) / ADP-dependent (S)-NAD(P)H-hydrate dehydratase (EC 4.2.1.136)', 0.04), ('3-hydroxyacyl-[acyl-carrier-protein] dehydratase, FabZ form (EC 4.2.1.59)', 0), ('Acyl carrier protein', 0), ('Transcriptional regulator, MerR family', 0), ('Manganese-dependent protein-tyrosine phosphatase (EC 3.1.3.48)', 0), ('Biotin carboxylase of acetyl-CoA carboxylase (EC 6.3.4.14)', 0), ('RNA methyltransferase, TrmA family', 0), ('Phage lysin, 1,4-beta-N-acetylmuramidase (EC 3.2.1.17) or lysozyme', 0), ('OsmC/Ohr family protein', 0), ('Lysozyme (EC 3.2.1.17)', 0), ('Acetyl-coenzyme A carboxyl transferase beta chain (EC 6.4.1.2)', 0), ('Conjugative transposon protein TraN', 0), ('Conjugative transposon protein TraE', 0), ('Cell envelope-associated transcriptional attenuator LytR-CpsA-Psr, subfamily F2', 0), ('Oligo-1,6-glucosidase (EC 3.2.1.10)', 0), ('Choloylglycine hydrolase family protein', 0), ('Thioredoxin family protein', 0), ('LtrC-like protein', 0), ('Antirestriction protein ArdA', 0), ('Similar to carbon monoxide dehydrogenase CooS subunit', 0), ('Short-chain alcohol dehydrogenase', 0), ('PTS system, IIB component', 0), ('Uncharacterized ABC1 family protein Bbr_0249', 0), ('Mobile element protein', 0), ('Phosphoglycerate mutase family 5', 0), ('Biotin carboxyl carrier protein of acetyl-CoA carboxylase', 0), ('ABC transporter, permease protein', 0), ('3-oxoacyl-[acyl-carrier protein] reductase (EC 1.1.1.100), FadG', 0)]\n"
     ]
    }
   ],
   "source": [
    "for tp_type in gene_change_null_counts_by_tp_type:\n",
    "    print(sorted(gene_change_null_counts_by_tp_type[tp_type].items(), key=lambda x: x[1], reverse=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gene_change_counts_by_tp_type' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mNameError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-e537997e2aea>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mtp_type\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_change_counts_by_tp_type\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m     \u001b[0moutput\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'%s/gene_changes_%s.tsv'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mconfig\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0manalysis_directory\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtp_type\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'w'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwrite\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\t'\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mjoin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'gene_name'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_changes'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_changes_null'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_gains'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_gains_null'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_losses'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'num_losses_null'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mgene\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgene_change_counts_by_tp_type\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtp_type\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0mnum_changes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_change_counts_by_tp_type\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mtp_type\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gene_change_counts_by_tp_type' is not defined"
     ]
    }
   ],
   "source": [
    "for tp_type in gene_change_counts_by_tp_type:\n",
    "    output = open('%s/gene_changes_%s.tsv' % (config.analysis_directory, tp_type), 'w')\n",
    "    output.write('\\t'.join(['gene_name', 'num_changes', 'num_changes_null', 'num_gains', 'num_gains_null', 'num_losses', 'num_losses_null']) + '\\n')\n",
    "    for gene in gene_change_counts_by_tp_type[tp_type]:\n",
    "        num_changes = gene_change_counts_by_tp_type[tp_type][gene]\n",
    "        num_losses = 0 if gene not in lost_genes_counts_by_tp_type[tp_type] else lost_genes_counts_by_tp_type[tp_type][gene]\n",
    "        num_gains = 0 if gene not in gained_genes_counts_by_tp_type[tp_type] else gained_genes_counts_by_tp_type[tp_type][gene]\n",
    "        num_changes_null = gene_change_null_counts_by_tp_type[tp_type][gene]\n",
    "        num_losses_null = 0 if gene not in lost_genes_counts_nulls_by_tp_type[tp_type] else lost_genes_counts_nulls_by_tp_type[tp_type][gene]        \n",
    "        num_gains_null = 0 if gene not in gained_genes_counts_nulls_by_tp_type[tp_type] else gained_genes_counts_nulls_by_tp_type[tp_type][gene]\n",
    "        output.write('\\t'.join([str(item) for item in [gene, num_changes, num_changes_null, num_gains, num_gains_null, num_losses, num_losses_null]]) + '\\n')\n",
    "    output.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Returns two arrays, one of ordered timepoints\n",
    "# and other of corresponding values\n",
    "def order_by_tp(tp_dict):\n",
    "    items = tp_dict.items()\n",
    "    mother_items = filter(lambda x: x[0][0] == 'M', items)\n",
    "    mother_sorted = sorted(mother_items, key=lambda x: int(x[0][1:]))\n",
    "    infant_items = filter(lambda x: x[0][0] == 'I', items)\n",
    "    infant_sorted = sorted(infant_items, key=lambda x: int(x[0][1:]))\n",
    "    comb_sorted = mother_sorted + infant_sorted\n",
    "    return comb_sorted if comb_sorted != [] else sorted(items, key=lambda x: int(x[0][1:]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Returns True if one value is below lower_threshold\n",
    "# and the other value is above upper_threshold\n",
    "def is_diff(val1, val2, lower_threshold, upper_threshold):\n",
    "    return (val1 <= lower_threshold and val2 >= upper_threshold) or (val1 >= upper_threshold and val2 <= lower_threshold)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Determines if list of values has reversion\n",
    "def has_reversion(vals, lower_threshold, upper_threshold):\n",
    "    ref_val = vals[0]\n",
    "    one_change = False\n",
    "    for val in vals[1:]:\n",
    "        # While first change has yet to occur\n",
    "        if one_change == False:\n",
    "            if is_diff(val, ref_val, lower_threshold, upper_threshold):\n",
    "                # If first change detected, set ref_val to new value\n",
    "                # and set one_change to True\n",
    "                ref_val = val\n",
    "                one_change = True\n",
    "        # First change has already occurred with ref_val reset\n",
    "        elif one_change == True:\n",
    "            if is_diff(val, ref_val, lower_threshold, upper_threshold):\n",
    "                # Second change (must be in opp. direction) detected\n",
    "                return True\n",
    "    return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Getting gene info for Bacteroides_intestinalis_61596...\n",
      "Getting gene info for Bacteroides_uniformis_57318...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-8-1950530fe417>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     11\u001b[0m         \u001b[0mgenome_ids\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmidas_db_utils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_ref_genome_ids\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     12\u001b[0m         \u001b[0mnon_shared_genes\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcore_gene_utils\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mparse_non_shared_pangenome_genes\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 13\u001b[0;31m         \u001b[0mgene_desc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mparse_patric\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mload_patric_gene_descriptions\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mgenome_ids\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnon_shared_genes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     14\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mcohort\u001b[0m \u001b[0;32min\u001b[0m \u001b[0msnp_modifications_by_site\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mspecies\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/u/project/ngarud/daisyche/mother_infant/scripts/utils/parse_patric.pyc\u001b[0m in \u001b[0;36mload_patric_gene_descriptions\u001b[0;34m(genome_ids, allowed_genes)\u001b[0m\n\u001b[1;32m    119\u001b[0m                                         \u001b[0;32mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"PATRIC file for \"\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0mgenome_id\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0;34m\" not found!\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    120\u001b[0m                                         \u001b[0;32mcontinue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 121\u001b[0;31m                                 \u001b[0mfile\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m#header\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    122\u001b[0m                                 \u001b[0;32mfor\u001b[0m \u001b[0mline\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    123\u001b[0m                                                 \u001b[0mitems\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msplit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"\\t\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/u/local/apps/python/2.7.13/lib/python2.7/gzip.pyc\u001b[0m in \u001b[0;36mreadline\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    462\u001b[0m         \u001b[0mbufs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    463\u001b[0m         \u001b[0;32mwhile\u001b[0m \u001b[0msize\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 464\u001b[0;31m             \u001b[0mc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreadsize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    465\u001b[0m             \u001b[0mi\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    466\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/u/local/apps/python/2.7.13/lib/python2.7/gzip.pyc\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    266\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    267\u001b[0m                 \u001b[0;32mwhile\u001b[0m \u001b[0msize\u001b[0m \u001b[0;34m>\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mextrasize\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 268\u001b[0;31m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreadsize\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    269\u001b[0m                     \u001b[0mreadsize\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax_read_chunk\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mreadsize\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    270\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mEOFError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/u/local/apps/python/2.7.13/lib/python2.7/gzip.pyc\u001b[0m in \u001b[0;36m_read\u001b[0;34m(self, size)\u001b[0m\n\u001b[1;32m    294\u001b[0m             \u001b[0;31m# if so, it's time to stop; no more members to read.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    295\u001b[0m             \u001b[0mpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfileobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtell\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m   \u001b[0;31m# Save current position\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 296\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfileobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseek\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m     \u001b[0;31m# Seek to end of file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mpos\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfileobj\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtell\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    298\u001b[0m                 \u001b[0;32mraise\u001b[0m \u001b[0mEOFError\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"Reached EOF\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "reversions = []\n",
    "lower_threshold = 0.2\n",
    "upper_threshold = 0.8\n",
    "\n",
    "# Enumerate reversions\n",
    "for species in snp_modifications_by_site:\n",
    "    \n",
    "    # Also prepare gene information\n",
    "    if sum([len(snp_modifications_by_site[species][cohort].keys()) for cohort in snp_modifications_by_site[species]]) != 0:\n",
    "        print(\"Getting gene info for %s...\" % species)\n",
    "        genome_ids = midas_db_utils.get_ref_genome_ids(species)\n",
    "        non_shared_genes = core_gene_utils.parse_non_shared_pangenome_genes(species)\n",
    "        gene_desc = parse_patric.load_patric_gene_descriptions(genome_ids, non_shared_genes)\n",
    "    \n",
    "    for cohort in snp_modifications_by_site[species]:\n",
    "        subdict = snp_modifications_by_site[species][cohort]\n",
    "        for subject in subdict:\n",
    "            for site in subdict[subject]:\n",
    "                if len(subdict[subject][site]) > 1:\n",
    "                    tp_freq_dict = {} # freq info only\n",
    "                    for snp_change in subdict[subject][site]:\n",
    "                        tp_pair, gene_id, var_type, A1, D1, A2, D2 = snp_change\n",
    "                        tp1, tp2 = tp_pair\n",
    "                        tp_freq_dict[tp1] = (A1/float(D1))\n",
    "                        tp_freq_dict[tp2] = (A2/float(D2))\n",
    "                    ordered_tp_freqs = order_by_tp(tp_freq_dict)\n",
    "                    gene_name = gene_desc[gene_id] if gene_id in gene_desc else 'No name'\n",
    "                    if has_reversion([freq for tp, freq in ordered_tp_freqs], lower_threshold, upper_threshold):\n",
    "                        reversions.append((species, cohort, subject, site, gene_id, gene_name, var_type, ordered_tp_freqs))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# revised to differentiate premie and nonpremie\n",
    "def tp_pair_to_tp_type(tp_pair, cohort):\n",
    "    tpa, tpb = tp_pair\n",
    "    tp_type = tpa[0] + tpb[0]\n",
    "    if tp_type == 'IM':\n",
    "        tp_type = 'MI'\n",
    "    if tp_type in ['II', 'MI']:\n",
    "        if cohort == 'olm':\n",
    "            tp_type += '-premie'\n",
    "        else:\n",
    "            tp_type += '-nonpremie'\n",
    "    return tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "cohorts = ['backhed', 'yassour', 'ferretti', 'hmp', 'shao', 'olm']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "num_snp_change_by_cohort = {cohort: 0 for cohort in cohorts}\n",
    "changed_gene_ids_by_cohort = {cohort: set() for cohort in cohorts}\n",
    "changed_gene_names_by_cohort = {cohort: set() for cohort in cohorts}\n",
    "changed_gene_name_counts_by_cohort = {cohort: defaultdict(int) for cohort in cohorts}\n",
    "\n",
    "tp_types = ['AA', 'II-premie', 'II-nonpremie','MM', 'MI-premie', 'MI-nonpremie']\n",
    "num_snp_change_by_tp_type = {tp_type: 0 for tp_type in tp_types}\n",
    "changed_gene_names_by_tp_type = {tp_type: set() for tp_type in tp_types}\n",
    "changed_gene_name_counts_by_tp_type = {tp_type: defaultdict(int) for tp_type in tp_types}\n",
    "\n",
    "# Enumerate genes\n",
    "for species in snp_modification_genes:\n",
    "    for cohort in snp_modification_genes[species]:\n",
    "        for subject in snp_modification_genes[species][cohort]:\n",
    "            subdict = snp_modification_genes[species][cohort][subject]\n",
    "            # For each sample pair in this cohort/species,\n",
    "            # have a count of how many SNP changes are\n",
    "            # in a particular gene\n",
    "            for tp_pair in subdict:\n",
    "                tp_type = tp_pair_to_tp_type(tp_pair, cohort)\n",
    "                for gene_id, gene_name in subdict[tp_pair]:\n",
    "                    snp_change_count = subdict[tp_pair][(gene_id, gene_name)]\n",
    "\n",
    "                    num_snp_change_by_cohort[cohort] += snp_change_count\n",
    "                    changed_gene_ids_by_cohort[cohort].add(gene_id)\n",
    "                    changed_gene_names_by_cohort[cohort].add(gene_name)\n",
    "                    changed_gene_name_counts_by_cohort[cohort][gene_name] += snp_change_count\n",
    "\n",
    "                    num_snp_change_by_tp_type[tp_type] += snp_change_count\n",
    "                    changed_gene_names_by_tp_type[tp_type].add(gene_name)\n",
    "                    changed_gene_name_counts_by_tp_type[tp_type][gene_name] += snp_change_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AA': 259,\n",
       " 'II-nonpremie': 276,\n",
       " 'II-premie': 38,\n",
       " 'MI-nonpremie': 101,\n",
       " 'MI-premie': 0,\n",
       " 'MM': 33}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_snp_change_by_tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA: 114\n",
      "II-premie: 28\n",
      "MI-premie: 0\n",
      "MM: 27\n",
      "MI-nonpremie: 54\n",
      "II-nonpremie: 107\n"
     ]
    }
   ],
   "source": [
    "for tp_type in changed_gene_names_by_tp_type:\n",
    "    print(\"%s: %i\" % (tp_type,len(changed_gene_names_by_tp_type[tp_type])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AA\n",
      "76 hypothetical protein\n",
      "14 Oligopeptide ABC transporter, permease protein OppC (TC 3.A.1.5.1)\n",
      "12 Biotin synthase (EC 2.8.1.6)\n",
      "10 HAD-superfamily hydrolase, subfamily IA, variant 3\n",
      "6 RND efflux system, inner membrane transporter\n",
      "6 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "5 Outer membrane protein oprM\n",
      "5 Oligopeptide ABC transporter, permease protein OppB (TC 3.A.1.5.1)\n",
      "5 Vitamin B12 ABC transporter, ATP-binding protein BtuD\n",
      "4 Ribonuclease J2 (endoribonuclease in RNA processing)\n",
      "3 Sensory box/GGDEF family protein\n",
      "3 N/A\n",
      "3 DNA gyrase subunit A (EC 5.99.1.3)\n",
      "2 Two-component system sensor histidine kinase\n",
      "2 Vitamin B12 ABC transporter, substrate-binding protein BtuF\n",
      "2 ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)\n",
      "2 Argininosuccinate lyase (EC 4.3.2.1)\n",
      "2 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "\n",
      "II-premie\n",
      "5 N/A\n",
      "3 Putative glutathione transporter, permease component; Dipeptide ABC transporter, permease protein DppC (TC 3.A.1.5.2)\n",
      "3 hypothetical protein\n",
      "2 Uncharacterized transporter YbxG\n",
      "2 Transcriptional regulator pfoR\n",
      "\n",
      "MI-premie\n",
      "\n",
      "MM\n",
      "7 hypothetical protein\n",
      "\n",
      "MI-nonpremie\n",
      "19 hypothetical protein\n",
      "13 N/A\n",
      "5 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "4 beta-glucosidase (EC 3.2.1.21)\n",
      "4 Electron transport complex protein RnfE\n",
      "3 putative TonB-dependent receptor\n",
      "2 Two-component system sensor histidine kinase/response regulator hybrid\n",
      "2 Two-component system sensor histidine kinase\n",
      "2 ThiF family protein, ubiquitin-activating enzyme\n",
      "2 Tyrosine-protein kinase (EC 2.7.10.2)\n",
      "2 RND efflux system, membrane fusion protein\n",
      "\n",
      "II-nonpremie\n",
      "39 hypothetical protein\n",
      "37 Transcription termination factor Rho\n",
      "25 Protoporphyrinogen oxidase\n",
      "22 Ribonucleotide reductase of class Ib (aerobic), alpha subunit (EC 1.17.4.1)\n",
      "16 FIG000875: Thioredoxin domain-containing protein EC-YbbN\n",
      "8 narrowly conserved hypothetical protein\n",
      "5 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "4 Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator\n",
      "4 DNA gyrase subunit B (EC 5.99.1.3)\n",
      "3 Tn916, NLP/P60 family protein\n",
      "3 Glycogen phosphorylase (EC 2.4.1.1)\n",
      "3 Aspartate ammonia-lyase (EC 4.3.1.1)\n",
      "3 Serine/threonine-protein kinase PknB (EC 2.7.11.1)\n",
      "2 RidA/YER057c/UK114 superfamily protein\n",
      "2 Sialidase (EC 3.2.1.18)\n",
      "2 ABC transporter, substrate-binding protein (cluster 1, maltose/g3p/polyamine/iron)\n",
      "2 beta-galactosidase (EC 3.2.1.23)\n",
      "2 Sporulation transcription regulator WhiA\n",
      "2 RND efflux system, membrane fusion protein\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "2 Hypothetical, related to broad specificity phosphatases COG0406\n",
      "2 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "2 large transmembrane protein possibly involved in transport\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for tp_type in changed_gene_name_counts_by_tp_type:\n",
    "    subdict = changed_gene_name_counts_by_tp_type[tp_type]\n",
    "    print(tp_type)\n",
    "    for gene_name, count in sorted(subdict.items(), key=lambda x: x[1], reverse=True):\n",
    "        if count > 1:\n",
    "            print(\"%i %s\" % (count, gene_name))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'backhed': 166,\n",
       " 'ferretti': 34,\n",
       " 'hmp': 259,\n",
       " 'olm': 38,\n",
       " 'shao': 153,\n",
       " 'yassour': 57}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_snp_change_by_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmp: 176\n",
      "olm: 31\n",
      "yassour: 48\n",
      "backhed: 114\n",
      "ferretti: 13\n",
      "shao: 68\n"
     ]
    }
   ],
   "source": [
    "for cohort in changed_gene_ids_by_cohort:\n",
    "    print(\"%s: %i\" % (cohort,len(changed_gene_ids_by_cohort[cohort])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmp: 114\n",
      "olm: 28\n",
      "yassour: 38\n",
      "backhed: 81\n",
      "ferretti: 12\n",
      "shao: 51\n"
     ]
    }
   ],
   "source": [
    "for cohort in changed_gene_names_by_cohort:\n",
    "    print(\"%s: %i\" % (cohort,len(changed_gene_names_by_cohort[cohort])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmp\n",
      "76 hypothetical protein\n",
      "14 Oligopeptide ABC transporter, permease protein OppC (TC 3.A.1.5.1)\n",
      "12 Biotin synthase (EC 2.8.1.6)\n",
      "10 HAD-superfamily hydrolase, subfamily IA, variant 3\n",
      "6 RND efflux system, inner membrane transporter\n",
      "6 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "5 Outer membrane protein oprM\n",
      "5 Oligopeptide ABC transporter, permease protein OppB (TC 3.A.1.5.1)\n",
      "5 Vitamin B12 ABC transporter, ATP-binding protein BtuD\n",
      "4 Ribonuclease J2 (endoribonuclease in RNA processing)\n",
      "3 Sensory box/GGDEF family protein\n",
      "3 N/A\n",
      "3 DNA gyrase subunit A (EC 5.99.1.3)\n",
      "2 Two-component system sensor histidine kinase\n",
      "2 Vitamin B12 ABC transporter, substrate-binding protein BtuF\n",
      "2 ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)\n",
      "2 Argininosuccinate lyase (EC 4.3.2.1)\n",
      "2 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "\n",
      "olm\n",
      "5 N/A\n",
      "3 Putative glutathione transporter, permease component; Dipeptide ABC transporter, permease protein DppC (TC 3.A.1.5.2)\n",
      "3 hypothetical protein\n",
      "2 Uncharacterized transporter YbxG\n",
      "2 Transcriptional regulator pfoR\n",
      "\n",
      "yassour\n",
      "10 hypothetical protein\n",
      "4 beta-glucosidase (EC 3.2.1.21)\n",
      "4 Electron transport complex protein RnfE\n",
      "3 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "2 Two-component system sensor histidine kinase/response regulator hybrid\n",
      "2 Tyrosine-protein kinase (EC 2.7.10.2)\n",
      "\n",
      "backhed\n",
      "29 hypothetical protein\n",
      "16 FIG000875: Thioredoxin domain-containing protein EC-YbbN\n",
      "8 narrowly conserved hypothetical protein\n",
      "5 N/A\n",
      "5 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "4 RND efflux system, membrane fusion protein\n",
      "3 Two-component system sensor histidine kinase\n",
      "3 Aspartate ammonia-lyase (EC 4.3.1.1)\n",
      "3 putative TonB-dependent receptor\n",
      "3 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "3 beta-galactosidase (EC 3.2.1.23)\n",
      "2 large transmembrane protein possibly involved in transport\n",
      "2 Malate dehydrogenase (EC 1.1.1.37)\n",
      "2 Esterase/lipase\n",
      "2 Lactose and galactose permease, GPH translocator family\n",
      "2 Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator\n",
      "2 RidA/YER057c/UK114 superfamily protein\n",
      "2 ABC transporter, substrate-binding protein (cluster 1, maltose/g3p/polyamine/iron)\n",
      "2 Cell division initiation protein DivIVA\n",
      "2 putative protein involved in capsular polysaccharide biosynthesis\n",
      "2 putative two-component system sensor but no kinase domain\n",
      "2 Sialidase (EC 3.2.1.18)\n",
      "2 Nitrogen regulatory protein P-II\n",
      "2 L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)\n",
      "2 Putative polysaccharide deacetylase\n",
      "\n",
      "ferretti\n",
      "21 Ribonucleotide reductase of class Ib (aerobic), alpha subunit (EC 1.17.4.1)\n",
      "2 ThiF family protein, ubiquitin-activating enzyme\n",
      "2 hypothetical protein\n",
      "\n",
      "shao\n",
      "37 Transcription termination factor Rho\n",
      "25 Protoporphyrinogen oxidase\n",
      "24 hypothetical protein\n",
      "9 N/A\n",
      "4 DNA gyrase subunit B (EC 5.99.1.3)\n",
      "3 Tn916, NLP/P60 family protein\n",
      "3 Serine/threonine-protein kinase PknB (EC 2.7.11.1)\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "2 Hypothetical, related to broad specificity phosphatases COG0406\n",
      "2 Glycogen phosphorylase (EC 2.4.1.1)\n",
      "2 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for cohort in changed_gene_name_counts_by_cohort:\n",
    "    subdict = changed_gene_name_counts_by_cohort[cohort]\n",
    "    print(cohort)\n",
    "    for gene_name, count in sorted(subdict.items(), key=lambda x: x[1], reverse=True):\n",
    "        if count > 1:\n",
    "            print(\"%i %s\" % (count, gene_name))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "grouped_cohorts = {'hmp': ['hmp'], 'non-premie': ['backhed', 'yassour', 'ferretti', 'shao'], 'premie': ['olm'], 'infant': ['backhed', 'yassour', 'ferretti', 'shao', 'olm']}\n",
    "changed_gene_name_counts_by_grouped_cohort = {grouped_cohort: defaultdict(int) for grouped_cohort in grouped_cohorts.keys()}\n",
    "\n",
    "# Group by premie and non-premie\n",
    "for cohort in changed_gene_name_counts_by_cohort:\n",
    "    subdict = changed_gene_name_counts_by_cohort[cohort]\n",
    "    for grouped_cohort in grouped_cohorts:\n",
    "        if cohort in grouped_cohorts[grouped_cohort]:\n",
    "            for gene_name in subdict:\n",
    "                changed_gene_name_counts_by_grouped_cohort[grouped_cohort][gene_name] += subdict[gene_name]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hmp\n",
      "76 hypothetical protein\n",
      "14 Oligopeptide ABC transporter, permease protein OppC (TC 3.A.1.5.1)\n",
      "12 Biotin synthase (EC 2.8.1.6)\n",
      "10 HAD-superfamily hydrolase, subfamily IA, variant 3\n",
      "6 RND efflux system, inner membrane transporter\n",
      "6 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "5 Outer membrane protein oprM\n",
      "5 Oligopeptide ABC transporter, permease protein OppB (TC 3.A.1.5.1)\n",
      "5 Vitamin B12 ABC transporter, ATP-binding protein BtuD\n",
      "4 Ribonuclease J2 (endoribonuclease in RNA processing)\n",
      "3 Sensory box/GGDEF family protein\n",
      "3 N/A\n",
      "3 DNA gyrase subunit A (EC 5.99.1.3)\n",
      "2 Two-component system sensor histidine kinase\n",
      "2 Vitamin B12 ABC transporter, substrate-binding protein BtuF\n",
      "2 ATP:Cob(I)alamin adenosyltransferase (EC 2.5.1.17)\n",
      "2 Argininosuccinate lyase (EC 4.3.2.1)\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "2 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "\n",
      "infant\n",
      "68 hypothetical protein\n",
      "37 Transcription termination factor Rho\n",
      "25 Protoporphyrinogen oxidase\n",
      "22 Ribonucleotide reductase of class Ib (aerobic), alpha subunit (EC 1.17.4.1)\n",
      "20 N/A\n",
      "16 FIG000875: Thioredoxin domain-containing protein EC-YbbN\n",
      "11 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "8 narrowly conserved hypothetical protein\n",
      "5 beta-glucosidase (EC 3.2.1.21)\n",
      "5 Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator\n",
      "4 Electron transport complex protein RnfE\n",
      "4 beta-galactosidase (EC 3.2.1.23)\n",
      "4 RND efflux system, membrane fusion protein\n",
      "4 DNA gyrase subunit B (EC 5.99.1.3)\n",
      "3 Two-component system sensor histidine kinase\n",
      "3 Tn916, NLP/P60 family protein\n",
      "3 Putative glutathione transporter, permease component; Dipeptide ABC transporter, permease protein DppC (TC 3.A.1.5.2)\n",
      "3 Glycogen phosphorylase (EC 2.4.1.1)\n",
      "3 putative TonB-dependent receptor\n",
      "3 Aspartate ammonia-lyase (EC 4.3.1.1)\n",
      "3 Serine/threonine-protein kinase PknB (EC 2.7.11.1)\n",
      "3 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "2 putative protein involved in capsular polysaccharide biosynthesis\n",
      "2 RidA/YER057c/UK114 superfamily protein\n",
      "2 L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)\n",
      "2 Sialidase (EC 3.2.1.18)\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "2 ABC transporter, substrate-binding protein (cluster 1, maltose/g3p/polyamine/iron)\n",
      "2 Two-component system sensor histidine kinase/response regulator hybrid\n",
      "2 Esterase/lipase\n",
      "2 Tyrosine-protein kinase (EC 2.7.10.2)\n",
      "2 Lactose and galactose permease, GPH translocator family\n",
      "2 Putative polysaccharide deacetylase\n",
      "2 ThiF family protein, ubiquitin-activating enzyme\n",
      "2 Sporulation transcription regulator WhiA\n",
      "2 Nitrogen regulatory protein P-II\n",
      "2 Phosphate regulon sensor protein PhoR (SphS) (EC 2.7.13.3)\n",
      "2 Uncharacterized transporter YbxG\n",
      "2 Cell division initiation protein DivIVA\n",
      "2 Malate dehydrogenase (EC 1.1.1.37)\n",
      "2 Transcriptional regulator pfoR\n",
      "2 Glycosyl transferase, family 2\n",
      "2 Hypothetical, related to broad specificity phosphatases COG0406\n",
      "2 large transmembrane protein possibly involved in transport\n",
      "2 putative two-component system sensor but no kinase domain\n",
      "\n",
      "premie\n",
      "5 N/A\n",
      "3 Putative glutathione transporter, permease component; Dipeptide ABC transporter, permease protein DppC (TC 3.A.1.5.2)\n",
      "3 hypothetical protein\n",
      "2 Uncharacterized transporter YbxG\n",
      "2 Transcriptional regulator pfoR\n",
      "\n",
      "non-premie\n",
      "65 hypothetical protein\n",
      "37 Transcription termination factor Rho\n",
      "25 Protoporphyrinogen oxidase\n",
      "22 Ribonucleotide reductase of class Ib (aerobic), alpha subunit (EC 1.17.4.1)\n",
      "16 FIG000875: Thioredoxin domain-containing protein EC-YbbN\n",
      "15 N/A\n",
      "11 Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "8 narrowly conserved hypothetical protein\n",
      "5 beta-glucosidase (EC 3.2.1.21)\n",
      "5 Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator\n",
      "4 Electron transport complex protein RnfE\n",
      "4 beta-galactosidase (EC 3.2.1.23)\n",
      "4 RND efflux system, membrane fusion protein\n",
      "4 DNA gyrase subunit B (EC 5.99.1.3)\n",
      "3 Two-component system sensor histidine kinase\n",
      "3 Tn916, NLP/P60 family protein\n",
      "3 Glycogen phosphorylase (EC 2.4.1.1)\n",
      "3 putative TonB-dependent receptor\n",
      "3 Aspartate ammonia-lyase (EC 4.3.1.1)\n",
      "3 Serine/threonine-protein kinase PknB (EC 2.7.11.1)\n",
      "3 Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family\n",
      "2 putative protein involved in capsular polysaccharide biosynthesis\n",
      "2 RidA/YER057c/UK114 superfamily protein\n",
      "2 Cell division initiation protein DivIVA\n",
      "2 L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)\n",
      "2 Sialidase (EC 3.2.1.18)\n",
      "2 ABC transporter, substrate-binding protein (cluster 1, maltose/g3p/polyamine/iron)\n",
      "2 Two-component system sensor histidine kinase/response regulator hybrid\n",
      "2 Esterase/lipase\n",
      "2 Tyrosine-protein kinase (EC 2.7.10.2)\n",
      "2 Putative polysaccharide deacetylase\n",
      "2 ThiF family protein, ubiquitin-activating enzyme\n",
      "2 Lactose and galactose permease, GPH translocator family\n",
      "2 Sporulation transcription regulator WhiA\n",
      "2 Nitrogen regulatory protein P-II\n",
      "2 DNA-directed RNA polymerase beta subunit (EC 2.7.7.6)\n",
      "2 large transmembrane protein possibly involved in transport\n",
      "2 Phosphate regulon sensor protein PhoR (SphS) (EC 2.7.13.3)\n",
      "2 Malate dehydrogenase (EC 1.1.1.37)\n",
      "2 Glycosyl transferase, family 2\n",
      "2 Hypothetical, related to broad specificity phosphatases COG0406\n",
      "2 putative two-component system sensor but no kinase domain\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for grouped_cohort in changed_gene_name_counts_by_grouped_cohort:\n",
    "    subdict = changed_gene_name_counts_by_grouped_cohort[grouped_cohort]\n",
    "    print(grouped_cohort)\n",
    "    for gene_name, count in sorted(subdict.items(), key=lambda x: x[1], reverse=True):\n",
    "        if count > 1:\n",
    "            print(\"%i %s\" % (count, gene_name))\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'1-acyl-sn-glycerol-3-phosphate acyltransferase': 1,\n",
       "             '16S rRNA (cytosine(1402)-N(4))-methyltransferase (EC 2.1.1.199)': 1,\n",
       "             'ABC transporter, ATP-binding protein (cluster 1, maltose/g3p/polyamine/iron); ABC transporter, ATP-binding protein (cluster 10, nitrate/sulfonate/bicarbonate)': 1,\n",
       "             'ABC transporter, permease protein 1 (cluster 1, maltose/g3p/polyamine/iron)': 1,\n",
       "             'ABC transporter, substrate-binding protein (cluster 1, maltose/g3p/polyamine/iron)': 2,\n",
       "             'ATP-dependent DNA helicase RecQ': 1,\n",
       "             'Acetylornithine deacetylase/Succinyl-diaminopimelate desuccinylase and related deacylases': 1,\n",
       "             'Acidobacterial duplicated orphan permease (function unknown)': 1,\n",
       "             'Acyl-[acyl-carrier-protein]--UDP-N-acetylglucosamine O-acyltransferase (EC 2.3.1.129)': 1,\n",
       "             'Alpha-1,2-mannosidase': 1,\n",
       "             'Aminopeptidase C (EC 3.4.22.40)': 1,\n",
       "             'Aminotransferase, DegT/DnrJ/EryC1/StrS family': 1,\n",
       "             'Anthranilate phosphoribosyltransferase (EC 2.4.2.18)': 1,\n",
       "             'Arabinan endo-1,5-alpha-L-arabinosidase (EC 3.2.1.99)': 1,\n",
       "             'Asparagine synthetase [glutamine-hydrolyzing] (EC 6.3.5.4) AsnB': 1,\n",
       "             'Aspartate ammonia-lyase (EC 4.3.1.1)': 3,\n",
       "             'Biotin carboxylase of acetyl-CoA carboxylase (EC 6.3.4.14)': 1,\n",
       "             'Cell division initiation protein DivIVA': 2,\n",
       "             'Cell division protein FtsL': 1,\n",
       "             'Cell surface glycan-binding lipoprotein, utilization system for glycans and polysaccharides (PUL), SusD family': 3,\n",
       "             'Chitinase (EC 3.2.1.14)': 1,\n",
       "             'Chromate transport protein': 1,\n",
       "             'Citrate/H+ symporter of CitMHS family': 1,\n",
       "             'Cobalt-zinc-cadmium resistance protein': 1,\n",
       "             'DNA ligase (NAD(+)) (EC 6.5.1.2)': 1,\n",
       "             'DNA-binding transcriptional activator DecR, AsnC family': 1,\n",
       "             'Di-tripeptide/cation symporter': 1,\n",
       "             'DnaD domain protein': 1,\n",
       "             'Esterase/lipase': 2,\n",
       "             'Excinuclease ABC subunit B': 1,\n",
       "             'FIG000875: Thioredoxin domain-containing protein EC-YbbN': 16,\n",
       "             'Glutaminyl-tRNA synthetase (EC 6.1.1.18)': 1,\n",
       "             'Glycerol-3-phosphate dehydrogenase [NAD(P)+] (EC 1.1.1.94)': 1,\n",
       "             'Glycogen phosphorylase (EC 2.4.1.1)': 1,\n",
       "             'Glycosyl transferase, family 2': 1,\n",
       "             'L-serine dehydratase, beta subunit (EC 4.3.1.17) / L-serine dehydratase, alpha subunit (EC 4.3.1.17)': 2,\n",
       "             'Lactose and galactose permease, GPH translocator family': 2,\n",
       "             'Malate dehydrogenase (EC 1.1.1.37)': 2,\n",
       "             'Methylated-DNA--protein-cysteine methyltransferase (EC 2.1.1.63)': 1,\n",
       "             'Multi antimicrobial extrusion protein (Na(+)/drug antiporter), MATE family of MDR efflux pumps': 1,\n",
       "             'N/A': 5,\n",
       "             'NifU-related domain containing protein': 1,\n",
       "             'Nitrogen regulatory protein P-II': 2,\n",
       "             'O-antigen chain length determinant protein WzzB': 1,\n",
       "             'OmpA family protein': 1,\n",
       "             'Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family': 5,\n",
       "             'Oxidoreductase, Gfo/Idh/MocA family': 1,\n",
       "             'Periplasmic ligand-binding sensor domain COG3292 / BaeS-type histidine kinase / OmpR-type DNA-binding response regulator': 2,\n",
       "             'Phosphate regulon sensor protein PhoR (SphS) (EC 2.7.13.3)': 1,\n",
       "             'Phosphotransferase involved in threonylcarbamoyladenosine t(6)A37 formation in tRNA': 1,\n",
       "             'Probable outer membrane receptor protein': 1,\n",
       "             'Putative NADH-dependent dehydrogenase': 1,\n",
       "             'Putative TonB-dependent receptor': 1,\n",
       "             'Putative helicase': 1,\n",
       "             'Putative large secreted protein SCO0341': 1,\n",
       "             'Putative outer membrane protein, probably involved in nutrient binding': 1,\n",
       "             'Putative polysaccharide deacetylase': 2,\n",
       "             'Putative sulfatase': 1,\n",
       "             'RND efflux system, membrane fusion protein': 4,\n",
       "             'RidA/YER057c/UK114 superfamily protein': 2,\n",
       "             'Sialidase (EC 3.2.1.18)': 2,\n",
       "             'Spermidine/putrescine import ABC transporter substrate-binding protein PotD (TC 3.A.1.11.1)': 1,\n",
       "             'Sporulation transcription regulator WhiA': 1,\n",
       "             'Tetratricopeptide repeat family protein': 1,\n",
       "             'ThiJ/PfpI family': 1,\n",
       "             'Transcriptional regulator': 1,\n",
       "             'Transcriptional regulator, AcrR family': 1,\n",
       "             'Transcriptional regulator, AraC family': 1,\n",
       "             'Two-component system sensor histidine kinase': 3,\n",
       "             'Two-component transcriptional response regulator, LuxR family': 1,\n",
       "             'UDP-glucose 6-dehydrogenase (EC 1.1.1.22)': 1,\n",
       "             'Valyl-tRNA synthetase (EC 6.1.1.9)': 1,\n",
       "             'alpha-L-fucosidase (EC 3.2.1.51)': 1,\n",
       "             'beta-galactosidase (EC 3.2.1.23)': 3,\n",
       "             'beta-glucosidase (EC 3.2.1.21)': 1,\n",
       "             'hypothetical protein': 29,\n",
       "             'large transmembrane protein possibly involved in transport': 2,\n",
       "             'narrowly conserved hypothetical protein': 8,\n",
       "             'putative TonB-dependent receptor': 3,\n",
       "             'putative protein involved in capsular polysaccharide biosynthesis': 2,\n",
       "             'putative two-component system sensor but no kinase domain': 2})"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "changed_gene_name_counts_by_cohort['backhed']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genes_tp_pairs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m\u001b[0m",
      "\u001b[0;31mNameError\u001b[0mTraceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-ba4a14aba8bd>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mgenes_cohort\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m{\u001b[0m\u001b[0mcohort\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mcohort\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcohorts\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mcohort\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenes_tp_pairs\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mgene_tuple\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenes_tp_pairs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcohort\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mtp_pair\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mgenes_tp_pairs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcohort\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mgene_tuple\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genes_tp_pairs' is not defined"
     ]
    }
   ],
   "source": [
    "genes_cohort = {cohort: set() for cohort in cohorts}\n",
    "\n",
    "for cohort in genes_tp_pairs:\n",
    "    for gene_tuple in genes_tp_pairs[cohort]:\n",
    "        for tp_pair in genes_tp_pairs[cohort][gene_tuple]:\n",
    "            tpa, tpb = tp_pair\n",
    "            tp_type = tpa[0] + tpb[0]\n",
    "            if tp_type == 'II':\n",
    "                genes_cohort[cohort].add(gene_tuple)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'backhed': set(),\n",
       " 'ferretti': set(),\n",
       " 'hmp': set(),\n",
       " 'olm': set(),\n",
       " 'shao': set(),\n",
       " 'yassour': set()}"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_cohort"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Enterococcus_faecalis_56297 | olm | Subject: N1_018\n",
      "Gene id: 1158976.3.peg.771 | Variant type: 4D\n",
      "Gene name: hypothetical protein\n",
      "\tI18: 0.826\n",
      "\tI19: 0.074\n",
      "\tI20: 0.074\n",
      "\tI24: 0.801\n",
      "\tI37: 0.118\n",
      "\n",
      "Klebsiella_pneumoniae_54788 | olm | Subject: N1_023\n",
      "Gene id: 1328400.3.peg.232 | Variant type: 1D\n",
      "Gene name: No name\n",
      "\tI15: 0.026\n",
      "\tI17: 0.897\n",
      "\tI18: 0.818\n",
      "\tI19: 0.088\n",
      "\tI21: 0.911\n",
      "\tI29: 0.172\n",
      "\n",
      "Citrobacter_freundii_56148 | olm | Subject: N1_009\n",
      "Gene id: 1114920.3.peg.2929 | Variant type: 4D\n",
      "Gene name: Phosphoenolpyruvate-dihydroxyacetone phosphotransferase operon regulatory protein DhaR\n",
      "\tI10: 0.870\n",
      "\tI13: 0.164\n",
      "\tI15: 0.091\n",
      "\tI16: 0.191\n",
      "\tI17: 0.944\n",
      "\n",
      "Bacteroides_vulgatus_57955 | olm | Subject: N4_097\n",
      "Gene id: 435590.9.peg.1561 | Variant type: 4D\n",
      "Gene name: Uncharacterized protein BT3327\n",
      "\tI12: 0.112\n",
      "\tI27: 0.826\n",
      "\tI34: 0.115\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M1098-M\n",
      "Gene id: 435590.9.peg.52 | Variant type: 1D\n",
      "Gene name: DNA internalization-related competence protein ComEC/Rec2\n",
      "\tM1: 1.000\n",
      "\tI2: 0.176\n",
      "\tI3: 0.176\n",
      "\tI4: 1.000\n",
      "\tI5: 0.962\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M1098-M\n",
      "Gene id: 435590.9.peg.543 | Variant type: 2D\n",
      "Gene name: Translation elongation factor P\n",
      "\tM1: 0.979\n",
      "\tI2: 0.154\n",
      "\tI3: 0.154\n",
      "\tI4: 1.000\n",
      "\tI5: 1.000\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M0808-M\n",
      "Gene id: 435590.9.peg.47 | Variant type: 4D\n",
      "Gene name: Keratan-sulfate endo-1,4-beta-galactosidase (EC 3.2.1.103)\n",
      "\tM1: 0.000\n",
      "\tM2: 0.885\n",
      "\tM3: 0.885\n",
      "\tI4: 0.094\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M0808-M\n",
      "Gene id: 435590.9.peg.1738 | Variant type: 1D\n",
      "Gene name: Transcriptional regulator, AraC family\n",
      "\tM1: 0.000\n",
      "\tM2: 0.941\n",
      "\tM3: 0.941\n",
      "\tI4: 0.020\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M0808-M\n",
      "Gene id: 435590.9.peg.2142 | Variant type: 2D\n",
      "Gene name: ATP-dependent DNA helicase RecS (RecQ family)\n",
      "\tM1: 0.011\n",
      "\tM2: 0.964\n",
      "\tM3: 0.964\n",
      "\tI4: 0.022\n",
      "\n",
      "Bacteroides_vulgatus_57955 | yassour | Subject: M0808-M\n",
      "Gene id: 435590.9.peg.2509 | Variant type: 4D\n",
      "Gene name: Outer membrane TonB-dependent transporter, utilization system for glycans and polysaccharides (PUL), SusC family\n",
      "\tM1: 0.000\n",
      "\tM2: 0.919\n",
      "\tM3: 0.919\n",
      "\tI4: 0.000\n",
      "\n",
      "Enterobacter_aerogenes_55704 | olm | Subject: N1_011\n",
      "Gene id: 1400143.3.peg.1706 | Variant type: 2D\n",
      "Gene name: Anaerobic dimethyl sulfoxide reductase chain B (EC 1.8.5.3), iron-sulfur binding subunit\n",
      "\tI11: 0.164\n",
      "\tI13: 0.958\n",
      "\tI15: 0.146\n",
      "\tI20: 0.976\n",
      "\tI21: 0.095\n",
      "\tI23: 0.980\n",
      "\tI33: 0.877\n",
      "\tI37: 0.936\n",
      "\tI46: 0.097\n",
      "\tI62: 0.140\n",
      "\tI69: 0.859\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Formatted printing of reversions\n",
    "for species, cohort, subject, site, gene_id, gene_name, var_type, ordered_tp_freqs in reversions:\n",
    "    \n",
    "    print(species + ' | ' + cohort + ' | Subject: ' + subject)\n",
    "    print(\"Gene id: \" + gene_id + ' | Variant type: ' + var_type)\n",
    "    print(\"Gene name: \" + gene_name)\n",
    "    for tp, freq in ordered_tp_freqs:\n",
    "        print('\\t' + tp + ': %.03f' % freq)\n",
    "    print('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
