{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading sample metadata...\n",
      "Done!\n"
     ]
    }
   ],
   "source": [
    "from utils import config, parse_midas_data, sample_utils as su, temporal_changes_utils, stats_utils, midas_db_utils, parse_patric\n",
    "from collections import defaultdict\n",
    "import numpy as np\n",
    "import pickle\n",
    "import sys\n",
    "import random\n",
    "from math import log10,ceil,log,exp\n",
    "\n",
    "import matplotlib.cm as cmx\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.gridspec as gridspec\n",
    "import matplotlib.colors as mcolors\n",
    "import matplotlib.patheffects as pe\n",
    "\n",
    "# Cohort list\n",
    "cohorts = ['backhed', 'ferretti', 'yassour', 'shao', 'olm', 'hmp']\n",
    "\n",
    "# Plot directory\n",
    "plot_dir = \"%s/\" % (config.analysis_directory)\n",
    "\n",
    "# Species list\n",
    "good_species_list = parse_midas_data.load_pickled_good_species_list()\n",
    "\n",
    "# Sample-subject-order maps\n",
    "sys.stderr.write(\"Loading sample metadata...\\n\")\n",
    "subject_sample_map = su.parse_subject_sample_map()\n",
    "sample_order_map = su.parse_sample_order_map()\n",
    "sample_subject_map = su.parse_sample_subject_map()\n",
    "sys.stderr.write(\"Done!\\n\")\n",
    "\n",
    "# Timepoint pair types\n",
    "tp_pair_names = ['MM', 'MI', 'II', 'AA']\n",
    "\n",
    "# Cohorts\n",
    "cohorts = ['backhed', 'ferretti', 'yassour', 'shao', 'hmp']\n",
    "mi_cohorts = ['backhed', 'ferretti', 'yassour', 'shao']\n",
    "\n",
    "# Samples for each cohort\n",
    "samples = {cohort: su.get_sample_names(cohort) for cohort in cohorts}\n",
    "hmp_samples = su.get_sample_names('hmp')\n",
    "mother_samples = su.get_sample_names('mother')\n",
    "infant_samples = su.get_sample_names('infant')\n",
    "olm_samples = su.get_sample_names('olm')\n",
    "infant_samples_no_olm = [sample for sample in infant_samples if sample not in olm_samples]\n",
    "mi_samples_no_olm = [sample for sample in (mother_samples + infant_samples) if sample not in olm_samples]\n",
    "\n",
    "# Sample-cohort map\n",
    "sample_cohort_map = su.parse_sample_cohort_map()\n",
    "\n",
    "# Sample-timepoint map\n",
    "mi_sample_day_dict = su.get_mi_sample_day_dict(exclude_cohorts=['olm'])\n",
    "mi_tp_sample_dict = su.get_mi_tp_sample_dict(exclude_cohorts=['olm']) # no binning\n",
    "mi_tp_sample_dict_binned, mi_tp_binned_labels = su.get_mi_tp_sample_dict(exclude_cohorts=['olm'], binned=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# ======================================================================\n",
    "# Load pickled data\n",
    "# ======================================================================\n",
    "\n",
    "# Parameters\n",
    "sweep_type = 'full' # assume full for now\n",
    "pp_prev_cohort = 'all'\n",
    "min_coverage = 0\n",
    "\n",
    "ddir = config.data_directory\n",
    "pdir = \"%s/pickles/cov%i_prev_%s/nonconsecutive\" % (ddir, min_coverage, pp_prev_cohort)\n",
    "\n",
    "snp_changes = pickle.load(open('%s/big_snp_changes_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "gene_changes = pickle.load(open('%s/big_gene_changes_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "snp_change_freqs = pickle.load(open('%s/snp_change_freqs_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "snp_change_freqs_with_opps = pickle.load(open('%s/snp_change_freqs_with_opps_full.pkl' % (pdir), 'rb'))\n",
    "snp_change_null_freqs = pickle.load(open('%s/snp_change_null_freqs_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "gene_gain_freqs = pickle.load(open('%s/gene_gain_freqs_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "gene_loss_freqs = pickle.load(open('%s/gene_loss_freqs_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "gene_loss_null_freqs = pickle.load(open('%s/gene_loss_null_freqs_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "between_snp_change_counts = pickle.load(open('%s/between_snp_change_counts_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "between_gene_change_counts = pickle.load(open('%s/between_gene_change_counts_%s.pkl' % (pdir, sweep_type), 'rb'))\n",
    "snp_change_present_gene_null = pickle.load(open('%s/snp_change_present_gene_null' % pdir, 'rb'))\n",
    "snp_change_between_host_null = pickle.load(open('%s/snp_change_between_host_null' % pdir, 'rb'))\n",
    "snp_change_pangenome_null = pickle.load(open('%s/snp_change_pangenome_null' % pdir, 'rb'))\n",
    "dnds_info = pickle.load(open('%s/dnds_info.pkl' % (pdir), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Consecutive data\n",
    "\n",
    "# Parameters\n",
    "sweep_type = 'full' # assume full for now\n",
    "pp_prev_cohort = 'all'\n",
    "min_coverage = 0\n",
    "\n",
    "ddir = config.data_directory\n",
    "pdir = \"%s/pickles/cov%i_prev_%s\" % (ddir, min_coverage, pp_prev_cohort)\n",
    "\n",
    "snp_changes_consecutive = pickle.load(open('%s/big_snp_changes_%s.pkl' % (pdir, sweep_type), 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Calculate number of days for a timpeoint pair\n",
    "\n",
    "mi_sample_day_dict = su.get_mi_sample_day_dict()\n",
    "\n",
    "def sample_pair_to_days(sample1, sample2):\n",
    "    days = mi_sample_day_dict[sample2] - mi_sample_day_dict[sample1]\n",
    "    if days < 0:\n",
    "        return np.abs(days)\n",
    "    return days\n",
    "\n",
    "def tp_pair_to_days(tp_pair):\n",
    "    tpa, tpb = tp_pair\n",
    "    o1 = float(tpa[1:])\n",
    "    o2 = float(tpb[1:])\n",
    "    return np.abs(o1-o2)\n",
    "\n",
    "# Rough approximation of HMP time intervals\n",
    "def adult_tp_pair_to_days(tp_pair):\n",
    "    tpa, tpb = tp_pair\n",
    "    return np.abs(int(tpa[-1:]) - int(tpb[-1:]))*183\n",
    "\n",
    "def tp_pair_to_tp_type(tp_pair):\n",
    "    tpa, tpb = tp_pair\n",
    "    tp_type = tpa[0] + tpb[0]\n",
    "    if tp_type == 'IM':\n",
    "        tp_type = 'MI'\n",
    "    return tp_type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Settings for prevalence plots\n",
    "\n",
    "modification_difference_threshold = config.modification_difference_threshold\n",
    "replacement_difference_threshold = config.replacement_difference_threshold\n",
    "default_num_bootstraps = 10000\n",
    "min_sample_size = 3\n",
    "min_haploid_sample_size = 10\n",
    "\n",
    "variant_types = ['1D','4D']\n",
    "within_host_type = 'consecutive' # consecutive timepoints (vs. longest)\n",
    "\n",
    "num_bootstraps = 10 # for gene change prevalence null\n",
    "\n",
    "# For partitioning SNVs according to prevalence\n",
    "derived_freq_bins = np.array([-1,0,0.01,0.1,0.5,0.9,0.99,1,2])\n",
    "derived_virtual_freqs = np.arange(0,len(derived_freq_bins)-1)\n",
    "derived_virtual_xticks = list(derived_virtual_freqs[:-1]+0.5)\n",
    "derived_virtual_xticklabels = ['0','.01','.1','.5','.9','.99','1']\n",
    "\n",
    "# For partitioning genes into different prevalence classes\n",
    "gene_freq_bins = np.array([-1,0.1,0.5,0.9,2])\n",
    "gene_freq_xticks = [-4, -3, -2, -1, 0, 1, 2, 3, 4]\n",
    "gene_freq_xticklabels = ['0','0.1','0.5', '0.9','1','0.9','0.5', '0.1','0']\n",
    "gene_gain_virtual_freqs = np.array([3.5,2.5,1.5,0.5])\n",
    "gene_loss_virtual_freqs = np.array([-3.5,-2.5,-1.5,-0.5])\n",
    "\n",
    "# Function\n",
    "def get_f_idx(f):\n",
    "    return ((f>derived_freq_bins[:-1])*\n",
    "            (f<=derived_freq_bins[1:])).argmax()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Do suspected de novo mutations in infants revert within hosts?\n",
    "\n",
    "# subject -> species -> (day1, day2) -> list of SNP change tuples OR 'replacement'\n",
    "host_snp_changes = defaultdict(dict)\n",
    "# subject -> species -> (day1, day2) -> list of (vartype, fdict) tuples OR 'replacement\n",
    "host_snp_changes_freqs = defaultdict(dict)\n",
    "\n",
    "for species in snp_changes:\n",
    "    for sample_i, sample_j in snp_changes[species]:\n",
    "        \n",
    "        tp_pair = su.sample_pair_to_tp_pair(sample_i, sample_j, sample_order_map, hmp_samples, mother_samples)\n",
    "        tp_type = tp_pair_to_tp_type(tp_pair)\n",
    "        if tp_type != 'MI' and tp_type != 'II': # Only look at mother-infant or infant-infant\n",
    "            continue\n",
    "        \n",
    "        subject = sample_subject_map[sample_j][:-2] # Combine mother and infant subjects!\n",
    "        day1, day2 = mi_sample_day_dict[sample_i], mi_sample_day_dict[sample_j]\n",
    "        \n",
    "        pdicts = snp_change_freqs_with_opps[species][(sample_i, sample_j)] # list of (vartype, freq_dict, opp_dict tuples)\n",
    "        npdict = snp_change_null_freqs[species][(sample_i, sample_j)] # dict: prev_cohort > list of (freq, weight) tuples\n",
    "        \n",
    "        snp_change_val = snp_changes[species][(sample_i, sample_j)]\n",
    "        \n",
    "        nonsyn_diffs, nonsyn_opps, syn_diffs, syn_opps = dnds_info[species][(sample_i, sample_j)]\n",
    "        \n",
    "        if type(snp_change_val) == type(1): # Replacement\n",
    "            try:\n",
    "                host_snp_changes[subject][species][(day1, day2, tp_type)] = 'replacement'\n",
    "                host_snp_changes_freqs[subject][species][(day1, day2, tp_type)] = 'replacement'\n",
    "            except:\n",
    "                host_snp_changes[subject][species] = {(day1, day2, tp_type): 'replacement'}\n",
    "                host_snp_changes_freqs[subject][species] = {(day1, day2, tp_type): 'replacement'}\n",
    "        else:\n",
    "            try:\n",
    "                host_snp_changes[subject][species][(day1, day2, tp_type)] = snp_changes[species][(sample_i, sample_j)]\n",
    "                host_snp_changes_freqs[subject][species][(day1, day2, tp_type)] = pdicts\n",
    "            except:\n",
    "                host_snp_changes[subject][species] = {(day1, day2, tp_type): snp_changes[species][(sample_i, sample_j)]}\n",
    "                host_snp_changes_freqs[subject][species] = {(day1, day2, tp_type): pdicts}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "num_host_species_pairs = 0\n",
    "desired_host_species_sites = []\n",
    "\n",
    "for subject in host_snp_changes:\n",
    "    for species in host_snp_changes[subject]:        \n",
    "        if len(host_snp_changes[subject][species].keys()) > 1:\n",
    "            num_host_species_pairs += 1\n",
    "            print(\"Subject: %s, Species: %s\" % (subject, species))\n",
    "            sites = []\n",
    "            for day1, day2, tp_type in sorted(host_snp_changes[subject][species].keys(), key=lambda x: x[0]):\n",
    "                print('\\n%s Day %i to day %i' % (tp_type, day1, day2))\n",
    "                snp_change_list = host_snp_changes[subject][species][(day1, day2, tp_type)]\n",
    "                snp_change_freqs_list = host_snp_changes_freqs[subject][species][(day1, day2, tp_type)]\n",
    "                for snp_change, x_snp_change_freqs in zip(snp_change_list, snp_change_freqs_list):\n",
    "                    gene_id, contig, location, vartype, a1, d1, a2, d2 = snp_change\n",
    "                    vartype2, freq_dict, opp_dict = x_snp_change_freqs\n",
    "                    if vartype != vartype2:\n",
    "                        print(\"Weird\")\n",
    "                    sites.append((gene_id, contig, location))\n",
    "                    print(\"Site: %s,%s,%s\\t| Vartype: %s\" % (gene_id, contig, location, vartype))\n",
    "                    print(\"\\t%i %i %i %i | %.02f -> %.02f\" % (a1, d1, a2, d2, float(a1)/d1, float(a2)/d2))\n",
    "                    print(\"\\tInfant prev: %.02f | HMP prev: %.02f | Mother prev: %.02f\" % (freq_dict['nonpremie'], freq_dict['hmp'], freq_dict['mother']))\n",
    "            desired_host_species_sites.append(([subject], species, sites))\n",
    "            print(\"\\n=============================================================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subject: M0305, Species: Bacteroides_uniformis_57318\n",
      "Mother: [0, 92]\n",
      "Infant: [14, 30]\n",
      "Pairs: [(0, 30, 'MI'), (14, 30, 'II'), (92, 30, 'MI'), (92, 14, 'MI'), (0, 14, 'MI')]\n",
      "\n",
      "Subject: A01805, Species: Bifidobacterium_adolescentis_56815\n",
      "Mother: [1]\n",
      "Infant: [4]\n",
      "Pairs: [(1, 4, 'MI')]\n",
      "\n",
      "Subject: C01758, Species: Escherichia_coli_58110\n",
      "Mother: [1]\n",
      "Infant: [21, 311]\n",
      "Pairs: [(1, 21, 'MI'), (21, 311, 'II'), (1, 311, 'MI')]\n",
      "\n",
      "Subject: C01758, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [1]\n",
      "Infant: [311]\n",
      "Pairs: [(1, 311, 'MI')]\n",
      "\n",
      "Subject: C01930, Species: Enterococcus_faecium_56947\n",
      "Mother: [1]\n",
      "Infant: [7]\n",
      "Pairs: [(1, 7, 'MI')]\n",
      "\n",
      "Subject: C01930, Species: Bacteroides_uniformis_57318\n",
      "Mother: [1]\n",
      "Infant: [271]\n",
      "Pairs: [(1, 271, 'MI')]\n",
      "\n",
      "Subject: 100, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 100, Species: Blautia_wexlerae_56130\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 39, Species: Bacteroides_caccae_53434\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 39, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 33, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 35, Species: Eubacterium_rectale_56927\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 35, Species: Roseburia_intestinalis_56239\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 35, Species: Bacteroides_eggerthii_54457\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: B01771, Species: Bacteroides_uniformis_57318\n",
      "Mother: [1]\n",
      "Infant: [7]\n",
      "Pairs: [(1, 7, 'MI')]\n",
      "\n",
      "Subject: CA_C10055, Species: Alistipes_putredinis_61533\n",
      "Mother: [1]\n",
      "Infant: [0]\n",
      "Pairs: [(1, 0, 'MI')]\n",
      "\n",
      "Subject: 98, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 98, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 122, 'MI'), (122, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 91, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 90, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 93, Species: Alistipes_finegoldii_56071\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 93, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 93, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 92, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 94, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 94, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 94, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 122, 'MI'), (122, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 97, Species: Prevotella_copri_61740\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 97, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 96, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: B01194, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 21]\n",
      "Pairs: [(1, 4, 'MI'), (7, 21, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (1, 21, 'MI')]\n",
      "\n",
      "Subject: B01194, Species: Bifidobacterium_catenulatum_58257\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 21]\n",
      "Pairs: [(1, 4, 'MI'), (7, 21, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (1, 21, 'MI')]\n",
      "\n",
      "Subject: 604136, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [1]\n",
      "Infant: [7, 10, 11, 14]\n",
      "Pairs: [(10, 11, 'II'), (7, 11, 'II'), (10, 14, 'II'), (11, 14, 'II'), (1, 11, 'MI'), (7, 10, 'II'), (7, 14, 'II'), (1, 7, 'MI'), (1, 10, 'MI'), (1, 14, 'MI')]\n",
      "\n",
      "Subject: C02143, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [1]\n",
      "Infant: [7, 237]\n",
      "Pairs: [(1, 237, 'MI'), (1, 7, 'MI'), (7, 237, 'II')]\n",
      "\n",
      "Subject: B00139, Species: Bifidobacterium_longum_57796\n",
      "Mother: [1]\n",
      "Infant: [21]\n",
      "Pairs: [(1, 21, 'MI')]\n",
      "\n",
      "Subject: C01840, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 21, 289]\n",
      "Pairs: [(1, 4, 'MI'), (7, 21, 'II'), (1, 289, 'MI'), (7, 289, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (21, 289, 'II'), (1, 21, 'MI'), (4, 289, 'II')]\n",
      "\n",
      "Subject: 27, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 20, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 20, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(2, 3, 'MI'), (122, 366, 'II'), (3, 366, 'II'), (2, 366, 'MI'), (3, 122, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 28, Species: Alistipes_finegoldii_56071\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 29, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 29, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: CA_C10009, Species: Dialister_invisus_61905\n",
      "Mother: [30]\n",
      "Infant: [0]\n",
      "Pairs: [(30, 0, 'MI')]\n",
      "\n",
      "Subject: CA_C10009, Species: Alistipes_putredinis_61533\n",
      "Mother: [30]\n",
      "Infant: [0]\n",
      "Pairs: [(30, 0, 'MI')]\n",
      "\n",
      "Subject: CA_C10007, Species: Bacteroides_ovatus_58035\n",
      "Mother: [0]\n",
      "Infant: [122]\n",
      "Pairs: [(0, 122, 'MI')]\n",
      "\n",
      "Subject: 88, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [3, 366]\n",
      "Pairs: [(2, 3, 'MI'), (3, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 88, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 88, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: A02110, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [1]\n",
      "Infant: [211]\n",
      "Pairs: [(1, 211, 'MI')]\n",
      "\n",
      "Subject: 82, Species: Bacteroides_thetaiotaomicron_56941\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 82, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 82, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 80, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 81, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 81, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(122, 366, 'II'), (2, 122, 'MI'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 87, Species: Akkermansia_muciniphila_55290\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 87, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: M1098, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [-92, 0]\n",
      "Infant: [0, 14, 30, 61, 92]\n",
      "Pairs: [(0, 14, 'MI'), (0, 92, 'MI'), (14, 61, 'II'), (14, 30, 'II'), (-92, 14, 'MI'), (0, 14, 'II'), (0, 61, 'MI'), (30, 92, 'II'), (0, 0, 'MI'), (61, 92, 'II'), (0, 30, 'MI'), (30, 61, 'II'), (-92, 61, 'MI'), (-92, 92, 'MI'), (0, 92, 'II'), (0, 30, 'II'), (-92, 0, 'MI'), (14, 92, 'II'), (0, 61, 'II'), (-92, 30, 'MI')]\n",
      "\n",
      "Subject: 7, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 7, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: B02243, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [1]\n",
      "Infant: [221]\n",
      "Pairs: [(1, 221, 'MI')]\n",
      "\n",
      "Subject: B00040, Species: Escherichia_coli_58110\n",
      "Mother: [1]\n",
      "Infant: [4]\n",
      "Pairs: [(1, 4, 'MI')]\n",
      "\n",
      "Subject: M0074, Species: Bacteroides_stercoris_56735\n",
      "Mother: [92]\n",
      "Infant: [0]\n",
      "Pairs: [(92, 0, 'MI')]\n",
      "\n",
      "Subject: M0074, Species: Bacteroides_uniformis_57318\n",
      "Mother: [-92, 0, 92]\n",
      "Infant: [14]\n",
      "Pairs: [(0, 14, 'MI'), (92, 14, 'MI'), (-92, 14, 'MI')]\n",
      "\n",
      "Subject: M0074, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [-92, 92]\n",
      "Infant: [0]\n",
      "Pairs: [(92, 0, 'MI'), (-92, 0, 'MI')]\n",
      "\n",
      "Subject: M0073, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [0]\n",
      "Infant: [14, 30, 61, 92]\n",
      "Pairs: [(0, 14, 'MI'), (14, 61, 'II'), (0, 61, 'MI'), (30, 92, 'II'), (14, 30, 'II'), (61, 92, 'II'), (0, 30, 'MI'), (30, 61, 'II'), (0, 92, 'MI'), (14, 92, 'II')]\n",
      "\n",
      "Subject: C01075, Species: Prevotella_copri_61740\n",
      "Mother: [1]\n",
      "Infant: [4]\n",
      "Pairs: [(1, 4, 'MI')]\n",
      "\n",
      "Subject: 13, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 13, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 13, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 13, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 15, Species: Ruminococcus_bromii_62047\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 15, Species: Bacteroides_caccae_53434\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 14, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 19, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 19, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(2, 3, 'MI'), (122, 366, 'II'), (3, 366, 'II'), (2, 366, 'MI'), (3, 122, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 19, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 18, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 18, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 18, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 18, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 18, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: M0327, Species: Bacteroides_caccae_53434\n",
      "Mother: [0]\n",
      "Infant: [30, 92]\n",
      "Pairs: [(0, 30, 'MI'), (0, 92, 'MI'), (30, 92, 'II')]\n",
      "\n",
      "Subject: M0450, Species: Bifidobacterium_adolescentis_56815\n",
      "Mother: [92]\n",
      "Infant: [61]\n",
      "Pairs: [(92, 61, 'MI')]\n",
      "\n",
      "Subject: C01696, Species: Citrobacter_freundii_56148\n",
      "Mother: [1]\n",
      "Infant: [7]\n",
      "Pairs: [(1, 7, 'MI')]\n",
      "\n",
      "Subject: B01234, Species: Bacteroides_uniformis_57318\n",
      "Mother: [1]\n",
      "Infant: [383]\n",
      "Pairs: [(1, 383, 'MI')]\n",
      "\n",
      "Subject: C01917, Species: Faecalibacterium_cf_62236\n",
      "Mother: [1]\n",
      "Infant: [311]\n",
      "Pairs: [(1, 311, 'MI')]\n",
      "\n",
      "Subject: C01917, Species: Escherichia_coli_58110\n",
      "Mother: [1]\n",
      "Infant: [4, 7]\n",
      "Pairs: [(1, 4, 'MI'), (4, 7, 'II'), (1, 7, 'MI')]\n",
      "\n",
      "Subject: 75, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 75, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 75, Species: Bifidobacterium_longum_57796\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 75, Species: Bifidobacterium_pseudocatenulatum_57754\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 74, Species: Bacteroides_eggerthii_54457\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 73, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 11, Species: Bifidobacterium_longum_57796\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 11, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(2, 3, 'MI'), (122, 366, 'II'), (3, 366, 'II'), (2, 366, 'MI'), (3, 122, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 10, Species: Bacteroides_xylanisolvens_57185\n",
      "Mother: [2]\n",
      "Infant: [3, 366]\n",
      "Pairs: [(2, 3, 'MI'), (3, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 12, Species: Alistipes_putredinis_61533\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: M0369, Species: Akkermansia_muciniphila_55290\n",
      "Mother: [-92]\n",
      "Infant: [61, 92]\n",
      "Pairs: [(61, 92, 'II'), (-92, 61, 'MI'), (-92, 92, 'MI')]\n",
      "\n",
      "Subject: BBS0067, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [1]\n",
      "Infant: [7]\n",
      "Pairs: [(1, 7, 'MI')]\n",
      "\n",
      "Subject: 3, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: CA_C10039, Species: Bacteroides_caccae_53434\n",
      "Mother: [0]\n",
      "Infant: [7]\n",
      "Pairs: [(0, 7, 'MI')]\n",
      "\n",
      "Subject: CA_C10035, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [3]\n",
      "Infant: [0]\n",
      "Pairs: [(3, 0, 'MI')]\n",
      "\n",
      "Subject: M0764, Species: Bacteroides_uniformis_57318\n",
      "Mother: [0]\n",
      "Infant: [30, 61, 92]\n",
      "Pairs: [(0, 61, 'MI'), (30, 92, 'II'), (61, 92, 'II'), (0, 30, 'MI'), (30, 61, 'II'), (0, 92, 'MI')]\n",
      "\n",
      "Subject: M0402, Species: Bifidobacterium_pseudocatenulatum_57754\n",
      "Mother: [92]\n",
      "Infant: [92]\n",
      "Pairs: [(92, 92, 'MI')]\n",
      "\n",
      "Subject: M0402, Species: Alistipes_onderdonkii_55464\n",
      "Mother: [-92, 92]\n",
      "Infant: [14]\n",
      "Pairs: [(-92, 14, 'MI'), (92, 14, 'MI')]\n",
      "\n",
      "Subject: M0402, Species: Bacteroides_stercoris_56735\n",
      "Mother: [-92, 92]\n",
      "Infant: [14]\n",
      "Pairs: [(-92, 14, 'MI'), (92, 14, 'MI')]\n",
      "\n",
      "Subject: B01409, Species: Bifidobacterium_longum_57796\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 21]\n",
      "Pairs: [(1, 4, 'MI'), (7, 21, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (1, 21, 'MI')]\n",
      "\n",
      "Subject: 61, Species: Bacteroides_stercoris_56735\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 122, 'MI'), (122, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 61, Species: Parabacteroides_merdae_56972\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 61, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 122, 'MI'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 67, Species: Bifidobacterium_longum_57796\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 68, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 68, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(122, 366, 'II'), (2, 122, 'MI'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 68, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 68, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3, 366]\n",
      "Pairs: [(2, 3, 'MI'), (3, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 69, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [3, 366]\n",
      "Pairs: [(2, 3, 'MI'), (3, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 515126, Species: Bacteroides_plebeius_61623\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 10, 11, 21]\n",
      "Pairs: [(10, 21, 'II'), (10, 11, 'II'), (7, 21, 'II'), (4, 11, 'II'), (7, 11, 'II'), (1, 11, 'MI'), (7, 10, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (1, 10, 'MI'), (1, 21, 'MI'), (1, 4, 'MI'), (4, 10, 'II'), (11, 21, 'II')]\n",
      "\n",
      "Subject: M0201, Species: Bifidobacterium_longum_57796\n",
      "Mother: [-92]\n",
      "Infant: [14, 61, 92]\n",
      "Pairs: [(14, 61, 'II'), (-92, 92, 'MI'), (61, 92, 'II'), (-92, 14, 'MI'), (14, 92, 'II'), (-92, 61, 'MI')]\n",
      "\n",
      "Subject: M0201, Species: Bacteroides_uniformis_57318\n",
      "Mother: [-92, 92]\n",
      "Infant: [61, 92]\n",
      "Pairs: [(61, 92, 'II'), (92, 92, 'MI'), (92, 61, 'MI'), (-92, 61, 'MI'), (-92, 92, 'MI')]\n",
      "\n",
      "Subject: 34, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: M0297, Species: Escherichia_coli_58110\n",
      "Mother: [0]\n",
      "Infant: [0, 14, 61, 92]\n",
      "Pairs: [(0, 14, 'MI'), (14, 61, 'II'), (0, 61, 'MI'), (0, 0, 'MI'), (61, 92, 'II'), (0, 14, 'II'), (0, 92, 'II'), (0, 92, 'MI'), (14, 92, 'II'), (0, 61, 'II')]\n",
      "\n",
      "Subject: 1, Species: Ruminococcus_bicirculans_59300\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 1, Species: Alistipes_onderdonkii_55464\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 1, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 1, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(2, 3, 'MI'), (122, 366, 'II'), (3, 366, 'II'), (2, 366, 'MI'), (3, 122, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: C01994, Species: Bacteroides_fragilis_54507\n",
      "Mother: [1]\n",
      "Infant: [4, 21, 285]\n",
      "Pairs: [(21, 285, 'II'), (4, 285, 'II'), (1, 4, 'MI'), (1, 285, 'MI'), (4, 21, 'II'), (1, 21, 'MI')]\n",
      "\n",
      "Subject: 4, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: A00942, Species: Bacteroides_massiliensis_44749\n",
      "Mother: [1]\n",
      "Infant: [4]\n",
      "Pairs: [(1, 4, 'MI')]\n",
      "\n",
      "Subject: 59, Species: Bifidobacterium_adolescentis_56815\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(2, 3, 'MI'), (122, 366, 'II'), (3, 366, 'II'), (2, 366, 'MI'), (3, 122, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 59, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 58, Species: Bacteroides_caccae_53434\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 55, Species: Bifidobacterium_bifidum_55065\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 54, Species: Bacteroides_faecis_58503\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(122, 366, 'II'), (2, 122, 'MI'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 57, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 56, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 51, Species: Ruminococcus_bromii_62047\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 50, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 53, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 53, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 53, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 53, Species: Bacteroides_caccae_53434\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: 53, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(122, 366, 'II'), (2, 122, 'MI'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 52, Species: Escherichia_coli_58110\n",
      "Mother: [2]\n",
      "Infant: [3]\n",
      "Pairs: [(2, 3, 'MI')]\n",
      "\n",
      "Subject: B01948, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [1]\n",
      "Infant: [7]\n",
      "Pairs: [(1, 7, 'MI')]\n",
      "\n",
      "Subject: M0808, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [-92, 0, 92]\n",
      "Infant: [14, 61]\n",
      "Pairs: [(0, 14, 'MI'), (14, 61, 'II'), (92, 61, 'MI'), (0, 61, 'MI'), (-92, 14, 'MI'), (92, 14, 'MI'), (-92, 61, 'MI')]\n",
      "\n",
      "Subject: BBS0014, Species: Bacteroides_uniformis_57318\n",
      "Mother: [1]\n",
      "Infant: [11]\n",
      "Pairs: [(1, 11, 'MI')]\n",
      "\n",
      "Subject: M0792, Species: Parabacteroides_merdae_56972\n",
      "Mother: [92]\n",
      "Infant: [61]\n",
      "Pairs: [(92, 61, 'MI')]\n",
      "\n",
      "Subject: M0792, Species: Parabacteroides_distasonis_56985\n",
      "Mother: [92]\n",
      "Infant: [14, 30, 61, 92]\n",
      "Pairs: [(14, 61, 'II'), (92, 61, 'MI'), (30, 92, 'II'), (92, 14, 'MI'), (61, 92, 'II'), (30, 61, 'II'), (92, 92, 'MI'), (14, 30, 'II'), (14, 92, 'II'), (92, 30, 'MI')]\n",
      "\n",
      "Subject: M0792, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [-92]\n",
      "Infant: [92]\n",
      "Pairs: [(-92, 92, 'MI')]\n",
      "\n",
      "Subject: CA_C10023, Species: Bacteroides_uniformis_57318\n",
      "Mother: [0]\n",
      "Infant: [30]\n",
      "Pairs: [(0, 30, 'MI')]\n",
      "\n",
      "Subject: CA_C10029, Species: Bacteroides_plebeius_61623\n",
      "Mother: [0]\n",
      "Infant: [7]\n",
      "Pairs: [(0, 7, 'MI')]\n",
      "\n",
      "Subject: CA_C10029, Species: Bacteroides_uniformis_57318\n",
      "Mother: [3]\n",
      "Infant: [0]\n",
      "Pairs: [(3, 0, 'MI')]\n",
      "\n",
      "Subject: 78, Species: Bacteroides_ovatus_58035\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 78, Species: Sutterella_wadsworthensis_56828\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 122, 'MI'), (122, 366, 'II'), (2, 366, 'MI')]\n",
      "\n",
      "Subject: 78, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: B01261, Species: Bifidobacterium_longum_57796\n",
      "Mother: [1]\n",
      "Infant: [4, 7, 21]\n",
      "Pairs: [(1, 4, 'MI'), (7, 21, 'II'), (4, 7, 'II'), (4, 21, 'II'), (1, 7, 'MI'), (1, 21, 'MI')]\n",
      "\n",
      "Subject: C01960, Species: Bifidobacterium_longum_57796\n",
      "Mother: [1]\n",
      "Infant: [266]\n",
      "Pairs: [(1, 266, 'MI')]\n",
      "\n",
      "Subject: 48, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 48, Species: Eubacterium_eligens_61678\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 48, Species: Bacteroides_cellulosilyticus_58046\n",
      "Mother: [2]\n",
      "Infant: [366]\n",
      "Pairs: [(2, 366, 'MI')]\n",
      "\n",
      "Subject: 48, Species: Bacteroides_vulgatus_57955\n",
      "Mother: [2]\n",
      "Infant: [3, 122]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 49, Species: Bacteroides_uniformis_57318\n",
      "Mother: [2]\n",
      "Infant: [122, 366]\n",
      "Pairs: [(2, 366, 'MI'), (122, 366, 'II'), (2, 122, 'MI')]\n",
      "\n",
      "Subject: 44, Species: Bacteroides_caccae_53434\n",
      "Mother: [2]\n",
      "Infant: [122]\n",
      "Pairs: [(2, 122, 'MI')]\n",
      "\n",
      "Subject: 5, Species: Bacteroides_fragilis_54507\n",
      "Mother: [2]\n",
      "Infant: [3, 122, 366]\n",
      "Pairs: [(3, 122, 'II'), (2, 3, 'MI'), (2, 122, 'MI'), (3, 366, 'II'), (2, 366, 'MI')]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Get a sense of available timepoints per host\n",
    "# so I filter out MI appropriately\n",
    "# Wait... should I not be restricting to modifications?\n",
    "# Including any nonconsecutive timepoint pair, calculation seems correct\n",
    "\n",
    "num_only_infant = 0\n",
    "num_have_mother = 0\n",
    "\n",
    "for subject in host_snp_changes:\n",
    "    for species in host_snp_changes[subject]:\n",
    "        m_days = set()\n",
    "        i_days = set()\n",
    "        for day1, day2, tp_type in sorted(host_snp_changes[subject][species].keys(), key=lambda x: x[0]):\n",
    "            if tp_type == 'MI':\n",
    "                m_days.add(day1)\n",
    "                i_days.add(day2)\n",
    "            elif tp_type == 'II':\n",
    "                i_days.add(day1)\n",
    "                i_days.add(day2)\n",
    "        \n",
    "        if len(m_days) > 0:\n",
    "            num_have_mother += 1\n",
    "            print(\"Subject: %s, Species: %s\" % (subject, species))\n",
    "            print(\"Mother: %s\" % str(sorted(list(m_days))))\n",
    "            print(\"Infant: %s\" % str(sorted(list(i_days))))\n",
    "            print(\"Pairs: %s\" % str(host_snp_changes[subject][species].keys()))\n",
    "            print('')\n",
    "        else:\n",
    "            num_only_infant += 1     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "145"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "num_only_infant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subject: 98, Species: Parabacteroides_distasonis_56985\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 658663.3.peg.45,ACUD01000001,46292\t| Vartype: 1D\n",
      "\t0 51 264 264 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.4287,ACUD01000012,72506\t| Vartype: 2D\n",
      "\t3 27 134 134 | 0.11 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.2184,ACUD01000004,144184\t| Vartype: 1D\n",
      "\t0 25 106 106 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 658663.3.peg.45,ACUD01000001,46292\t| Vartype: 1D\n",
      "\t0 51 47 47 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.2024,ACUD01000003,296362\t| Vartype: 1D\n",
      "\t0 49 54 61 | 0.00 -> 0.89\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.2311,ACUD01000004,294606\t| Vartype: 1D\n",
      "\t0 37 33 40 | 0.00 -> 0.82\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.4435,ACUD01000013,52774\t| Vartype: 1D\n",
      "\t0 54 45 56 | 0.00 -> 0.80\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.4287,ACUD01000012,72506\t| Vartype: 2D\n",
      "\t3 27 22 22 | 0.11 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.1731,ACUD01000002,1644232\t| Vartype: 1D\n",
      "\t0 29 18 21 | 0.00 -> 0.86\n",
      "\tInfant prev: 0.02 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.2184,ACUD01000004,144184\t| Vartype: 1D\n",
      "\t0 25 20 20 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.3032,ACUD01000005,59806\t| Vartype: 1D\n",
      "\t0 40 40 45 | 0.00 -> 0.89\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "II Day 122 to day 366\n",
      "Site: 658663.3.peg.2024,ACUD01000003,296362\t| Vartype: 1D\n",
      "\t0 284 54 61 | 0.00 -> 0.89\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.4435,ACUD01000013,52774\t| Vartype: 1D\n",
      "\t5 197 45 56 | 0.03 -> 0.80\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 658663.3.peg.3032,ACUD01000005,59806\t| Vartype: 1D\n",
      "\t0 239 40 45 | 0.00 -> 0.89\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "=============================================================================\n",
      "Subject: 94, Species: Bacteroides_vulgatus_57955\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 435590.9.peg.1463,NC_009614,1876991\t| Vartype: 2D\n",
      "\t33 252 400 400 | 0.13 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: 0.01 | Mother prev: 0.02\n",
      "Site: 435590.9.peg.2517,NC_009614,3131479\t| Vartype: 1D\n",
      "\t19 149 193 193 | 0.13 -> 1.00\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.546,NC_009614,749817\t| Vartype: 4D\n",
      "\t22 173 252 252 | 0.13 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1324,NC_009614,1712073\t| Vartype: 1D\n",
      "\t30 186 351 351 | 0.16 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.2465,NC_009614,3068361\t| Vartype: 1D\n",
      "\t40 257 485 485 | 0.16 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 435590.9.peg.1463,NC_009614,1876991\t| Vartype: 2D\n",
      "\t33 252 99 99 | 0.13 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: 0.01 | Mother prev: 0.02\n",
      "Site: 435590.9.peg.2517,NC_009614,3131479\t| Vartype: 1D\n",
      "\t19 149 58 58 | 0.13 -> 1.00\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.546,NC_009614,749817\t| Vartype: 4D\n",
      "\t22 173 72 72 | 0.13 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1324,NC_009614,1712073\t| Vartype: 1D\n",
      "\t30 186 87 87 | 0.16 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.3058,NC_009614,3760793\t| Vartype: 1D\n",
      "\t0 129 50 50 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.3988,NC_009614,4789805\t| Vartype: 2D\n",
      "\t5 82 20 20 | 0.06 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.2465,NC_009614,3068361\t| Vartype: 1D\n",
      "\t40 257 148 148 | 0.16 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "II Day 122 to day 366\n",
      "Site: 435590.9.peg.3058,NC_009614,3760793\t| Vartype: 1D\n",
      "\t0 195 50 50 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "=============================================================================\n",
      "Subject: C02143, Species: Bifidobacterium_bifidum_55065\n",
      "\n",
      "MI Day 1 to day 237\n",
      "Site: 500634.3.peg.1861,AWSW01000054,37945\t| Vartype: 1D\n",
      "\t0 28 127 145 | 0.00 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "MI Day 1 to day 7\n",
      "Site: 500634.3.peg.945,AWSW01000030,35925\t| Vartype: 4D\n",
      "\t0 40 50 56 | 0.00 -> 0.89\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 500634.3.peg.1636,AWSW01000046,21960\t| Vartype: 3D\n",
      "\t0 19 32 38 | 0.00 -> 0.84\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 500634.3.peg.875,AWSW01000027,7916\t| Vartype: 1D\n",
      "\t32 32 1 57 | 1.00 -> 0.02\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.97\n",
      "Site: 500634.3.peg.952,AWSW01000030,45351\t| Vartype: 1D\n",
      "\t34 34 4 79 | 1.00 -> 0.05\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.97\n",
      "Site: 500634.3.peg.1619,AWSW01000046,4226\t| Vartype: 1D\n",
      "\t33 33 2 53 | 1.00 -> 0.04\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.97\n",
      "\n",
      "II Day 7 to day 237\n",
      "Site: 500634.3.peg.875,AWSW01000027,7916\t| Vartype: 1D\n",
      "\t1 57 125 145 | 0.02 -> 0.86\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "Site: 500634.3.peg.1861,AWSW01000054,37945\t| Vartype: 1D\n",
      "\t1 50 127 145 | 0.02 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 500634.3.peg.952,AWSW01000030,45351\t| Vartype: 1D\n",
      "\t4 79 152 171 | 0.05 -> 0.89\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "Site: 500634.3.peg.1619,AWSW01000046,4226\t| Vartype: 1D\n",
      "\t2 53 115 137 | 0.04 -> 0.84\n",
      "\tInfant prev: 0.00 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "Site: 500634.3.peg.1636,AWSW01000046,21960\t| Vartype: 3D\n",
      "\t32 38 14 135 | 0.84 -> 0.10\n",
      "\tInfant prev: 1.50 | HMP prev: 1.50 | Mother prev: 1.50\n",
      "\n",
      "=============================================================================\n",
      "Subject: C01840, Species: Bacteroides_vulgatus_57955\n",
      "\n",
      "MI Day 1 to day 4\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 1 to day 289\n",
      "Site: 435590.9.peg.3037,NC_009614,3742405\t| Vartype: 1D\n",
      "\t0 38 45 53 | 0.00 -> 0.85\n",
      "\tInfant prev: 0.31 | HMP prev: 0.43 | Mother prev: 0.22\n",
      "\n",
      "MI Day 1 to day 7\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 1 to day 21\n",
      "NO CHANGE\n",
      "\n",
      "II Day 4 to day 7\n",
      "Site: 435590.9.peg.1813,NC_009614,2242607\t| Vartype: 1D\n",
      "\t28 155 69 83 | 0.18 -> 0.83\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: 0.00\n",
      "Site: 435590.9.peg.3172,NC_009614,3888045\t| Vartype: 1D\n",
      "\t20 105 42 48 | 0.19 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1932,NC_009614,2435682\t| Vartype: 2D\n",
      "\t25 147 52 57 | 0.17 -> 0.91\n",
      "\tInfant prev: 0.39 | HMP prev: 0.15 | Mother prev: 0.28\n",
      "\n",
      "II Day 4 to day 21\n",
      "Site: 435590.9.peg.1813,NC_009614,2242607\t| Vartype: 1D\n",
      "\t28 155 31 31 | 0.18 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: 0.00\n",
      "Site: 435590.9.peg.3172,NC_009614,3888045\t| Vartype: 1D\n",
      "\t20 105 13 13 | 0.19 -> 1.00\n",
      "\tInfant prev: 1.50 | HMP prev: 1.50 | Mother prev: 1.50\n",
      "Site: 435590.9.peg.1932,NC_009614,2435682\t| Vartype: 2D\n",
      "\t25 147 30 30 | 0.17 -> 1.00\n",
      "\tInfant prev: 0.39 | HMP prev: 0.15 | Mother prev: 0.28\n",
      "\n",
      "II Day 4 to day 289\n",
      "Site: 435590.9.peg.1813,NC_009614,2242607\t| Vartype: 1D\n",
      "\t28 155 76 76 | 0.18 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: 0.00\n",
      "Site: 435590.9.peg.3037,NC_009614,3742405\t| Vartype: 1D\n",
      "\t0 146 45 53 | 0.00 -> 0.85\n",
      "\tInfant prev: 0.31 | HMP prev: 0.43 | Mother prev: 0.22\n",
      "Site: 435590.9.peg.3172,NC_009614,3888045\t| Vartype: 1D\n",
      "\t20 105 57 57 | 0.19 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1932,NC_009614,2435682\t| Vartype: 2D\n",
      "\t25 147 51 51 | 0.17 -> 1.00\n",
      "\tInfant prev: 0.39 | HMP prev: 0.15 | Mother prev: 0.28\n",
      "\n",
      "II Day 7 to day 21\n",
      "NO CHANGE\n",
      "\n",
      "II Day 7 to day 289\n",
      "Site: 435590.9.peg.3037,NC_009614,3742405\t| Vartype: 1D\n",
      "\t0 58 45 53 | 0.00 -> 0.85\n",
      "\tInfant prev: 0.31 | HMP prev: 0.43 | Mother prev: 0.22\n",
      "\n",
      "II Day 21 to day 289\n",
      "Site: 435590.9.peg.3037,NC_009614,3742405\t| Vartype: 1D\n",
      "\t0 20 45 53 | 0.00 -> 0.85\n",
      "\tInfant prev: 0.31 | HMP prev: 0.43 | Mother prev: 0.22\n",
      "\n",
      "=============================================================================\n",
      "Subject: 20, Species: Bacteroides_vulgatus_57955\n",
      "\n",
      "MI Day 2 to day 3\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 435590.9.peg.1738,NC_009614,2160159\t| Vartype: 2D\n",
      "\t0 63 138 138 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.13 | HMP prev: 0.14 | Mother prev: 0.12\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 435590.9.peg.1738,NC_009614,2160159\t| Vartype: 2D\n",
      "\t0 63 42 42 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.13 | HMP prev: 0.14 | Mother prev: 0.12\n",
      "\n",
      "II Day 3 to day 366\n",
      "Site: 435590.9.peg.1738,NC_009614,2160159\t| Vartype: 2D\n",
      "\t0 88 138 138 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.13 | HMP prev: 0.14 | Mother prev: 0.12\n",
      "\n",
      "II Day 3 to day 122\n",
      "Site: 435590.9.peg.1738,NC_009614,2160159\t| Vartype: 2D\n",
      "\t0 88 42 42 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.13 | HMP prev: 0.14 | Mother prev: 0.12\n",
      "Site: 435590.9.peg.1906,NC_009614,2388971\t| Vartype: 4D\n",
      "\t53 58 4 28 | 0.91 -> 0.14\n",
      "\tInfant prev: 0.79 | HMP prev: 0.77 | Mother prev: 0.71\n",
      "\n",
      "II Day 122 to day 366\n",
      "NO CHANGE\n",
      "\n",
      "=============================================================================\n",
      "Subject: 10, Species: Bacteroides_xylanisolvens_57185\n",
      "\n",
      "MI Day 2 to day 3\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 556258.5.peg.4082,NZ_EQ973261,14239\t| Vartype: 1D\n",
      "\t0 25 130 130 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "II Day 3 to day 366\n",
      "Site: 556258.5.peg.4082,NZ_EQ973261,14239\t| Vartype: 1D\n",
      "\t0 95 130 130 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "=============================================================================\n",
      "Subject: 1, Species: Bacteroides_fragilis_54507\n",
      "\n",
      "MI Day 2 to day 3\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 2 to day 366\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 1339327.3.peg.2283,JGDJ01000171,5687\t| Vartype: 1D\n",
      "\t3 53 373 377 | 0.06 -> 0.99\n",
      "\tInfant prev: 0.01 | HMP prev: 0.01 | Mother prev: 0.02\n",
      "\n",
      "II Day 3 to day 366\n",
      "NO CHANGE\n",
      "\n",
      "II Day 3 to day 122\n",
      "Site: 1339327.3.peg.2283,JGDJ01000171,5687\t| Vartype: 1D\n",
      "\t23 402 373 377 | 0.06 -> 0.99\n",
      "\tInfant prev: 0.01 | HMP prev: 0.01 | Mother prev: 0.02\n",
      "\n",
      "II Day 122 to day 366\n",
      "Site: 1339327.3.peg.2283,JGDJ01000171,5687\t| Vartype: 1D\n",
      "\t373 377 1 19 | 0.99 -> 0.05\n",
      "\tInfant prev: 0.99 | HMP prev: 0.99 | Mother prev: 0.98\n",
      "\n",
      "=============================================================================\n",
      "Subject: 59, Species: Bifidobacterium_adolescentis_56815\n",
      "\n",
      "MI Day 2 to day 3\n",
      "REPLACEMENT\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 592977.3.peg.642,JGZQ01000005,14361\t| Vartype: 1D\n",
      "\t3 38 37 37 | 0.08 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.860,JGZQ01000006,69020\t| Vartype: 1D\n",
      "\t7 44 66 66 | 0.16 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.1216,JGZQ01000008,284119\t| Vartype: 2D\n",
      "\t0 21 46 46 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.1129,JGZQ01000008,186577\t| Vartype: 4D\n",
      "\t23 28 0 38 | 0.82 -> 0.00\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 592977.3.peg.39,JGZQ01000001,53358\t| Vartype: 1D\n",
      "\t0 42 272 288 | 0.00 -> 0.94\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.1732,JGZQ01000009,58203\t| Vartype: 4D\n",
      "\t0 33 186 192 | 0.00 -> 0.97\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "II Day 3 to day 366\n",
      "REPLACEMENT\n",
      "\n",
      "II Day 3 to day 122\n",
      "REPLACEMENT\n",
      "\n",
      "II Day 122 to day 366\n",
      "Site: 592977.3.peg.1705,JGZQ01000009,29026\t| Vartype: 4D\n",
      "\t0 209 58 58 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.642,JGZQ01000005,14361\t| Vartype: 1D\n",
      "\t0 296 37 37 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.860,JGZQ01000006,69020\t| Vartype: 1D\n",
      "\t0 302 66 66 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.1216,JGZQ01000008,284119\t| Vartype: 2D\n",
      "\t0 176 46 46 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 592977.3.peg.39,JGZQ01000001,53358\t| Vartype: 1D\n",
      "\t272 288 0 46 | 0.94 -> 0.00\n",
      "\tInfant prev: 1.50 | HMP prev: 1.50 | Mother prev: 1.50\n",
      "Site: 592977.3.peg.1129,JGZQ01000008,186577\t| Vartype: 4D\n",
      "\t225 230 0 38 | 0.98 -> 0.00\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "Site: 592977.3.peg.1732,JGZQ01000009,58203\t| Vartype: 4D\n",
      "\t186 192 0 48 | 0.97 -> 0.00\n",
      "\tInfant prev: 1.50 | HMP prev: 1.50 | Mother prev: 1.50\n",
      "\n",
      "=============================================================================\n",
      "Subject: M0808, Species: Bacteroides_vulgatus_57955\n",
      "\n",
      "MI Day -92 to day 14\n",
      "NO CHANGE\n",
      "\n",
      "MI Day -92 to day 61\n",
      "Site: 435590.9.peg.2142,NC_009614,2683430\t| Vartype: 2D\n",
      "\t1 89 27 28 | 0.01 -> 0.96\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1738,NC_009614,2159955\t| Vartype: 1D\n",
      "\t0 84 32 34 | 0.00 -> 0.94\n",
      "\tInfant prev: 0.00 | HMP prev: 0.00 | Mother prev: 0.01\n",
      "Site: 435590.9.peg.2509,NC_009614,3117210\t| Vartype: 4D\n",
      "\t0 71 34 37 | 0.00 -> 0.92\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.47,NC_009614,64759\t| Vartype: 4D\n",
      "\t0 96 46 52 | 0.00 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.2142,NC_009614,2683119\t| Vartype: 4D\n",
      "\t55 55 2 21 | 1.00 -> 0.10\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "Site: 435590.9.peg.2119,NC_009614,2654453\t| Vartype: 4D\n",
      "\t58 58 1 38 | 1.00 -> 0.03\n",
      "\tInfant prev: 0.82 | HMP prev: 0.97 | Mother prev: 0.85\n",
      "Site: 435590.9.peg.4057,NC_009614,4863932\t| Vartype: 1D\n",
      "\t72 72 0 12 | 1.00 -> 0.00\n",
      "\tInfant prev: 0.99 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "\n",
      "MI Day 0 to day 14\n",
      "NO CHANGE\n",
      "\n",
      "MI Day 0 to day 61\n",
      "Site: 435590.9.peg.2142,NC_009614,2683430\t| Vartype: 2D\n",
      "\t0 53 27 28 | 0.00 -> 0.96\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1738,NC_009614,2159955\t| Vartype: 1D\n",
      "\t0 47 32 34 | 0.00 -> 0.94\n",
      "\tInfant prev: 0.00 | HMP prev: 0.00 | Mother prev: 0.01\n",
      "Site: 435590.9.peg.2509,NC_009614,3117210\t| Vartype: 4D\n",
      "\t1 52 34 37 | 0.02 -> 0.92\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.47,NC_009614,64759\t| Vartype: 4D\n",
      "\t0 46 46 52 | 0.00 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.2142,NC_009614,2683119\t| Vartype: 4D\n",
      "\t50 51 2 21 | 0.98 -> 0.10\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "Site: 435590.9.peg.2119,NC_009614,2654453\t| Vartype: 4D\n",
      "\t45 45 1 38 | 1.00 -> 0.03\n",
      "\tInfant prev: 0.82 | HMP prev: 0.97 | Mother prev: 0.85\n",
      "Site: 435590.9.peg.4057,NC_009614,4863932\t| Vartype: 1D\n",
      "\t46 46 0 12 | 1.00 -> 0.00\n",
      "\tInfant prev: 0.99 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "\n",
      "II Day 14 to day 61\n",
      "Site: 435590.9.peg.2142,NC_009614,2683430\t| Vartype: 2D\n",
      "\t0 222 27 28 | 0.00 -> 0.96\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1738,NC_009614,2159955\t| Vartype: 1D\n",
      "\t0 272 32 34 | 0.00 -> 0.94\n",
      "\tInfant prev: 0.00 | HMP prev: 0.00 | Mother prev: 0.01\n",
      "Site: 435590.9.peg.2509,NC_009614,3117210\t| Vartype: 4D\n",
      "\t0 294 34 37 | 0.00 -> 0.92\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.47,NC_009614,64759\t| Vartype: 4D\n",
      "\t0 400 46 52 | 0.00 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.2142,NC_009614,2683119\t| Vartype: 4D\n",
      "\t233 233 2 21 | 1.00 -> 0.10\n",
      "\tInfant prev: 1.00 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "Site: 435590.9.peg.2119,NC_009614,2654453\t| Vartype: 4D\n",
      "\t269 270 1 38 | 1.00 -> 0.03\n",
      "\tInfant prev: 0.82 | HMP prev: 0.97 | Mother prev: 0.85\n",
      "Site: 435590.9.peg.4057,NC_009614,4863932\t| Vartype: 1D\n",
      "\t213 213 0 12 | 1.00 -> 0.00\n",
      "\tInfant prev: 0.99 | HMP prev: 1.50 | Mother prev: 0.99\n",
      "\n",
      "MI Day 92 to day 61\n",
      "Site: 435590.9.peg.2142,NC_009614,2683430\t| Vartype: 2D\n",
      "\t1 45 27 28 | 0.02 -> 0.96\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.1738,NC_009614,2159955\t| Vartype: 1D\n",
      "\t1 49 32 34 | 0.02 -> 0.94\n",
      "\tInfant prev: 0.00 | HMP prev: 0.00 | Mother prev: 0.01\n",
      "Site: 435590.9.peg.2509,NC_009614,3117210\t| Vartype: 4D\n",
      "\t0 50 34 37 | 0.00 -> 0.92\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 435590.9.peg.47,NC_009614,64759\t| Vartype: 4D\n",
      "\t5 53 46 52 | 0.09 -> 0.88\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "MI Day 92 to day 14\n",
      "NO CHANGE\n",
      "\n",
      "=============================================================================\n",
      "Subject: 5, Species: Bacteroides_fragilis_54507\n",
      "\n",
      "MI Day 2 to day 3\n",
      "Site: 1339327.3.peg.4212,JGDJ01000260,9849\t| Vartype: 1D\n",
      "\t1 26 385 385 | 0.04 -> 1.00\n",
      "\tInfant prev: 0.03 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "\n",
      "MI Day 2 to day 122\n",
      "Site: 1339327.3.peg.4212,JGDJ01000260,9849\t| Vartype: 1D\n",
      "\t1 26 105 105 | 0.04 -> 1.00\n",
      "\tInfant prev: 0.03 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "Site: 1339327.3.peg.2899,JGDJ01000204,3609\t| Vartype: 4D\n",
      "\t0 30 144 144 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "MI Day 2 to day 366\n",
      "Site: 1339327.3.peg.4212,JGDJ01000260,9849\t| Vartype: 1D\n",
      "\t1 26 47 47 | 0.04 -> 1.00\n",
      "\tInfant prev: 0.03 | HMP prev: -0.50 | Mother prev: 0.03\n",
      "Site: 1339327.3.peg.3412,JGDJ01000232,16279\t| Vartype: 1D\n",
      "\t0 32 55 55 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 1339327.3.peg.2899,JGDJ01000204,3609\t| Vartype: 4D\n",
      "\t0 30 34 34 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 1339327.3.peg.3553,JGDJ01000241,5833\t| Vartype: 1D\n",
      "\t0 44 64 64 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 1339327.3.peg.887,JGDJ01000131,22145\t| Vartype: 1D\n",
      "\t0 22 52 52 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "II Day 3 to day 122\n",
      "NO CHANGE\n",
      "\n",
      "II Day 3 to day 366\n",
      "Site: 1339327.3.peg.3412,JGDJ01000232,16279\t| Vartype: 1D\n",
      "\t0 382 55 55 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 1339327.3.peg.3553,JGDJ01000241,5833\t| Vartype: 1D\n",
      "\t0 367 64 64 | 0.00 -> 1.00\n",
      "\tInfant prev: -0.50 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "Site: 1339327.3.peg.887,JGDJ01000131,22145\t| Vartype: 1D\n",
      "\t1 527 52 52 | 0.00 -> 1.00\n",
      "\tInfant prev: 0.01 | HMP prev: -0.50 | Mother prev: -0.50\n",
      "\n",
      "=============================================================================\n"
     ]
    }
   ],
   "source": [
    "num_host_species_pairs = 0\n",
    "num_host_species_pairs_by_cohort = defaultdict(int)\n",
    "num_host_species_pairs_by_species = defaultdict(int)\n",
    "\n",
    "for subject in host_snp_changes:\n",
    "    \n",
    "    # Really hacky way to get cohort of this subject\n",
    "    for possible_subject in [subject+'IS',subject+'-I',subject+'-C']:\n",
    "        if possible_subject in subject_sample_map:\n",
    "            for sample in subject_sample_map[possible_subject]:\n",
    "                if sample in sample_cohort_map:\n",
    "                    cohort = sample_cohort_map[sample]\n",
    "    \n",
    "    for species in host_snp_changes[subject]:        \n",
    "        if len(host_snp_changes[subject][species].keys()) > 1:\n",
    "            \n",
    "            # Check if there is at least one infant-infant modification for this host\n",
    "            # AND mother information\n",
    "            # If not, skip/not interested\n",
    "            has_modification = False\n",
    "            has_mother = False\n",
    "            for day1, day2, tp_type in host_snp_changes[subject][species]:\n",
    "                val = host_snp_changes[subject][species][(day1, day2, tp_type)]\n",
    "                if val != 'replacement' and len(val) > 0 and tp_type == 'II':\n",
    "                    has_modification = True\n",
    "                if tp_type == 'MI':\n",
    "                    has_mother = True\n",
    "            \n",
    "            if has_modification == False or has_mother == False:\n",
    "                continue\n",
    "            \n",
    "            num_host_species_pairs += 1\n",
    "            num_host_species_pairs_by_cohort[cohort] += 1\n",
    "            num_host_species_pairs_by_species[species] += 1\n",
    "            \n",
    "            print(\"Subject: %s, Species: %s\" % (subject, species))\n",
    "            \n",
    "            for day1, day2, tp_type in sorted(host_snp_changes[subject][species].keys(), key=lambda x: x[0]):\n",
    "                print('\\n%s Day %i to day %i' % (tp_type, day1, day2))\n",
    "                val = host_snp_changes[subject][species][(day1, day2, tp_type)]\n",
    "                if val == 'replacement':\n",
    "                    print(\"REPLACEMENT\")\n",
    "                elif len(val) == 0:\n",
    "                    print(\"NO CHANGE\")\n",
    "                else:\n",
    "                    snp_change_list = val                    \n",
    "                    snp_change_freqs_list = host_snp_changes_freqs[subject][species][(day1, day2, tp_type)]\n",
    "                    for snp_change, x_snp_change_freqs in zip(snp_change_list, snp_change_freqs_list):\n",
    "                        gene_id, contig, location, vartype, a1, d1, a2, d2 = snp_change\n",
    "                        vartype2, freq_dict, opp_dict = x_snp_change_freqs\n",
    "                        if vartype != vartype2:\n",
    "                            print(\"Weird\")\n",
    "                        print(\"Site: %s,%s,%s\\t| Vartype: %s\" % (gene_id, contig, location, vartype))\n",
    "                        print(\"\\t%i %i %i %i | %.02f -> %.02f\" % (a1, d1, a2, d2, float(a1)/d1, float(a2)/d2))\n",
    "                        print(\"\\tInfant prev: %.02f | HMP prev: %.02f | Mother prev: %.02f\" % (freq_dict['nonpremie'], freq_dict['hmp'], freq_dict['mother']))\n",
    "            print(\"\\n=============================================================================\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "defaultdict(<type 'int'>, {'yassour': 1, 'shao': 2, 'backhed': 7})\n",
      "defaultdict(<type 'int'>, {'Bacteroides_fragilis_54507': 2, 'Bifidobacterium_adolescentis_56815': 1, 'Bacteroides_xylanisolvens_57185': 1, 'Parabacteroides_distasonis_56985': 1, 'Bifidobacterium_bifidum_55065': 1, 'Bacteroides_vulgatus_57955': 4})\n"
     ]
    }
   ],
   "source": [
    "print(num_host_species_pairs)\n",
    "print(num_host_species_pairs_by_cohort)\n",
    "print(num_host_species_pairs_by_species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on Lactobacillus_paracasei_55666....\n",
      "Working on Leclercia_adecarboxylata_62497....\n",
      "Working on Bacteroides_sartorii_54642....\n",
      "Working on Megamonas_hypermegale_57114....\n",
      "Working on Bacteroides_intestinalis_61596....\n",
      "Weird - present\n",
      "Working on Bacteroides_uniformis_57318....\n",
      "Weird - pangenome\n",
      "Weird - present\n",
      "Weird - present\n",
      "Working on Bifidobacterium_bifidum_55065....\n",
      "Working on Clostridium_bolteae_57158....\n",
      "Working on Streptococcus_gallolyticus_57748....\n",
      "Working on Clostridium_paraputrificum_59909....\n",
      "Working on Subdoligranulum_sp_62068....\n",
      "Working on Enterococcus_faecium_56710....\n",
      "Working on Streptococcus_lutetiensis_58501....\n",
      "Working on Phascolarctobacterium_succinatutens_61948....\n",
      "Working on Faecalibacterium_prausnitzii_62201....\n",
      "Working on Faecalibacterium_prausnitzii_61481....\n",
      "Working on Streptococcus_anginosus_58223....\n",
      "Working on Blautia_wexlerae_56130....\n",
      "Working on Dorea_longicatena_61473....\n",
      "Working on Staphylococcus_warneri_58053....\n",
      "Working on Enterococcus_faecalis_55915....\n",
      "Working on Streptococcus_salivarius_58037....\n",
      "Working on Alistipes_indistinctus_62207....\n",
      "Working on Clostridiales_bacterium_52743....\n",
      "Working on Streptococcus_thermophilus_54772....\n",
      "Working on Clostridium_butyricum_56361....\n",
      "Working on Oscillibacter_sp_60799....\n",
      "Working on Erysipelotrichaceae_bacterium_56596....\n",
      "Working on Ruminococcus_torques_62045....\n",
      "Working on Roseburia_intestinalis_56239....\n",
      "Working on Lactobacillus_gasseri_55553....\n",
      "Working on Clostridiales_bacterium_56470....\n",
      "Working on Lachnospiraceae_bacterium_51870....\n",
      "Working on Collinsella_aerofaciens_61484....\n",
      "Working on Bacteroides_stercoris_56735....\n",
      "Working on Bifidobacterium_scardovii_44511....\n",
      "Working on Staphylococcus_aureus_56630....\n",
      "Working on Propionibacterium_acnes_57367....\n",
      "Working on Bacteroides_clarus_62282....\n",
      "Working on Barnesiella_intestinihominis_62208....\n",
      "Working on Enterobacter_asburiae_58142....\n",
      "Working on Morganella_morganii_57363....\n",
      "Working on Oscillospiraceae_bacterium_54867....\n",
      "Working on Bacteroides_eggerthii_54457....\n",
      "Working on Ruminococcus_obeum_61472....\n",
      "Working on Eggerthella_lenta_56463....\n",
      "Working on Klebsiella_oxytoca_57801....\n",
      "Working on Bifidobacterium_longum_57796....\n",
      "Working on Collinsella_sp_62205....\n",
      "Working on Acidaminococcus_intestini_54097....\n",
      "Working on Lactobacillus_rhamnosus_57549....\n",
      "Working on Veillonella_atypica_58169....\n",
      "Working on Bacteroides_faecis_58503....\n",
      "Working on Eubacterium_rectale_56927....\n",
      "Working on Bacteroides_acidifaciens_59693....\n",
      "Working on Clostridium_sordellii_57678....\n",
      "Working on Anaerostipes_hadrus_55206....\n",
      "Working on Pantoea_sp_58264....\n",
      "Working on Clostridium_difficile_57608....\n",
      "Working on Ruminococcus_sp_55468....\n",
      "Working on Klebsiella_oxytoca_54123....\n",
      "Working on Alistipes_putredinis_61533....\n",
      "Weird - present\n",
      "Weird - present\n",
      "Working on Enterobacter_asburiae_57214....\n",
      "Working on Eubacterium_hallii_61477....\n",
      "Working on Citrobacter_freundii_56776....\n",
      "Working on Bacteroides_coprocola_61586....\n",
      "Working on Enterococcus_avium_44557....\n",
      "Working on Citrobacter_koseri_61132....\n",
      "Working on Burkholderiales_bacterium_56577....\n",
      "Working on Ruminococcus_bromii_62047....\n",
      "Working on Klebsiella_oxytoca_56762....\n",
      "Working on Enterococcus_faecalis_56297....\n",
      "Working on Bifidobacterium_adolescentis_56815....\n",
      "Working on Bacteroides_plebeius_61623....\n",
      "Working on Bacteroides_thetaiotaomicron_56941....\n",
      "Working on Bacteroides_ovatus_58035....\n",
      "Working on Paraprevotella_xylaniphila_62280....\n",
      "Working on Odoribacter_splanchnicus_62174....\n",
      "Working on Streptococcus_parasanguinis_58487....\n",
      "Working on Bacteroides_cellulosilyticus_58046....\n",
      "Working on Enterobacter_cloacae_57303....\n",
      "Working on Klebsiella_pneumoniae_54788....\n",
      "Working on Propionibacterium_avidum_56659....\n",
      "Working on Alistipes_sp_59510....\n",
      "Working on Finegoldia_magna_57293....\n",
      "Working on Veillonella_dispar_61763....\n",
      "Working on Escherichia_coli_58110....\n",
      "Working on Prevotella_buccae_56058....\n",
      "Working on Clostridium_sp_61556....\n",
      "Working on Clostridium_hathewayi_61827....\n",
      "Working on Propionibacterium_sp_59451....\n",
      "Working on Peptostreptococcus_anaerobius_55004....\n",
      "Working on Hafnia_alvei_56921....\n",
      "Working on Lactobacillus_crispatus_56887....\n",
      "Working on Enterobacter_cloacae_60571....\n",
      "Working on Clostridium_nexile_61654....\n",
      "Working on Bifidobacterium_dentium_55719....\n",
      "Working on Enterococcus_casseliflavus_56514....\n",
      "Working on Streptococcus_parasanguinis_62481....\n",
      "Working on Bifidobacterium_breve_57133....\n",
      "Working on Dialister_invisus_61905....\n",
      "Working on Veillonella_sp_62404....\n",
      "Working on Odoribacter_laneus_62216....\n",
      "Working on Bacteroides_caccae_53434....\n",
      "Working on Alistipes_shahii_62199....\n",
      "Working on Parabacteroides_johnsonii_55217....\n",
      "Working on Streptococcus_agalactiae_55593....\n",
      "Working on Streptococcus_salivarius_58024....\n",
      "Working on Roseburia_inulinivorans_61943....\n",
      "Working on Bacteroides_xylanisolvens_57185....\n",
      "Working on Veillonella_parvula_58184....\n",
      "Working on Haemophilus_parainfluenzae_57123....\n",
      "Working on Haemophilus_parainfluenzae_62356....\n",
      "Working on Enterococcus_gallinarum_52312....\n",
      "Working on Negativicoccus_succinicivorans_57482....\n",
      "Working on Ruminococcus_gnavus_57638....\n",
      "Working on Ruminococcus_lactaris_55568....\n",
      "Working on Akkermansia_muciniphila_55290....\n",
      "Working on Clostridium_perfringens_56840....\n",
      "Working on Peptoniphilus_rhinitidis_62391....\n",
      "Working on Bifidobacterium_pseudocatenulatum_57754....\n",
      "Weird - present\n",
      "Weird - pangenome\n",
      "Weird - pangenome\n",
      "Working on Paraprevotella_clara_33712....\n",
      "Working on Clostridium_hathewayi_55515....\n",
      "Working on Proteus_mirabilis_55046....\n",
      "Working on Escherichia_coli_57907....\n",
      "Working on Escherichia_coli_53492....\n",
      "Working on Butyrivibrio_crossotus_61674....\n",
      "Working on Bacteroidales_bacterium_58650....\n",
      "Working on Staphylococcus_haemolyticus_57278....\n",
      "Working on Faecalibacterium_prausnitzii_57453....\n",
      "Working on Klebsiella_pneumoniae_54688....\n",
      "Working on Enterobacter_hormaechei_50416....\n",
      "Working on Roseburia_hominis_61877....\n",
      "Working on Lactobacillus_gasseri_55805....\n",
      "Working on Staphylococcus_sp_59500....\n",
      "Working on Coprococcus_sp_62244....\n",
      "Working on Clostridium_sp_60465....\n",
      "Working on Haemophilus_parainfluenzae_62468....\n",
      "Working on Eubacterium_eligens_61678....\n",
      "Working on Neisseria_subflava_61760....\n",
      "Working on Alistipes_finegoldii_56071....\n",
      "Working on Clostridium_sp_61482....\n",
      "Working on Alistipes_sp_60764....\n",
      "Working on Veillonella_ratti_62429....\n",
      "Working on Clostridium_sp_57418....\n",
      "Working on Eubacterium_siraeum_57634....\n",
      "Weird - present\n",
      "Weird - pangenome\n",
      "Working on Faecalibacterium_cf_62236....\n",
      "Working on Alistipes_onderdonkii_55464....\n",
      "Working on Escherichia_albertii_56276....\n",
      "Working on Actinomyces_urogenitalis_53864....\n",
      "Working on Bacteroides_salyersiae_54873....\n",
      "Working on Corynebacterium_sp_48834....\n",
      "Working on Serratia_marcescens_58205....\n",
      "Working on Streptococcus_vestibularis_56030....\n",
      "Working on Guyana_massiliensis_60772....\n",
      "Working on Coprococcus_comes_61587....\n",
      "Working on Enterococcus_faecium_56947....\n",
      "Working on Butyricimonas_virosa_58742....\n",
      "Working on Enterobacter_cloacae_58148....\n",
      "Working on Clostridium_bartlettii_61535....\n",
      "Working on Ruminococcus_bicirculans_59300....\n",
      "Working on Raoultella_planticola_57814....\n",
      "Weird - pangenome\n",
      "Working on Pseudomonas_aeruginosa_57148....\n",
      "Working on Lactobacillus_fermentum_54035....\n",
      "Working on Lactobacillus_plantarum_54610....\n",
      "Working on Parabacteroides_goldsteinii_56831....\n",
      "Working on Rothia_mucilaginosa_58144....\n",
      "Working on Eubacterium_biforme_61684....\n",
      "Working on Serratia_liquefaciens_58210....\n",
      "Working on Clostridium_baratii_60475....\n",
      "Working on Bacteroides_fragilis_56548....\n",
      "Working on Citrobacter_freundii_56148....\n",
      "Working on Clostridiales_bacterium_59663....\n",
      "Working on Staphylococcus_epidermidis_57523....\n",
      "Working on Bacteroides_vulgatus_57955....\n",
      "Working on Klebsiella_pneumoniae_57337....\n",
      "Working on Phascolarctobacterium_sp_59817....\n",
      "Working on Prevotella_melaninogenica_58075....\n",
      "Working on Bacteroides_fragilis_54507....\n",
      "Weird - present\n",
      "Weird - between\n",
      "Working on Bifidobacterium_animalis_58116....\n",
      "Working on Prevotella_copri_61740....\n",
      "Working on Staphylococcus_hominis_57055....\n",
      "Working on Enterococcus_hirae_55131....\n",
      "Working on Enterobacter_aerogenes_55704....\n",
      "Working on Bacteroides_massiliensis_44749....\n",
      "Working on Streptococcus_salivarius_58022....\n",
      "Working on Sutterella_wadsworthensis_56828....\n",
      "Working on Streptococcus_sp_60086....\n",
      "Working on Escherichia_hermannii_58626....\n",
      "Working on Bifidobacterium_catenulatum_58257....\n",
      "Working on Rothia_mucilaginosa_62109....\n",
      "Working on Varibaculum_cambriense_60384....\n",
      "Working on Bacteroides_finegoldii_57739....\n",
      "Working on Parabacteroides_merdae_56972....\n",
      "Working on Bradyrhizobium_sp_61126....\n",
      "Working on Enterobacter_cloacae_55612....\n",
      "Working on Streptococcus_parasanguinis_62442....\n",
      "Working on Bacteroides_coprophilus_61767....\n",
      "Working on Enterococcus_durans_52177....\n",
      "Working on Veillonella_sp_62611....\n",
      "Working on Pantoea_sp_60700....\n",
      "Working on Citrobacter_braakii_56022....\n",
      "Working on Eubacterium_ventriosum_61474....\n",
      "Working on Finegoldia_magna_61710....\n",
      "Working on Coprobacter_fastidiosus_56550....\n",
      "Working on Streptococcus_parasanguinis_58126....\n",
      "Working on Bacteroides_nordii_55557....\n",
      "Working on Streptococcus_sp_62701....\n",
      "Working on Enterobacter_sp_59869....\n",
      "Working on Enterobacter_cloacae_55011....\n",
      "Working on Veillonella_parvula_57794....\n",
      "Working on Staphylococcus_lugdunensis_54043....\n",
      "Working on Sutterella_wadsworthensis_62218....\n",
      "Working on Clostridium_leptum_61499....\n",
      "Working on Parabacteroides_distasonis_56985....\n",
      "Working on Actinomyces_sp_62581....\n",
      "Working on Anaerococcus_obesiensis_58051....\n"
     ]
    }
   ],
   "source": [
    "# Parallelism\n",
    "\n",
    "custom_cohorts_ordered = ['MI', 'Day 0-Week 1', 'Week 1-Month 1', 'Month 1-Year 1', 'II(other)', 'MM', 'AA']\n",
    "variant_types = ['1D', '2D', '3D', '4D']\n",
    "\n",
    "def custom_cohort_of_sample_pair(custom_cohort_tests, sample_i, sample_j):\n",
    "    for cohort in custom_cohort_tests:\n",
    "        if custom_cohort_tests[cohort](sample_i, sample_j):\n",
    "            return cohort\n",
    "\n",
    "custom_cohort_tests = {}\n",
    "custom_cohort_tests['MI'] = lambda sample_i, sample_j: (sample_i in mother_samples and sample_j in infant_samples_no_olm)\n",
    "# custom_cohort_tests['MI'] = lambda sample_i, sample_j: ((sample_i in mother_samples and sample_j in infant_samples_no_olm) and mi_sample_day_dict[sample_i] >= 0 and mi_sample_day_dict[sample_j] <= 7)\n",
    "custom_cohort_tests['Day 0-Week 1'] = lambda sample_i, sample_j: ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm) and (mi_sample_day_dict[sample_i] >= 0 and mi_sample_day_dict[sample_j] <= 7))\n",
    "custom_cohort_tests['Week 1-Month 1'] = lambda sample_i, sample_j: ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm) and (mi_sample_day_dict[sample_i] >= 7 and mi_sample_day_dict[sample_j] <= 31))\n",
    "custom_cohort_tests['Month 1-Year 1'] = lambda sample_i, sample_j: ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm) and (mi_sample_day_dict[sample_i] >= 31 and mi_sample_day_dict[sample_j] <= 370))\n",
    "custom_cohort_tests['II(other)'] = lambda sample_i, sample_j: ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm))\n",
    "custom_cohort_tests['MM'] = lambda sample_i, sample_j: (sample_i in mother_samples and sample_j in mother_samples)\n",
    "custom_cohort_tests['AA'] = lambda sample_i, sample_j: (sample_i in hmp_samples and sample_j in hmp_samples)\n",
    "\n",
    "# cohort -> gene -> variant type -> count of SNP changes\n",
    "genes_by_cohort = {cohort: {} for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_present_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_between_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_pangenome_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "\n",
    "# cohort -> gene -> count of unique hosts\n",
    "num_host_genes_by_cohort = {cohort: defaultdict(set) for cohort in custom_cohorts_ordered}\n",
    "\n",
    "for species in snp_changes_consecutive:\n",
    "    print(\"Working on %s....\" % species)\n",
    "    genome_ids = midas_db_utils.get_ref_genome_ids(species)\n",
    "    # load the gene descriptions for all genomes coresponding to this speceis:\n",
    "    gene_descriptions=parse_patric.load_patric_gene_descriptions(genome_ids)\n",
    "    \n",
    "    for sample_i, sample_j in snp_changes_consecutive[species]:\n",
    "        val = snp_changes_consecutive[species][(sample_i, sample_j)]\n",
    "        subject_tuple = (sample_subject_map[sample_i], sample_subject_map[sample_j])\n",
    "        custom_cohort = custom_cohort_of_sample_pair(custom_cohort_tests, sample_i, sample_j)\n",
    "        if type(val) == type([]): # not a replacement\n",
    "            \n",
    "            # Store actual changed genes\n",
    "            for gene_id, contig, position, variant_type, A1, D1, A2, D2 in val:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    if desc not in genes_by_cohort[custom_cohort]:\n",
    "                        genes_by_cohort[custom_cohort][desc] = {vartype: 0 for vartype in variant_types}\n",
    "                    genes_by_cohort[custom_cohort][desc][variant_type] += 1\n",
    "                    num_host_genes_by_cohort[custom_cohort][desc].add(subject_tuple)\n",
    "                except:\n",
    "                    print(\"Weird\")\n",
    "                    continue\n",
    "            \n",
    "            # Store present gene null\n",
    "            for gene_id in snp_change_present_gene_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_present_gene_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_present_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - present\")\n",
    "                    continue\n",
    "            \n",
    "            # Store between host null\n",
    "            for gene_id in snp_change_between_host_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_between_host_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_between_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - between\")\n",
    "                    continue\n",
    "            \n",
    "            # Store pangenome null\n",
    "            for gene_id in snp_change_pangenome_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_pangenome_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_pangenome_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - pangenome\")\n",
    "                    continue\n",
    "\n",
    "# pickle.dump(genes_by_cohort, open('%s/genes_by_cohort.pkl' % config.analysis_directory, 'wb'))\n",
    "# pickle.dump(num_host_genes_by_cohort, open('%s/num_host_genes_by_cohort.pkl' % config.analysis_directory, 'wb'))\n",
    "# pickle.dump(genes_by_cohort_present_null, open('%s/genes_by_cohort_present_null.pkl' % config.analysis_directory, 'wb'))\n",
    "# pickle.dump(genes_by_cohort_between_null, open('%s/genes_by_cohort_between_null.pkl' % config.analysis_directory, 'wb'))\n",
    "# pickle.dump(genes_by_cohort_pangenome_null, open('%s/genes_by_cohort_pangenome_null.pkl' % config.analysis_directory, 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "pickle.dump(genes_by_cohort, open('%s/genes_by_cohort_with_vartype.pkl' % config.analysis_directory, 'wb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "genes_by_cohort = pickle.load(open('%s/genes_by_cohort.pkl' % config.analysis_directory, 'rb'))\n",
    "num_host_genes_by_cohort = pickle.load(open('%s/num_host_genes_by_cohort.pkl' % config.analysis_directory, 'rb'))\n",
    "genes_by_cohort_present_null = pickle.load(open('%s/genes_by_cohort_present_null.pkl' % config.analysis_directory, 'rb'))\n",
    "genes_by_cohort_between_null = pickle.load(open('%s/genes_by_cohort_between_null.pkl' % config.analysis_directory, 'rb'))\n",
    "genes_by_cohort_pangenome_null = pickle.load(open('%s/genes_by_cohort_pangenome_null.pkl' % config.analysis_directory, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "for cat in custom_cohorts_ordered:\n",
    "    f = open('%s/%s_snp_change_gene_annotation_v2.tsv' % (config.analysis_directory, cat.replace(' ', '_')), 'w')\n",
    "    f.write('\\t'.join(['count', 'count_by_vartype', 'num_unique_hosts', 'present_null', 'between_null', 'pangenome_null', 'gene']) + '\\n')\n",
    "    for gene, count_dict in sorted(genes_by_cohort[cat].items(), key=lambda x: sum(x[1].values()), reverse=True):\n",
    "        total_count = sum(count_dict.values())\n",
    "        vartype_str = '|'.join([str(genes_by_cohort[cat][gene][vt]) for vt in variant_types])\n",
    "        present_null_count = genes_by_cohort_present_null[cat][gene]\n",
    "        between_null_count = genes_by_cohort_between_null[cat][gene]\n",
    "        pangenome_null_count = genes_by_cohort_pangenome_null[cat][gene]\n",
    "        num_unique_hosts = len(num_host_genes_by_cohort[cat][gene])\n",
    "        if num_unique_hosts > total_count:\n",
    "            print(\"Huh???\")\n",
    "            print(gene)\n",
    "        f.write('\\t'.join([str(x) for x in [total_count, vartype_str, num_unique_hosts, present_null_count, between_null_count, pangenome_null_count, gene]]) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'1D': 0, '2D': 1, '3D': 0, '4D': 0}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes_by_cohort['AA']['1,4-alpha-glucan (glycogen) branching enzyme, GH-13-type (EC 2.4.1.18)']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Working on Lactobacillus_paracasei_55666....\n",
      "Working on Leclercia_adecarboxylata_62497....\n",
      "Working on Bacteroides_sartorii_54642....\n",
      "Working on Megamonas_hypermegale_57114....\n",
      "Working on Bacteroides_intestinalis_61596....\n",
      "Working on Bacteroides_uniformis_57318....\n",
      "Weird - pangenome\n",
      "Working on Bifidobacterium_bifidum_55065....\n",
      "Working on Clostridium_bolteae_57158....\n",
      "Working on Streptococcus_gallolyticus_57748....\n",
      "Working on Clostridium_paraputrificum_59909....\n",
      "Working on Subdoligranulum_sp_62068....\n",
      "Working on Enterococcus_faecium_56710....\n",
      "Working on Streptococcus_lutetiensis_58501....\n",
      "Working on Phascolarctobacterium_succinatutens_61948....\n",
      "Working on Faecalibacterium_prausnitzii_62201....\n",
      "Working on Faecalibacterium_prausnitzii_61481....\n",
      "Working on Streptococcus_anginosus_58223....\n",
      "Working on Blautia_wexlerae_56130....\n",
      "Working on Dorea_longicatena_61473....\n",
      "Working on Staphylococcus_warneri_58053....\n",
      "Working on Enterococcus_faecalis_55915....\n",
      "Working on Streptococcus_salivarius_58037....\n",
      "Working on Alistipes_indistinctus_62207....\n",
      "Working on Clostridiales_bacterium_52743....\n",
      "Working on Streptococcus_thermophilus_54772....\n",
      "Working on Clostridium_butyricum_56361....\n",
      "Working on Oscillibacter_sp_60799....\n",
      "Working on Erysipelotrichaceae_bacterium_56596....\n",
      "Working on Ruminococcus_torques_62045....\n",
      "Working on Roseburia_intestinalis_56239....\n",
      "Working on Lactobacillus_gasseri_55553....\n",
      "Working on Clostridiales_bacterium_56470....\n",
      "Working on Lachnospiraceae_bacterium_51870....\n",
      "Working on Collinsella_aerofaciens_61484....\n",
      "Working on Bacteroides_stercoris_56735....\n",
      "Working on Bifidobacterium_scardovii_44511....\n",
      "Working on Staphylococcus_aureus_56630....\n",
      "Working on Propionibacterium_acnes_57367....\n",
      "Working on Bacteroides_clarus_62282....\n",
      "Working on Barnesiella_intestinihominis_62208....\n",
      "Working on Enterobacter_asburiae_58142....\n",
      "Working on Morganella_morganii_57363....\n",
      "Working on Oscillospiraceae_bacterium_54867....\n",
      "Working on Bacteroides_eggerthii_54457....\n",
      "Working on Ruminococcus_obeum_61472....\n",
      "Working on Eggerthella_lenta_56463....\n",
      "Working on Klebsiella_oxytoca_57801....\n",
      "Working on Bifidobacterium_longum_57796....\n",
      "Working on Collinsella_sp_62205....\n",
      "Working on Acidaminococcus_intestini_54097....\n",
      "Working on Lactobacillus_rhamnosus_57549....\n",
      "Working on Veillonella_atypica_58169....\n",
      "Working on Bacteroides_faecis_58503....\n",
      "Working on Eubacterium_rectale_56927....\n",
      "Working on Bacteroides_acidifaciens_59693....\n",
      "Working on Clostridium_sordellii_57678....\n",
      "Working on Anaerostipes_hadrus_55206....\n",
      "Working on Pantoea_sp_58264....\n",
      "Working on Clostridium_difficile_57608....\n",
      "Working on Ruminococcus_sp_55468....\n",
      "Working on Klebsiella_oxytoca_54123....\n",
      "Working on Alistipes_putredinis_61533....\n",
      "Working on Enterobacter_asburiae_57214....\n",
      "Working on Eubacterium_hallii_61477....\n",
      "Working on Citrobacter_freundii_56776....\n",
      "Working on Bacteroides_coprocola_61586....\n",
      "Working on Enterococcus_avium_44557....\n",
      "Working on Citrobacter_koseri_61132....\n",
      "Working on Burkholderiales_bacterium_56577....\n",
      "Working on Ruminococcus_bromii_62047....\n",
      "Working on Klebsiella_oxytoca_56762....\n",
      "Working on Enterococcus_faecalis_56297....\n",
      "Working on Bifidobacterium_adolescentis_56815....\n",
      "Working on Bacteroides_plebeius_61623....\n",
      "Working on Bacteroides_thetaiotaomicron_56941....\n",
      "Working on Bacteroides_ovatus_58035....\n",
      "Working on Paraprevotella_xylaniphila_62280....\n",
      "Working on Odoribacter_splanchnicus_62174....\n",
      "Working on Streptococcus_parasanguinis_58487....\n",
      "Working on Bacteroides_cellulosilyticus_58046....\n",
      "Working on Enterobacter_cloacae_57303....\n",
      "Working on Klebsiella_pneumoniae_54788....\n",
      "Working on Propionibacterium_avidum_56659....\n",
      "Working on Alistipes_sp_59510....\n",
      "Working on Finegoldia_magna_57293....\n",
      "Working on Veillonella_dispar_61763....\n",
      "Working on Escherichia_coli_58110....\n",
      "Working on Prevotella_buccae_56058....\n",
      "Working on Clostridium_sp_61556....\n",
      "Working on Clostridium_hathewayi_61827....\n",
      "Working on Propionibacterium_sp_59451....\n",
      "Working on Peptostreptococcus_anaerobius_55004....\n",
      "Working on Hafnia_alvei_56921....\n",
      "Working on Lactobacillus_crispatus_56887....\n",
      "Working on Enterobacter_cloacae_60571....\n",
      "Working on Clostridium_nexile_61654....\n",
      "Working on Bifidobacterium_dentium_55719....\n",
      "Working on Enterococcus_casseliflavus_56514....\n",
      "Working on Streptococcus_parasanguinis_62481....\n",
      "Working on Bifidobacterium_breve_57133....\n",
      "Working on Dialister_invisus_61905....\n",
      "Working on Veillonella_sp_62404....\n",
      "Working on Odoribacter_laneus_62216....\n",
      "Working on Bacteroides_caccae_53434....\n",
      "Working on Alistipes_shahii_62199....\n",
      "Working on Parabacteroides_johnsonii_55217....\n",
      "Working on Streptococcus_agalactiae_55593....\n",
      "Working on Streptococcus_salivarius_58024....\n",
      "Working on Roseburia_inulinivorans_61943....\n",
      "Working on Bacteroides_xylanisolvens_57185....\n",
      "Working on Veillonella_parvula_58184....\n",
      "Working on Haemophilus_parainfluenzae_57123....\n",
      "Working on Haemophilus_parainfluenzae_62356....\n",
      "Working on Enterococcus_gallinarum_52312....\n",
      "Working on Negativicoccus_succinicivorans_57482....\n",
      "Working on Ruminococcus_gnavus_57638....\n",
      "Working on Ruminococcus_lactaris_55568....\n",
      "Working on Akkermansia_muciniphila_55290....\n",
      "Working on Clostridium_perfringens_56840....\n",
      "Working on Peptoniphilus_rhinitidis_62391....\n",
      "Working on Bifidobacterium_pseudocatenulatum_57754....\n",
      "Weird - present\n",
      "Weird - pangenome\n",
      "Weird - pangenome\n",
      "Working on Paraprevotella_clara_33712....\n",
      "Working on Clostridium_hathewayi_55515....\n",
      "Working on Proteus_mirabilis_55046....\n",
      "Working on Escherichia_coli_57907....\n",
      "Working on Escherichia_coli_53492....\n",
      "Working on Butyrivibrio_crossotus_61674....\n",
      "Working on Bacteroidales_bacterium_58650....\n",
      "Working on Staphylococcus_haemolyticus_57278....\n",
      "Working on Faecalibacterium_prausnitzii_57453....\n",
      "Working on Klebsiella_pneumoniae_54688....\n",
      "Working on Enterobacter_hormaechei_50416....\n",
      "Working on Roseburia_hominis_61877....\n",
      "Working on Lactobacillus_gasseri_55805....\n",
      "Working on Staphylococcus_sp_59500....\n",
      "Working on Coprococcus_sp_62244....\n",
      "Working on Clostridium_sp_60465....\n",
      "Working on Haemophilus_parainfluenzae_62468....\n",
      "Working on Eubacterium_eligens_61678....\n",
      "Working on Neisseria_subflava_61760....\n",
      "Working on Alistipes_finegoldii_56071....\n",
      "Working on Clostridium_sp_61482....\n",
      "Working on Alistipes_sp_60764....\n",
      "Working on Veillonella_ratti_62429....\n",
      "Working on Clostridium_sp_57418....\n",
      "Working on Eubacterium_siraeum_57634....\n",
      "Working on Faecalibacterium_cf_62236....\n",
      "Working on Alistipes_onderdonkii_55464....\n",
      "Working on Escherichia_albertii_56276....\n",
      "Working on Actinomyces_urogenitalis_53864....\n",
      "Working on Bacteroides_salyersiae_54873....\n",
      "Working on Corynebacterium_sp_48834....\n",
      "Working on Serratia_marcescens_58205....\n",
      "Working on Streptococcus_vestibularis_56030....\n",
      "Working on Guyana_massiliensis_60772....\n",
      "Working on Coprococcus_comes_61587....\n",
      "Working on Enterococcus_faecium_56947....\n",
      "Working on Butyricimonas_virosa_58742....\n",
      "Working on Enterobacter_cloacae_58148....\n",
      "Working on Clostridium_bartlettii_61535....\n",
      "Working on Ruminococcus_bicirculans_59300....\n",
      "Working on Raoultella_planticola_57814....\n",
      "Weird - pangenome\n",
      "Working on Pseudomonas_aeruginosa_57148....\n",
      "Working on Lactobacillus_fermentum_54035....\n",
      "Working on Lactobacillus_plantarum_54610....\n",
      "Working on Parabacteroides_goldsteinii_56831....\n",
      "Working on Rothia_mucilaginosa_58144....\n",
      "Working on Eubacterium_biforme_61684....\n",
      "Working on Serratia_liquefaciens_58210....\n",
      "Working on Clostridium_baratii_60475....\n",
      "Working on Bacteroides_fragilis_56548....\n",
      "Working on Citrobacter_freundii_56148....\n",
      "Working on Clostridiales_bacterium_59663....\n",
      "Working on Staphylococcus_epidermidis_57523....\n",
      "Working on Bacteroides_vulgatus_57955....\n",
      "Working on Klebsiella_pneumoniae_57337....\n",
      "Working on Phascolarctobacterium_sp_59817....\n",
      "Working on Prevotella_melaninogenica_58075....\n",
      "Working on Bacteroides_fragilis_54507....\n",
      "Weird - present\n",
      "Working on Bifidobacterium_animalis_58116....\n",
      "Working on Prevotella_copri_61740....\n",
      "Working on Staphylococcus_hominis_57055....\n",
      "Working on Enterococcus_hirae_55131....\n",
      "Working on Enterobacter_aerogenes_55704....\n",
      "Working on Bacteroides_massiliensis_44749....\n",
      "Working on Streptococcus_salivarius_58022....\n",
      "Working on Sutterella_wadsworthensis_56828....\n",
      "Working on Streptococcus_sp_60086....\n",
      "Working on Escherichia_hermannii_58626....\n",
      "Working on Bifidobacterium_catenulatum_58257....\n",
      "Working on Rothia_mucilaginosa_62109....\n",
      "Working on Varibaculum_cambriense_60384....\n",
      "Working on Bacteroides_finegoldii_57739....\n",
      "Working on Parabacteroides_merdae_56972....\n",
      "Working on Bradyrhizobium_sp_61126....\n",
      "Working on Enterobacter_cloacae_55612....\n",
      "Working on Streptococcus_parasanguinis_62442....\n",
      "Working on Bacteroides_coprophilus_61767....\n",
      "Working on Enterococcus_durans_52177....\n",
      "Working on Veillonella_sp_62611....\n",
      "Working on Pantoea_sp_60700....\n",
      "Working on Citrobacter_braakii_56022....\n",
      "Working on Eubacterium_ventriosum_61474....\n",
      "Working on Finegoldia_magna_61710....\n",
      "Working on Coprobacter_fastidiosus_56550....\n",
      "Working on Streptococcus_parasanguinis_58126....\n",
      "Working on Bacteroides_nordii_55557....\n",
      "Working on Streptococcus_sp_62701....\n",
      "Working on Enterobacter_sp_59869....\n",
      "Working on Enterobacter_cloacae_55011....\n",
      "Working on Veillonella_parvula_57794....\n",
      "Working on Staphylococcus_lugdunensis_54043....\n",
      "Working on Sutterella_wadsworthensis_62218....\n",
      "Working on Clostridium_leptum_61499....\n",
      "Working on Parabacteroides_distasonis_56985....\n",
      "Working on Actinomyces_sp_62581....\n",
      "Working on Anaerococcus_obesiensis_58051....\n"
     ]
    }
   ],
   "source": [
    "# Parallelism - ok just make one big II category\n",
    "\n",
    "custom_cohorts_ordered = ['II']\n",
    "variant_types = ['1D', '2D', '3D', '4D']\n",
    "\n",
    "def custom_cohort_of_sample_pair(custom_cohort_tests, sample_i, sample_j):\n",
    "    for cohort in custom_cohort_tests:\n",
    "        if custom_cohort_tests[cohort](sample_i, sample_j):\n",
    "            return cohort\n",
    "\n",
    "custom_cohort_tests = {}\n",
    "custom_cohort_tests['II'] = lambda sample_i, sample_j: ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm))\n",
    "\n",
    "# cohort -> gene -> variant type -> count of SNP changes\n",
    "genes_by_cohort = {cohort: {} for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_present_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_between_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "genes_by_cohort_pangenome_null = {cohort: defaultdict(int) for cohort in custom_cohorts_ordered}\n",
    "\n",
    "# cohort -> gene -> count of unique hosts\n",
    "num_host_genes_by_cohort = {cohort: defaultdict(set) for cohort in custom_cohorts_ordered}\n",
    "\n",
    "for species in snp_changes_consecutive:\n",
    "    print(\"Working on %s....\" % species)\n",
    "    genome_ids = midas_db_utils.get_ref_genome_ids(species)\n",
    "    # load the gene descriptions for all genomes coresponding to this speceis:\n",
    "    gene_descriptions=parse_patric.load_patric_gene_descriptions(genome_ids)\n",
    "    \n",
    "    for sample_i, sample_j in snp_changes_consecutive[species]:\n",
    "        val = snp_changes_consecutive[species][(sample_i, sample_j)]\n",
    "        subject_tuple = (sample_subject_map[sample_i], sample_subject_map[sample_j])\n",
    "        if not ((sample_i in infant_samples_no_olm and sample_j in infant_samples_no_olm)):\n",
    "            continue\n",
    "        else:\n",
    "            custom_cohort = 'II'\n",
    "        if type(val) == type([]): # not a replacement\n",
    "            \n",
    "            # Store actual changed genes\n",
    "            for gene_id, contig, position, variant_type, A1, D1, A2, D2 in val:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    if desc not in genes_by_cohort[custom_cohort]:\n",
    "                        genes_by_cohort[custom_cohort][desc] = {vartype: 0 for vartype in variant_types}\n",
    "                    genes_by_cohort[custom_cohort][desc][variant_type] += 1\n",
    "                    num_host_genes_by_cohort[custom_cohort][desc].add(subject_tuple)\n",
    "                except:\n",
    "                    print(\"Weird\")\n",
    "                    continue\n",
    "            \n",
    "            # Store present gene null\n",
    "            for gene_id in snp_change_present_gene_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_present_gene_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_present_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - present\")\n",
    "                    continue\n",
    "            \n",
    "            # Store between host null\n",
    "            for gene_id in snp_change_between_host_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_between_host_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_between_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - between\")\n",
    "                    continue\n",
    "            \n",
    "            # Store pangenome null\n",
    "            for gene_id in snp_change_pangenome_null[species][(sample_i, sample_j)]:\n",
    "                try:\n",
    "                    desc = gene_descriptions[gene_id]\n",
    "                    count = snp_change_pangenome_null[species][(sample_i, sample_j)][gene_id]\n",
    "                    genes_by_cohort_pangenome_null[custom_cohort][desc] += count\n",
    "                except:\n",
    "                    print(\"Weird - pangenome\")\n",
    "                    continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "genes_by_cohort = pickle.load(open('%s/genes_by_cohort_with_vartype.pkl' % config.analysis_directory, 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Also include number of unique hosts\n",
    "    \n",
    "for cat in custom_cohorts_ordered:\n",
    "    f = open('%s/%s_snp_change_gene_annotation_v2.tsv' % (config.analysis_directory, cat.replace(' ', '_')), 'w')\n",
    "    f.write('\\t'.join(['count', 'count_by_vartype', 'num_unique_hosts', 'present_null', 'between_null', 'pangenome_null', 'gene']) + '\\n')\n",
    "    for gene, count_dict in sorted(genes_by_cohort[cat].items(), key=lambda x: sum(x[1].values()), reverse=True):\n",
    "        total_count = sum(count_dict.values())\n",
    "        vartype_str = '|'.join([str(genes_by_cohort[cat][gene][vt]) for vt in variant_types])\n",
    "        present_null_count = genes_by_cohort_present_null[cat][gene]\n",
    "        between_null_count = genes_by_cohort_between_null[cat][gene]\n",
    "        pangenome_null_count = genes_by_cohort_pangenome_null[cat][gene]\n",
    "        num_unique_hosts = len(num_host_genes_by_cohort[cat][gene])\n",
    "        if num_unique_hosts > total_count:\n",
    "            print(\"Huh???\")\n",
    "            print(gene)\n",
    "        f.write('\\t'.join([str(x) for x in [total_count, vartype_str, num_unique_hosts, present_null_count, between_null_count, pangenome_null_count, gene]]) + '\\n')\n",
    "    f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "# Here's what imma plot:\n",
    "# For every gene that changes in at least two unique hosts in infants,\n",
    "# and which changes a greater number of times than all three null expectations,\n",
    "# get proportion increase in occurrence from present null as\n",
    "# (actual - present_null) / (present_null)\n",
    "# and order by proportion increase..?\n",
    "\n",
    "gene_prop_increase = {}\n",
    "\n",
    "for gene, count in genes_by_cohort[cat].items():\n",
    "    present_null_count = genes_by_cohort_present_null[cat][gene]\n",
    "    between_null_count = genes_by_cohort_between_null[cat][gene]\n",
    "    pangenome_null_count = genes_by_cohort_pangenome_null[cat][gene]\n",
    "    num_unique_hosts = len(num_host_genes_by_cohort[cat][gene])\n",
    "    if num_unique_hosts >= 2 and count > present_null_count and count > between_null_count and count > pangenome_null_count:\n",
    "        gene_prop_increase[gene] = float(count - present_null_count)/float(present_null_count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "for gene, prop_increase in sorted(gene_prop_increase.items(), key=lambda x: x[1], reverse=True):\n",
    "    print(gene)\n",
    "    present_null_count = genes_by_cohort_present_null[cat][gene]\n",
    "    count = genes_by_cohort[cat][gene]\n",
    "    print('Present null: %.02f\\t| Actual: %i\\t| Num hosts: %i\\t| Increase: %.02fx' % (present_null_count, count, prop_increase) + '\\n')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
